<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../img/favicon.ico" />
    <title>biopipen.ns.rnaseq - biopipen</title>
    <link rel="stylesheet" href="../../../css/theme.css" />
    <link rel="stylesheet" href="../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="//use.fontawesome.com/releases/v5.8.1/css/all.css" rel="stylesheet" />
        <link href="//use.fontawesome.com/releases/v5.8.1/css/v4-shims.css" rel="stylesheet" />
        <link href="../../../css/mkapi-common.css" rel="stylesheet" />
        <link href="../../../style.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "biopipen.ns.rnaseq";
        var mkdocs_page_input_path = "api/source/biopipen.ns.rnaseq.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.configure({ ignoreUnescapedHTML: true }); hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../..">
          <img src="../../../img/logo.png" class="logo" alt="Logo"/>
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="nav-scrollable-wrapper">
      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Pipelines</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../../pipelines/cnvkit_pipeline/">CNVkit Pipeline</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../../pipelines/cellranger_pipeline/">CellRanger Pipeline</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">API</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.utils</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils/">biopipen.utils</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.gene/">biopipen.utils.gene</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.reference/">biopipen.utils.reference</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.misc/">biopipen.utils.misc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.vcf/">biopipen.utils.vcf</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.core</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core/">biopipen.core</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.proc/">biopipen.core.proc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.filters/">biopipen.core.filters</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.testing/">biopipen.core.testing</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.config/">biopipen.core.config</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.defaults/">biopipen.core.defaults</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.ns</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns/">biopipen.ns</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.gene/">biopipen.ns.gene</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cellranger/">biopipen.ns.cellranger</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.scrna_metabolic_landscape/">biopipen.ns.scrna_metabolic_landscape</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.protein/">biopipen.ns.protein</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.delim/">biopipen.ns.delim</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.rnaseq/">biopipen.ns.rnaseq</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.plot/">biopipen.ns.plot</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.tcgamaf/">biopipen.ns.tcgamaf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.regulatory/">biopipen.ns.regulatory</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.stats/">biopipen.ns.stats</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cellranger_pipeline/">biopipen.ns.cellranger_pipeline</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.bed/">biopipen.ns.bed</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnv/">biopipen.ns.cnv</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.snp/">biopipen.ns.snp</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.web/">biopipen.ns.web</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.misc/">biopipen.ns.misc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnvkit/">biopipen.ns.cnvkit</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnvkit_pipeline/">biopipen.ns.cnvkit_pipeline</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.tcr/">biopipen.ns.tcr</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.scrna/">biopipen.ns.scrna</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.vcf/">biopipen.ns.vcf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.gsea/">biopipen.ns.gsea</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.bam/">biopipen.ns.bam</a>
                </li>
    </ul>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../../CHANGELOG/">Change log</a>
                </li>
              </ul>
      </div>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../..">biopipen</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../.." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">biopipen.ns.rnaseq</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/pwwang/biopipen/edit/master/docs/api/source/biopipen.ns.rnaseq.md">Edit on pwwang/biopipen</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="mkapi-node">
<h1 id="" class="mkapi-object mkapi-object-code plain">
<span class="mkapi-object-kind mkapi-object-kind-code">SOURCE CODE</span>
<span class="mkapi-object-prefix">biopipen.ns.</span><span class="mkapi-object-name">rnaseq</span>
<span id="biopipen.ns.rnaseq"></span><a class="mkapi-docs-link" href="../../biopipen.ns.rnaseq">DOCS</a>
</h1>

<div class="mkapi-code">
<div class="codehilite"><pre><span></span><code><span class="sd">&quot;&quot;&quot;RNA-seq data analysis&quot;&quot;&quot;</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">..core.proc</span><span class="w"> </span><span class="kn">import</span> <span class="n">Proc</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">..core.config</span><span class="w"> </span><span class="kn">import</span> <span class="n">config</span>


<span class="k">class</span><span class="w"> </span><span class="nc">UnitConversion</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.rnaseq.UnitConversion"></span><a class="mkapi-docs-link" title="biopipen.ns.rnaseq.UnitConversion" href="../../biopipen.ns.rnaseq/#biopipen.ns.rnaseq.UnitConversion">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert expression value units back and forth</span>

<span class="sd">    See &lt;https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/&gt;</span>
<span class="sd">    and &lt;https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/Expression_mRNA_Pipeline/#fpkm&gt;.</span>

<span class="sd">    Following converstions are supported -</span>
<span class="sd">    * `count -&gt; cpm, fpkm/rpkm, fpkmuq/rpkmrq, tpm, tmm`</span>
<span class="sd">    * `fpkm/rpkm -&gt; count, tpm, cpm`</span>
<span class="sd">    * `tpm -&gt; count, fpkm/rpkm, cpm`</span>
<span class="sd">    * `cpm -&gt; count, fpkm/rpkm, tpm`</span>
<span class="sd">    NOTE that during some conversions, `sum(counts/effLen)` is approximated to</span>
<span class="sd">    `sum(counts)/sum(effLen) * length(effLen))`</span>

<span class="sd">    You can also use this process to just transform the expression values, e.g., take</span>
<span class="sd">    log2 of the expression values. In this case, you can set `inunit` and `outunit` to</span>
<span class="sd">    `count` and `log2(count + 1)` respectively.</span>

<span class="sd">    Input:</span>
<span class="sd">        infile: Input file containing expression values</span>
<span class="sd">            The file should be a matrix with rows representing genes and columns</span>
<span class="sd">            representing samples.</span>
<span class="sd">            It could be an RDS file containing a data frame or a matrix, or a</span>
<span class="sd">            text file containing a matrix with tab as the delimiter. The text</span>
<span class="sd">            file can be gzipped.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: Output file containing the converted expression values</span>
<span class="sd">            The file will be a matrix with rows representing genes and columns</span>
<span class="sd">            representing samples.</span>

<span class="sd">    Envs:</span>
<span class="sd">        inunit: The input unit of the expression values.</span>
<span class="sd">            You can also use an expression to indicate the input unit, e.g.,</span>
<span class="sd">            `log2(counts + 1)`. The expression should be like `A * fn(B*X + C) + D`,</span>
<span class="sd">            where `A`, `B`, `C` and `D` are constants, `fn` is a function, and X is</span>
<span class="sd">            the input unit.</span>
<span class="sd">            Currently only `expr`, `sqrt`, `log2`, `log10` and `log` are supported as</span>
<span class="sd">            functions.</span>
<span class="sd">            Supported input units are:</span>
<span class="sd">            * counts/count/rawcounts/rawcount: raw counts.</span>
<span class="sd">            * cpm: counts per million.</span>
<span class="sd">            * fpkm/rpkm: fragments per kilobase of transcript per million.</span>
<span class="sd">            * fpkmuq/rpkmuq: upper quartile normalized FPKM/RPKM.</span>
<span class="sd">            * tpm: transcripts per million.</span>
<span class="sd">            * tmm: trimmed mean of M-values.</span>
<span class="sd">        outunit: The output unit of the expression values. An expression can also be</span>
<span class="sd">            used for transformation (e.g. `log2(tpm + 1)`). If `inunit` is `count`,</span>
<span class="sd">            then this means we are converting raw counts to tpm, and transforming it</span>
<span class="sd">            to `log2(tpm + 1)` as the output. Any expression supported by `R` can be</span>
<span class="sd">            used. Same units as `inunit` are supported.</span>
<span class="sd">        refexon: Path to the reference exon gff file.</span>
<span class="sd">        meanfl (type=auto): A file containing the mean fragment length for each sample</span>
<span class="sd">            by rows (samples as rowname), without header.</span>
<span class="sd">            Or a fixed universal estimated number (1 used by TCGA).</span>
<span class="sd">        nreads (type=auto): The estimatied total number of reads for each sample.</span>
<span class="sd">            or you can pass a file with the number for each sample by rows</span>
<span class="sd">            (samples as rowname), without header.</span>
<span class="sd">            When converting `fpkm/rpkm -&gt; count`, it should be total reads of that sample.</span>
<span class="sd">            When converting `cpm -&gt; count`: it should be total reads of that sample.</span>
<span class="sd">            When converting `tpm -&gt; count`: it should be total reads of that sample.</span>
<span class="sd">            When converting `tpm -&gt; cpm`: it should be total reads of that sample.</span>
<span class="sd">            When converting `tpm -&gt; fpkm/rpkm`: it should be `sum(fpkm)` of that sample.</span>
<span class="sd">            It is not used when converting `count -&gt; cpm, fpkm/rpkm, tpm`.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.infile | basename}}&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;inunit&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;outunit&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;refexon&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">refexon</span><span class="p">,</span>
        <span class="s2">&quot;meanfl&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
        <span class="s2">&quot;nreads&quot;</span><span class="p">:</span> <span class="mi">1_000_000</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/rnaseq/UnitConversion.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">Simulation</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.rnaseq.Simulation"></span><a class="mkapi-docs-link" title="biopipen.ns.rnaseq.Simulation" href="../../biopipen.ns.rnaseq/#biopipen.ns.rnaseq.Simulation">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Simulate RNA-seq data using ESCO/RUVcorr package</span>

<span class="sd">    Input:</span>
<span class="sd">        ngenes: Number of genes to simulate</span>
<span class="sd">        nsamples: Number of samples to simulate</span>
<span class="sd">            If you want to force the process to re-simulate for the same</span>
<span class="sd">            `ngenes` and `nsamples`, you can set a different value for `envs.seed`.</span>
<span class="sd">            Note that the samples will be shown as cells in the output (since</span>
<span class="sd">            the simulation is designed for single-cell RNA-seq data).</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: Output file containing the simulated data with rows representing</span>
<span class="sd">            genes and columns representing samples.</span>
<span class="sd">        outdir: Output directory containing the simulated data</span>
<span class="sd">            `sim.rds` and `True.rds` will be generated.</span>
<span class="sd">            For `ESCO`, `sim.rds` contains the simulated data in a</span>
<span class="sd">            `SingleCellExperiment` object, and `True.rds` contains the matrix of true</span>
<span class="sd">            counts.</span>
<span class="sd">            For `RUVcorr`, `sim.rds` contains the simulated data in list with</span>
<span class="sd">            `Truth`, A matrix containing the values of Xβ; `Y` A matrix containing the</span>
<span class="sd">            values in `Y`; `Noise` A matrix containing the values in `Wα`; `Sigma`</span>
<span class="sd">            A matrix containing the true gene-gene correlations, as defined by Xβ; and</span>
<span class="sd">            `Info` A matrix containing some of the general information about the</span>
<span class="sd">            simulation.</span>
<span class="sd">            For all matrices, rows represent genes and columns represent samples.</span>

<span class="sd">    Envs:</span>
<span class="sd">        tool (choice): Which tool to use for simulation.</span>
<span class="sd">            - ESCO: uses the [ESCO](https://github.com/JINJINT/ESCO) package.</span>
<span class="sd">            - RUVcorr: uses the [RUVcorr](https://rdrr.io/bioc/RUVcorr/) package.</span>
<span class="sd">        ncores (type=int): Number of cores to use.</span>
<span class="sd">        seed (type=int): Random seed.</span>
<span class="sd">            If not set, seed will not be set.</span>
<span class="sd">        esco_args (ns): Additional arguments to pass to the simulation function.</span>
<span class="sd">            - save (choice): Which type of data to save to `out.outfile`.</span>
<span class="sd">                - `simulated-truth`: saves the simulated true counts.</span>
<span class="sd">                - `zero-inflated`: saves the zero-inflated counts.</span>
<span class="sd">                - `down-sampled`: saves the down-sampled counts.</span>
<span class="sd">            - type (choice): Which type of heterogenounity to use.</span>
<span class="sd">                - single: produces a single population.</span>
<span class="sd">                - group: produces distinct groups.</span>
<span class="sd">                - tree: produces distinct groups but admits a tree structure.</span>
<span class="sd">                - traj: produces distinct groups but admits a smooth trajectory</span>
<span class="sd">                    structure.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://rdrr.io/github/JINJINT/ESCO/man/escoParams.html&gt;.</span>
<span class="sd">        ruvcorr_args (ns): Additional arguments to pass to the simulation</span>
<span class="sd">            function.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://rdrr.io/bioc/RUVcorr/man/simulateGEdata.html&gt;.</span>
<span class="sd">        transpose_output (flag): If set, the output will be transposed.</span>
<span class="sd">        index_start (type=int): The index to start from when naming the samples.</span>
<span class="sd">            Affects the sample names in `out.outfile` only.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;ngenes:var, nsamples:var&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;outfile:file:{{in.ngenes}}x{{in.nsamples}}.sim/simulated.txt&quot;</span><span class="p">,</span>
        <span class="s2">&quot;outdir:dir:{{in.ngenes}}x{{in.nsamples}}.sim&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;tool&quot;</span><span class="p">:</span> <span class="s2">&quot;RUVcorr&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;single&quot;</span><span class="p">,</span>
        <span class="s2">&quot;esco_args&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;dropout-type&quot;</span><span class="p">:</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span>
            <span class="s2">&quot;save&quot;</span><span class="p">:</span> <span class="s2">&quot;simulated-truth&quot;</span><span class="p">,</span>
            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;single&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;ruvcorr_args&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;seed&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;transpose_output&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;index_start&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/rnaseq/Simulation.R&quot;</span>
</code></pre></div>
</div>
</div>
              
            </div>
          </div><footer>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  <div class="footer-notes">
    <a href="https://github.com/pwwang/biopipen" title="Open repository in GitHub">Github Repo</a>
    <div class="footer-sep"></div>Built with <a href="https://www.mkdocs.org/">MkDocs</a> using theme <a href="https://github.com/pwwang/mkdocs-rtd">mkdocs-rtd</a>.
  </div>
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
  </span>
</div>
    <script src="../../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../../..";</script>
    <script src="../../../js/theme_extra.js"></script>
    <script src="../../../js/theme.js"></script>
    <script src="../../../js/version-select.js"></script>
      <script src="../../../js/mkapi.js"></script>
      <script src="../../../extra.js"></script>
      <script src="../../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
