<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../img/favicon.ico" />
    <title>biopipen.ns.scrna - biopipen</title>
    <link rel="stylesheet" href="../../../css/theme.css" />
    <link rel="stylesheet" href="../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="//use.fontawesome.com/releases/v5.8.1/css/all.css" rel="stylesheet" />
        <link href="//use.fontawesome.com/releases/v5.8.1/css/v4-shims.css" rel="stylesheet" />
        <link href="../../../css/mkapi-common.css" rel="stylesheet" />
        <link href="../../../style.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "biopipen.ns.scrna";
        var mkdocs_page_input_path = "api/source/biopipen.ns.scrna.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.configure({ ignoreUnescapedHTML: true }); hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../..">
          <img src="../../../img/logo.png" class="logo" alt="Logo"/>
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="nav-scrollable-wrapper">
      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Pipelines</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../../pipelines/cnvkit_pipeline/">CNVkit Pipeline</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../../pipelines/cellranger_pipeline/">CellRanger Pipeline</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">API</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.utils</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils/">biopipen.utils</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.gene/">biopipen.utils.gene</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.reference/">biopipen.utils.reference</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.misc/">biopipen.utils.misc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.vcf/">biopipen.utils.vcf</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.core</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core/">biopipen.core</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.proc/">biopipen.core.proc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.filters/">biopipen.core.filters</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.testing/">biopipen.core.testing</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.config/">biopipen.core.config</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.defaults/">biopipen.core.defaults</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.ns</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns/">biopipen.ns</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.gene/">biopipen.ns.gene</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cellranger/">biopipen.ns.cellranger</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.scrna_metabolic_landscape/">biopipen.ns.scrna_metabolic_landscape</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.protein/">biopipen.ns.protein</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.delim/">biopipen.ns.delim</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.rnaseq/">biopipen.ns.rnaseq</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.plot/">biopipen.ns.plot</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.tcgamaf/">biopipen.ns.tcgamaf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.regulatory/">biopipen.ns.regulatory</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.stats/">biopipen.ns.stats</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cellranger_pipeline/">biopipen.ns.cellranger_pipeline</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.bed/">biopipen.ns.bed</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnv/">biopipen.ns.cnv</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.snp/">biopipen.ns.snp</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.web/">biopipen.ns.web</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.misc/">biopipen.ns.misc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnvkit/">biopipen.ns.cnvkit</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnvkit_pipeline/">biopipen.ns.cnvkit_pipeline</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.tcr/">biopipen.ns.tcr</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.scrna/">biopipen.ns.scrna</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.vcf/">biopipen.ns.vcf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.gsea/">biopipen.ns.gsea</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.bam/">biopipen.ns.bam</a>
                </li>
    </ul>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../../CHANGELOG/">Change log</a>
                </li>
              </ul>
      </div>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../..">biopipen</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../.." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">biopipen.ns.scrna</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/pwwang/biopipen/edit/master/docs/api/source/biopipen.ns.scrna.md">Edit on pwwang/biopipen</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="mkapi-node">
<h1 id="" class="mkapi-object mkapi-object-code plain">
<span class="mkapi-object-kind mkapi-object-kind-code">SOURCE CODE</span>
<span class="mkapi-object-prefix">biopipen.ns.</span><span class="mkapi-object-name">scrna</span>
<span id="biopipen.ns.scrna"></span><a class="mkapi-docs-link" href="../../biopipen.ns.scrna">DOCS</a>
</h1>

<div class="mkapi-code">
<div class="codehilite"><pre><span></span><code><span class="sd">&quot;&quot;&quot;Tools to analyze single-cell RNA&quot;&quot;&quot;</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pipen.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">mark</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">..core.proc</span><span class="w"> </span><span class="kn">import</span> <span class="n">Proc</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">..core.config</span><span class="w"> </span><span class="kn">import</span> <span class="n">config</span>


<span class="k">class</span><span class="w"> </span><span class="nc">SeuratLoading</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.SeuratLoading"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.SeuratLoading" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.SeuratLoading">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Seurat - Loading data</span>

<span class="sd">    Deprecated, should be superseded by SeuratPreparing</span>

<span class="sd">    Input:</span>
<span class="sd">        metafile: The metadata of the samples</span>
<span class="sd">            A tab-delimited file</span>
<span class="sd">            Two columns are required:</span>
<span class="sd">            - `Sample` to specify the sample names.</span>
<span class="sd">            - `RNAData` to assign the path of the data to the samples</span>
<span class="sd">              The path will be read by `Read10X()` from `Seurat`</span>

<span class="sd">    Output:</span>
<span class="sd">        rdsfile: The RDS file with a list of Seurat object</span>

<span class="sd">    Envs:</span>
<span class="sd">        qc: The QC filter for each sample.</span>
<span class="sd">            This will be passed to `subset(obj, subset=&lt;qc&gt;)`.</span>
<span class="sd">            For example</span>
<span class="sd">            `nFeature_RNA &gt; 200 &amp; nFeature_RNA &lt; 2500 &amp; percent.mt &lt; 5`</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;metafile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;rdsfile:file:{{in.metafile | stem}}.seurat.RDS&quot;</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;qc&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">}</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/SeuratLoading.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">SeuratPreparing</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.SeuratPreparing"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.SeuratPreparing" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.SeuratPreparing">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load, prepare and apply QC to data, using `Seurat`</span>

<span class="sd">    This process will -</span>
<span class="sd">    - Prepare the seurat object</span>
<span class="sd">    - Apply QC to the data</span>
<span class="sd">    - Integrate the data from different samples</span>

<span class="sd">    See also</span>
<span class="sd">    - &lt;https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#standard-pre-processing-workflow-1)&gt;</span>
<span class="sd">    - &lt;https://satijalab.org/seurat/articles/integration_introduction&gt;</span>

<span class="sd">    This process will read the scRNA-seq data, based on the information provided by</span>
<span class="sd">    `SampleInfo`, specifically, the paths specified by the `RNAData` column.</span>
<span class="sd">    Those paths should be either paths to directoies containing `matrix.mtx`,</span>
<span class="sd">    `barcodes.tsv` and `features.tsv` files that can be loaded by</span>
<span class="sd">    [`Seurat::Read10X()`](https://satijalab.org/seurat/reference/read10x),</span>
<span class="sd">    or paths of loom files that can be loaded by `SeuratDisk::LoadLoom()`, or paths to</span>
<span class="sd">    `h5` files that can be loaded by</span>
<span class="sd">    [`Seurat::Read10X_h5()`](https://satijalab.org/seurat/reference/read10x_h5).</span>

<span class="sd">    Each sample will be loaded individually and then merged into one `Seurat` object, and then perform QC.</span>

<span class="sd">    In order to perform QC, some additional columns are added to the meta data of the `Seurat` object. They are:</span>

<span class="sd">    - `precent.mt`: The percentage of mitochondrial genes.</span>
<span class="sd">    - `percent.ribo`: The percentage of ribosomal genes.</span>
<span class="sd">    - `precent.hb`: The percentage of hemoglobin genes.</span>
<span class="sd">    - `percent.plat`: The percentage of platelet genes.</span>

<span class="sd">    For integration, two routes are available:</span>

<span class="sd">    - [Performing integration on datasets normalized with `SCTransform`](https://satijalab.org/seurat/articles/seurat5_integration#perform-streamlined-one-line-integrative-analysis)</span>
<span class="sd">    - [Using `NormalizeData` and `FindIntegrationAnchors`](https://satijalab.org/seurat/articles/seurat5_integration#layers-in-the-seurat-v5-object)</span>

<span class="sd">    /// Note</span>
<span class="sd">    When using `SCTransform`, the default Assay will be set to `SCT` in output, rather than `RNA`.</span>
<span class="sd">    If you are using `cca` or `rpca` interation, the default assay will be `integrated`.</span>
<span class="sd">    ///</span>

<span class="sd">    /// Note</span>
<span class="sd">    From `biopipen` v0.23.0, this requires `Seurat` v5.0.0 or higher.</span>
<span class="sd">    ///</span>

<span class="sd">    Input:</span>
<span class="sd">        metafile: The metadata of the samples</span>
<span class="sd">            A tab-delimited file</span>
<span class="sd">            Two columns are required:</span>
<span class="sd">            `Sample` to specify the sample names.</span>
<span class="sd">            `RNAData` to assign the path of the data to the samples</span>
<span class="sd">            The path will be read by `Read10X()` from `Seurat`, or the path</span>
<span class="sd">            to the h5 file that can be read by `Read10X_h5()` from `Seurat`.</span>
<span class="sd">            It can also be an RDS or qs2 file containing a `Seurat` object.</span>
<span class="sd">            Note that it must has a column named `Sample` in the meta.data to specify the sample names.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The qs2 file with the Seurat object with all samples integrated.</span>
<span class="sd">            Note that the cell ids are prefixied with sample names.</span>

<span class="sd">    Envs:</span>
<span class="sd">        ncores (type=int): Number of cores to use.</span>
<span class="sd">            Used in `future::plan(strategy = &quot;multicore&quot;, workers = &lt;ncores&gt;)`</span>
<span class="sd">            to parallelize some Seurat procedures.</span>
<span class="sd">        mutaters (type=json): The mutaters to mutate the metadata to the cells.</span>
<span class="sd">            These new columns will be added to the metadata of the Seurat object and</span>
<span class="sd">            will be saved in the output file.</span>
<span class="sd">        min_cells (type=int): The minimum number of cells that a gene must be</span>
<span class="sd">            expressed in to be kept. This is used in `Seurat::CreateSeuratObject()`.</span>
<span class="sd">            Futher QC (`envs.cell_qc`, `envs.gene_qc`) will be performed after this.</span>
<span class="sd">            It doesn&#39;t work when data is loaded from loom files or RDS/qs2 files.</span>
<span class="sd">        min_features (type=int): The minimum number of features that a cell must</span>
<span class="sd">            express to be kept. This is used in `Seurat::CreateSeuratObject()`.</span>
<span class="sd">            Futher QC (`envs.cell_qc`, `envs.gene_qc`) will be performed after this.</span>
<span class="sd">            It doesn&#39;t work when data is loaded from loom files or RDS/qs2 files.</span>
<span class="sd">        cell_qc: Filter expression to filter cells, using</span>
<span class="sd">            `tidyrseurat::filter()`.</span>
<span class="sd">            It can also be a dictionary of expressions, where the names of the list are</span>
<span class="sd">            sample names.</span>
<span class="sd">            You can have a default expression in the list with the name &quot;DEFAULT&quot; for</span>
<span class="sd">            the samples that are not listed.</span>
<span class="sd">            Available QC keys include `nFeature_RNA`, `nCount_RNA`,</span>
<span class="sd">            `percent.mt`, `percent.ribo`, `percent.hb`, and `percent.plat`.</span>

<span class="sd">            /// Tip | Example</span>
<span class="sd">            Including the columns added above, all available QC keys include</span>
<span class="sd">            `nFeature_RNA`, `nCount_RNA`, `percent.mt`, `percent.ribo`, `percent.hb`,</span>
<span class="sd">            and `percent.plat`. For example:</span>

<span class="sd">            ```toml</span>
<span class="sd">            [SeuratPreparing.envs]</span>

<span class="sd">            cell_qc = &quot;nFeature_RNA &gt; 200 &amp; percent.mt &lt; 5&quot;</span>
<span class="sd">            ```</span>
<span class="sd">            will keep cells with more than 200 genes and less than 5%% mitochondrial</span>
<span class="sd">            genes.</span>
<span class="sd">            ///</span>

<span class="sd">        gene_qc (ns): Filter genes.</span>
<span class="sd">            `gene_qc` is applied after `cell_qc`.</span>
<span class="sd">            - min_cells: The minimum number of cells that a gene must be</span>
<span class="sd">                expressed in to be kept.</span>
<span class="sd">            - excludes: The genes to exclude. Multiple genes can be specified by</span>
<span class="sd">                comma separated values, or as a list.</span>

<span class="sd">            /// Tip | Example</span>
<span class="sd">            ```toml</span>
<span class="sd">            [SeuratPreparing.envs]</span>

<span class="sd">            gene_qc = { min_cells = 3 }</span>
<span class="sd">            ```</span>
<span class="sd">            will keep genes that are expressed in at least 3 cells.</span>
<span class="sd">            ///</span>

<span class="sd">        ccs_args (ns): Arguments for `RunSeuratCellCycleScoring()`.</span>
<span class="sd">            When &quot;S.Score&quot; and/or &quot;G2M.Score&quot; are specified in `envs.SCTransform.vars-to-regress` when `envs.use_sct = TRUE`</span>
<span class="sd">            or in `envs.ScaleData.vars-to-regress` when `envs.use_sct = FALSE`, this will be automatically enabled</span>
<span class="sd">            to calculate the cell cycle scores and regress them out in the following transformation step.</span>
<span class="sd">            Otherwise if this is empty, cell cycle scoring and the normalization before cell cycle scoring will be skipped.</span>
<span class="sd">            - trans_args (ns): Arguments for `RunSeuratTransformation()` to normalize the data before cell cycle scoring.</span>
<span class="sd">                `use_sct` will be `FALSE` by default (not using `envs.use_sct`).</span>
<span class="sd">                See &lt;https://github.com/satijalab/seurat/issues/7694&gt;.</span>
<span class="sd">                - &lt;more&gt;: See &lt;https://pwwang.github.io/biopipen.utils.R/reference/RunSeuratTransformation.html&gt;.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://pwwang.github.io/biopipen.utils.R/reference/RunSeuratCellCycleScoring.html&gt;.</span>

<span class="sd">        qc_plots (type=json): The plots for QC metrics.</span>
<span class="sd">            It should be a json (or python dict) with the keys as the names of the plots and</span>
<span class="sd">            the values also as dicts with the following keys:</span>
<span class="sd">            * kind: The kind of QC. Either `gene` or `cell` (default).</span>
<span class="sd">            * devpars: The device parameters for the plot. A dict with `res`, `height`, and `width`.</span>
<span class="sd">            * more_formats: The formats to save the plots other than `png`.</span>
<span class="sd">            * save_code: Whether to save the code to reproduce the plot.</span>
<span class="sd">            * other arguments passed to</span>
<span class="sd">            [`biopipen.utils::VizSeuratCellQC`](https://pwwang.github.io/biopipen.utils.R/reference/VizSeuratCellQC.html)</span>
<span class="sd">            when `kind` is `cell` or</span>
<span class="sd">            [`biopipen.utils::VizSeuratGeneQC`](https://pwwang.github.io/biopipen.utils.R/reference/VizSeuratGeneQC.html)</span>
<span class="sd">            when `kind` is `gene`.</span>

<span class="sd">        use_sct (flag): Whether use SCTransform routine to integrate samples or not.</span>
<span class="sd">            Before the following procedures, the `RNA` layer will be split by samples.</span>

<span class="sd">            If `False`, following procedures will be performed in the order:</span>
<span class="sd">            * [`NormalizeData`](https://satijalab.org/seurat/reference/normalizedata).</span>
<span class="sd">            * [`FindVariableFeatures`](https://satijalab.org/seurat/reference/findvariablefeatures).</span>
<span class="sd">            * [`ScaleData`](https://satijalab.org/seurat/reference/scaledata).</span>
<span class="sd">            See &lt;https://satijalab.org/seurat/articles/seurat5_integration#layers-in-the-seurat-v5-object&gt;</span>
<span class="sd">            and &lt;https://satijalab.org/seurat/articles/pbmc3k_tutorial.html&gt;</span>

<span class="sd">            If `True`, following procedures will be performed in the order:</span>
<span class="sd">            * [`SCTransform`](https://satijalab.org/seurat/reference/sctransform).</span>
<span class="sd">            See &lt;https://satijalab.org/seurat/articles/seurat5_integration#perform-streamlined-one-line-integrative-analysis&gt;</span>

<span class="sd">        no_integration (flag): Whether to skip integration or not.</span>
<span class="sd">        NormalizeData (ns): Arguments for [`NormalizeData()`](https://satijalab.org/seurat/reference/normalizedata).</span>
<span class="sd">            `object` is specified internally, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/normalizedata&gt;</span>

<span class="sd">        FindVariableFeatures (ns): Arguments for [`FindVariableFeatures()`](https://satijalab.org/seurat/reference/findvariablefeatures).</span>
<span class="sd">            `object` is specified internally, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/findvariablefeatures&gt;</span>

<span class="sd">        ScaleData (ns): Arguments for [`ScaleData()`](https://satijalab.org/seurat/reference/scaledata).</span>
<span class="sd">            `object` and `features` is specified internally, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/scaledata&gt;</span>

<span class="sd">        RunPCA (ns): Arguments for [`RunPCA()`](https://satijalab.org/seurat/reference/runpca).</span>
<span class="sd">            `object` and `features` is specified internally, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            - npcs (type=int): The number of PCs to compute.</span>
<span class="sd">                For each sample, `npcs` will be no larger than the number of columns - 1.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/runpca&gt;</span>

<span class="sd">        SCTransform (ns): Arguments for [`SCTransform()`](https://satijalab.org/seurat/reference/sctransform).</span>
<span class="sd">            `object` is specified internally, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            - return-only-var-genes: Whether to return only variable genes.</span>
<span class="sd">            - min_cells: The minimum number of cells that a gene must be expressed in to be kept.</span>
<span class="sd">                A hidden argument of `SCTransform` to filter genes.</span>
<span class="sd">                If you try to keep all genes in the `RNA` assay, you can set `min_cells` to `0` and</span>
<span class="sd">                `return-only-var-genes` to `False`.</span>
<span class="sd">                See &lt;https://github.com/satijalab/seurat/issues/3598#issuecomment-715505537&gt;</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/sctransform&gt;</span>

<span class="sd">        IntegrateLayers (ns): Arguments for [`IntegrateLayers()`](https://satijalab.org/seurat/reference/integratelayers).</span>
<span class="sd">            `object` is specified internally, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            When `use_sct` is `True`, `normalization-method` defaults to `SCT`.</span>
<span class="sd">            - method (choice): The method to use for integration.</span>
<span class="sd">                - CCAIntegration: Use `Seurat::CCAIntegration`.</span>
<span class="sd">                - CCA: Same as `CCAIntegration`.</span>
<span class="sd">                - cca: Same as `CCAIntegration`.</span>
<span class="sd">                - RPCAIntegration: Use `Seurat::RPCAIntegration`.</span>
<span class="sd">                - RPCA: Same as `RPCAIntegration`.</span>
<span class="sd">                - rpca: Same as `RPCAIntegration`.</span>
<span class="sd">                - HarmonyIntegration: Use `Seurat::HarmonyIntegration`.</span>
<span class="sd">                - Harmony: Same as `HarmonyIntegration`.</span>
<span class="sd">                - harmony: Same as `HarmonyIntegration`.</span>
<span class="sd">                - FastMNNIntegration: Use `Seurat::FastMNNIntegration`.</span>
<span class="sd">                - FastMNN: Same as `FastMNNIntegration`.</span>
<span class="sd">                - fastmnn: Same as `FastMNNIntegration`.</span>
<span class="sd">                - scVIIntegration: Use `Seurat::scVIIntegration`.</span>
<span class="sd">                - scVI: Same as `scVIIntegration`.</span>
<span class="sd">                - scvi: Same as `scVIIntegration`.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/integratelayers&gt;</span>

<span class="sd">        doublet_detector (choice): The doublet detector to use.</span>
<span class="sd">            - none: Do not use any doublet detector.</span>
<span class="sd">            - DoubletFinder: Use `DoubletFinder` to detect doublets.</span>
<span class="sd">            - doubletfinder: Same as `DoubletFinder`.</span>
<span class="sd">            - scDblFinder: Use `scDblFinder` to detect doublets.</span>
<span class="sd">            - scdblfinder: Same as `scDblFinder`.</span>

<span class="sd">        DoubletFinder (ns): Arguments to run [`DoubletFinder`](https://github.com/chris-mcginnis-ucsf/DoubletFinder).</span>
<span class="sd">            See also &lt;https://demultiplexing-doublet-detecting-docs.readthedocs.io/en/latest/DoubletFinder.html&gt;.</span>
<span class="sd">            - PCs (type=int): Number of PCs to use for &#39;doubletFinder&#39; function.</span>
<span class="sd">            - doublets (type=float): Number of expected doublets as a proportion of the pool size.</span>
<span class="sd">            - pN (type=float): Number of doublets to simulate as a proportion of the pool size.</span>
<span class="sd">            - ncores (type=int): Number of cores to use for `DoubletFinder::paramSweep`.</span>
<span class="sd">                Set to `None` to use `envs.ncores`.</span>
<span class="sd">                Since parallelization of the function usually exhausts memory, if big `envs.ncores` does not work</span>
<span class="sd">                for `DoubletFinder`, set this to a smaller number.</span>

<span class="sd">        scDblFinder (ns): Arguments to run [`scDblFinder`](https://rdrr.io/bioc/scDblFinder/man/scDblFinder.html).</span>
<span class="sd">            - dbr (type=float): The expected doublet rate.</span>
<span class="sd">            - ncores (type=int): Number of cores to use for `scDblFinder`.</span>
<span class="sd">                Set to `None` to use `envs.ncores`.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://rdrr.io/bioc/scDblFinder/man/scDblFinder.html&gt;.</span>

<span class="sd">        cache (type=auto): Whether to cache the information at different steps.</span>
<span class="sd">            If `True`, the seurat object will be cached in the job output directory, which will be not cleaned up when job is rerunning.</span>
<span class="sd">            The cached seurat object will be saved as `&lt;signature&gt;.&lt;kind&gt;.RDS` file, where `&lt;signature&gt;` is the signature determined by</span>
<span class="sd">            the input and envs of the process.</span>
<span class="sd">            See &lt;https://github.com/satijalab/seurat/issues/7849&gt;, &lt;https://github.com/satijalab/seurat/issues/5358&gt; and</span>
<span class="sd">            &lt;https://github.com/satijalab/seurat/issues/6748&gt; for more details also about reproducibility issues.</span>
<span class="sd">            To not use the cached seurat object, you can either set `cache` to `False` or delete the cached file at</span>
<span class="sd">            `&lt;signature&gt;.RDS` in the cache directory.</span>

<span class="sd">    Requires:</span>
<span class="sd">        r-seurat:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(Seurat)&quot;)</span>
<span class="sd">        r-future:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(future)&quot;)</span>
<span class="sd">        r-bracer:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(bracer)&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;metafile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.metafile | stem}}.seurat.qs&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs_depth</span> <span class="o">=</span> <span class="mi">4</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;min_cells&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s2">&quot;min_features&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s2">&quot;cell_qc&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>  <span class="c1"># &quot;nFeature_RNA &gt; 200 &amp; percent.mt &lt; 5&quot;,</span>
        <span class="s2">&quot;gene_qc&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;min_cells&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;excludes&quot;</span><span class="p">:</span> <span class="p">[]},</span>
        <span class="s2">&quot;ccs_args&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;qc_plots&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Violin Plots&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;kind&quot;</span><span class="p">:</span> <span class="s2">&quot;cell&quot;</span><span class="p">,</span>
                <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;violin&quot;</span><span class="p">,</span>
                <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;height&quot;</span><span class="p">:</span> <span class="mi">600</span><span class="p">,</span> <span class="s2">&quot;width&quot;</span><span class="p">:</span> <span class="mi">1200</span><span class="p">},</span>
            <span class="p">},</span>
            <span class="s2">&quot;Scatter Plots&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;kind&quot;</span><span class="p">:</span> <span class="s2">&quot;cell&quot;</span><span class="p">,</span>
                <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;scatter&quot;</span><span class="p">,</span>
                <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;height&quot;</span><span class="p">:</span> <span class="mi">800</span><span class="p">,</span> <span class="s2">&quot;width&quot;</span><span class="p">:</span> <span class="mi">1200</span><span class="p">},</span>
            <span class="p">},</span>
            <span class="s2">&quot;Ridge Plots&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;kind&quot;</span><span class="p">:</span> <span class="s2">&quot;cell&quot;</span><span class="p">,</span>
                <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;ridge&quot;</span><span class="p">,</span>
                <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;height&quot;</span><span class="p">:</span> <span class="mi">800</span><span class="p">,</span> <span class="s2">&quot;width&quot;</span><span class="p">:</span> <span class="mi">1200</span><span class="p">},</span>
            <span class="p">},</span>
            <span class="s2">&quot;Distribution of number of cells a gene is expressed in&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;kind&quot;</span><span class="p">:</span> <span class="s2">&quot;gene&quot;</span><span class="p">,</span>
                <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;histogram&quot;</span><span class="p">,</span>
                <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;height&quot;</span><span class="p">:</span> <span class="mi">1200</span><span class="p">,</span> <span class="s2">&quot;width&quot;</span><span class="p">:</span> <span class="mi">1200</span><span class="p">},</span>
            <span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;use_sct&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;no_integration&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;NormalizeData&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;FindVariableFeatures&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;ScaleData&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;RunPCA&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;SCTransform&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;return-only-var-genes&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;min_cells&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="s2">&quot;verbose&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;IntegrateLayers&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;harmony&quot;</span><span class="p">},</span>
        <span class="s2">&quot;doublet_detector&quot;</span><span class="p">:</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span>
        <span class="s2">&quot;DoubletFinder&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;PCs&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span> <span class="s2">&quot;pN&quot;</span><span class="p">:</span> <span class="mf">0.25</span><span class="p">,</span> <span class="s2">&quot;doublets&quot;</span><span class="p">:</span> <span class="mf">0.075</span><span class="p">,</span> <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
        <span class="s2">&quot;scDblFinder&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;dbr&quot;</span><span class="p">:</span> <span class="mf">0.075</span><span class="p">,</span> <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
        <span class="s2">&quot;cache&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">tmpdir</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/SeuratPreparing.R&quot;</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/common.svelte&quot;</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">SeuratClustering</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.SeuratClustering"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.SeuratClustering" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.SeuratClustering">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Determine the clusters of cells without reference using Seurat FindClusters</span>
<span class="sd">    procedure.</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object loaded by SeuratPreparing</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The seurat object with cluster information at `seurat_clusters` or</span>
<span class="sd">            the name specified by `envs.ident`</span>

<span class="sd">    Envs:</span>
<span class="sd">        ncores (type=int;order=-100): Number of cores to use.</span>
<span class="sd">            Used in `future::plan(strategy = &quot;multicore&quot;, workers = &lt;ncores&gt;)`</span>
<span class="sd">            to parallelize some Seurat procedures.</span>
<span class="sd">            See also: &lt;https://satijalab.org/seurat/articles/future_vignette.html&gt;</span>
<span class="sd">        ident: The name in the metadata to save the cluster labels.</span>
<span class="sd">            A shortcut for `envs[&quot;FindClusters&quot;][&quot;cluster.name&quot;]`.</span>
<span class="sd">        python: The path to the python executable where `umap-learn` is installed.</span>
<span class="sd">            Needed for `RunUMAP` when `umap.method` is set to `umap-learn`.</span>
<span class="sd">        RunUMAP (ns): Arguments for [`RunUMAP()`](https://satijalab.org/seurat/reference/runumap).</span>
<span class="sd">            `object` is specified internally, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            `dims=N` will be expanded to `dims=1:N`; The maximal value of `N` will be the minimum of `N` and the number of columns - 1 for each sample.</span>
<span class="sd">            You can also specify `features` instead of `dims` to use specific features for UMAP. It can be</span>
<span class="sd">            a list with the following fields: `order` (the order of the markers to use for UMAP, e.g. &quot;desc(abs(avg_log2FC))&quot;, and</span>
<span class="sd">            `n` (the number of total features to use for UMAP, e.g. 30).</span>
<span class="sd">            If `features` is a list, it will run `biopipen.utils::RunSeuratDEAnalysis` to get the markers</span>
<span class="sd">            for each group, and then select the top `n`/`ngroups` features for each group based on the `order` field.</span>
<span class="sd">            If `features` is a numeric value, it will be treated as the `n` field</span>
<span class="sd">            in the list above, with the default `order` being &quot;desc(abs(avg_log2FC))&quot;.</span>
<span class="sd">            - dims (type=int): The number of PCs to use</span>
<span class="sd">            - reduction: The reduction to use for UMAP.</span>
<span class="sd">                If not provided, `sobj@misc$integrated_new_reduction` will be used.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/runumap&gt;</span>
<span class="sd">        RunPCA (ns): Arguments for [`RunPCA()`](https://satijalab.org/seurat/reference/runpca).</span>
<span class="sd">        FindNeighbors (ns): Arguments for [`FindNeighbors()`](https://satijalab.org/seurat/reference/findneighbors).</span>
<span class="sd">            `object` is specified internally, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            - reduction: The reduction to use.</span>
<span class="sd">                If not provided, `sobj@misc$integrated_new_reduction` will be used.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/findneighbors&gt;</span>
<span class="sd">        FindClusters (ns): Arguments for [`FindClusters()`](https://satijalab.org/seurat/reference/findclusters).</span>
<span class="sd">            `object` is specified internally, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            The cluster labels will be saved in cluster names and prefixed with &quot;c&quot;.</span>
<span class="sd">            The first cluster will be &quot;c1&quot;, instead of &quot;c0&quot;.</span>
<span class="sd">            - resolution (type=auto): The resolution of the clustering. You can have multiple resolutions as a list or as a string separated by comma.</span>
<span class="sd">                Ranges are also supported, for example: `0.1:0.5:0.1` will generate `0.1, 0.2, 0.3, 0.4, 0.5`. The step can be omitted, defaulting to 0.1.</span>
<span class="sd">                The results will be saved in `&lt;ident&gt;_&lt;resolution&gt;`.</span>
<span class="sd">                The final resolution will be used to define the clusters at `&lt;ident&gt;`.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/findclusters&gt;</span>
<span class="sd">        cache (type=auto): Where to cache the information at different steps.</span>
<span class="sd">            If `True`, the seurat object will be cached in the job output directory, which will be not cleaned up when job is rerunning.</span>
<span class="sd">            Set to `False` to not cache the results.</span>

<span class="sd">    Requires:</span>
<span class="sd">        r-seurat:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(Seurat)&quot;)</span>
<span class="sd">        r-tidyr:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(tidyr)&quot;)</span>
<span class="sd">        r-dplyr:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(dplyr)&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.srtobj | stem}}.qs&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;ident&quot;</span><span class="p">:</span> <span class="s2">&quot;seurat_clusters&quot;</span><span class="p">,</span>
        <span class="s2">&quot;python&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span><span class="p">,</span>
        <span class="s2">&quot;RunPCA&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;RunUMAP&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;FindNeighbors&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;FindClusters&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;resolution&quot;</span><span class="p">:</span> <span class="mf">0.8</span><span class="p">},</span>
        <span class="s2">&quot;cache&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">tmpdir</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/SeuratClustering.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">SeuratSubClustering</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.SeuratSubClustering"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.SeuratSubClustering" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.SeuratSubClustering">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find clusters of a subset of cells.</span>

<span class="sd">    It&#39;s unlike [`Seurat::FindSubCluster`], which only finds subclusters of a single</span>
<span class="sd">    cluster. Instead, it will perform the whole clustering procedure on the subset of</span>
<span class="sd">    cells. One can use metadata to specify the subset of cells to perform clustering on.</span>

<span class="sd">    For the subset of cells, the reductions will be re-performed on the subset of cells,</span>
<span class="sd">    and then the clustering will be performed on the subset of cells. The reduction</span>
<span class="sd">    will be saved in `object@reduction$&lt;casename&gt;.&lt;reduction&gt;` of the original object and the</span>
<span class="sd">    clustering will be saved in the metadata of the original object using the casename</span>
<span class="sd">    as the column name.</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object in RDS or qs/qs2 format.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The seurat object with the subclustering information in qs/qs2 format.</span>

<span class="sd">    Envs:</span>
<span class="sd">        ncores (type=int;order=-100): Number of cores to use.</span>
<span class="sd">            Used in `future::plan(strategy = &quot;multicore&quot;, workers = &lt;ncores&gt;)`</span>
<span class="sd">            to parallelize some Seurat procedures.</span>
<span class="sd">        python: The path to the python executable where `umap-learn` is installed.</span>
<span class="sd">            Needed for `RunUMAP` when `umap.method` is set to `umap-learn`.</span>
<span class="sd">        mutaters (type=json): The mutaters to mutate the metadata to subset the cells.</span>
<span class="sd">            The mutaters will be applied in the order specified.</span>
<span class="sd">        subset: An expression to subset the cells, will be passed to</span>
<span class="sd">            [`tidyseurat::filter()`](https://stemangiola.github.io/tidyseurat/reference/filter.html).</span>
<span class="sd">        RunPCA (ns): Arguments for [`RunPCA()`](https://satijalab.org/seurat/reference/runpca).</span>
<span class="sd">            `object` is specified internally as the subset object, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/runpca&gt;</span>
<span class="sd">        RunUMAP (ns): Arguments for [`RunUMAP()`](https://satijalab.org/seurat/reference/runumap).</span>
<span class="sd">            `object` is specified internally as the subset object, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            `dims=N` will be expanded to `dims=1:N`; The maximal value of `N` will be the minimum of `N` and the number of columns - 1 for each sample.</span>
<span class="sd">            You can also specify `features` instead of `dims` to use specific features for UMAP. It can be</span>
<span class="sd">            a list with the following fields: `order` (the order of the markers to use for UMAP, e.g. &quot;desc(abs(avg_log2FC))&quot;, and</span>
<span class="sd">            `n` (the number of total features to use for UMAP, e.g. 30).</span>
<span class="sd">            If `features` is a list, it will run `biopipen.utils::RunSeuratDEAnalysis` to get the markers</span>
<span class="sd">            for each group, and then select the top `n`/`ngroups` features for each group based on the `order` field.</span>
<span class="sd">            If `features` is a numeric value, it will be treated as the `n` field</span>
<span class="sd">            in the list above, with the default `order` being &quot;desc(abs(avg_log2FC))&quot;.</span>
<span class="sd">            - dims (type=int): The number of PCs to use</span>
<span class="sd">            - reduction: The reduction to use for UMAP.</span>
<span class="sd">                If not provided, `sobj@misc$integrated_new_reduction` will be used.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/runumap&gt;</span>
<span class="sd">        FindNeighbors (ns): Arguments for [`FindNeighbors()`](https://satijalab.org/seurat/reference/findneighbors).</span>
<span class="sd">            `object` is specified internally, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            - reduction: The reduction to use.</span>
<span class="sd">                If not provided, `object@misc$integrated_new_reduction` will be used.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/findneighbors&gt;</span>
<span class="sd">        FindClusters (ns): Arguments for [`FindClusters()`](https://satijalab.org/seurat/reference/findclusters).</span>
<span class="sd">            `object` is specified internally, and `-` in the key will be replaced with `.`.</span>
<span class="sd">            The cluster labels will be prefixed with &quot;s&quot;. The first cluster will be &quot;s1&quot;, instead of &quot;s0&quot;.</span>
<span class="sd">            - resolution (type=auto): The resolution of the clustering. You can have multiple resolutions as a list or as a string separated by comma.</span>
<span class="sd">                Ranges are also supported, for example: `0.1:0.5:0.1` will generate `0.1, 0.2, 0.3, 0.4, 0.5`. The step can be omitted, defaulting to 0.1.</span>
<span class="sd">                The results will be saved in `&lt;casename&gt;_&lt;resolution&gt;`.</span>
<span class="sd">                The final resolution will be used to define the clusters at `&lt;casename&gt;`.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/findclusters&gt;</span>
<span class="sd">        cache (type=auto): Whether to cache the results.</span>
<span class="sd">            If `True`, the seurat object will be cached in the job output directory, which will be not cleaned up when job is rerunning.</span>
<span class="sd">            Set to `False` to not cache the results.</span>
<span class="sd">        cases (type=json): The cases to perform subclustering.</span>
<span class="sd">            Keys are the names of the cases and values are the dicts inherited from `envs` except `mutaters` and `cache`.</span>
<span class="sd">            If empty, a case with name `subcluster` will be created with default parameters.</span>
<span class="sd">            The case name will be passed to `biopipen.utils::SeuratSubCluster()` as `name`.</span>
<span class="sd">            It will be used as the prefix for the reduction name, keys and cluster names.</span>
<span class="sd">            For reduction keys, it will be `toupper(&lt;name&gt;)` + &quot;PC_&quot; and `toupper(&lt;name&gt;)` + &quot;UMAP_&quot;.</span>
<span class="sd">            For cluster names, it will be `&lt;name&gt;` + &quot;.&quot; + resolution.</span>
<span class="sd">            And the final cluster name will be `&lt;name&gt;`.</span>
<span class="sd">            Note that the `name` should be alphanumeric and anything other than alphanumeric will be removed.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.srtobj | stem}}.qs&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs_depth</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;python&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span><span class="p">,</span>
        <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;RunPCA&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;RunUMAP&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;FindNeighbors&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;FindClusters&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;resolution&quot;</span><span class="p">:</span> <span class="mf">0.8</span><span class="p">},</span>
        <span class="s2">&quot;cache&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">tmpdir</span><span class="p">,</span>
        <span class="s2">&quot;cases&quot;</span><span class="p">:</span> <span class="p">{},</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/SeuratSubClustering.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">SeuratClusterStats</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.SeuratClusterStats"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.SeuratClusterStats" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.SeuratClusterStats">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Statistics of the clustering.</span>

<span class="sd">    Including the number/fraction of cells in each cluster, the gene expression values</span>
<span class="sd">    and dimension reduction plots. It&#39;s also possible to perform stats on</span>
<span class="sd">    TCR clones/clusters or other metadata for each T-cell cluster.</span>

<span class="sd">    Examples:</span>
<span class="sd">        ### Clustree Plot</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.clustrees.&quot;Clustree Plot&quot;]</span>
<span class="sd">        prefix = &quot;seurat_clusters&quot;</span>
<span class="sd">        devpars = {height = 500}</span>
<span class="sd">        ```</span>

<span class="sd">        ![Clustree Plot](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/clustrees/seurat_clusters.clustree.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Number of cells in each cluster (Bar Chart)</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.stats.&quot;Number of cells in each cluster (Bar Chart)&quot;]</span>
<span class="sd">        plot_type = &quot;bar&quot;</span>
<span class="sd">        x_text_angle = 90</span>
<span class="sd">        ```</span>

<span class="sd">        ![Number of cells in each cluster (Bar Chart)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/stats/Number-of-cells-in-each-cluster-Bar-Chart-.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Number of cells in each cluster by Sample (Bar Chart)</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.stats.&quot;Number of cells in each cluster by Sample (Bar Chart)&quot;]</span>
<span class="sd">        plot_type = &quot;bar&quot;</span>
<span class="sd">        group_by = &quot;Sample&quot;</span>
<span class="sd">        x_text_angle = 90</span>
<span class="sd">        ```</span>

<span class="sd">        ![Number of cells in each cluster by Sample (Bar Chart)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/stats/Number-of-cells-in-each-cluster-by-Sample-Bar-Chart-.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Number of cells in each cluster by Diagnosis</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.stats.&quot;Number of cells in each cluster by Diagnosis&quot;]</span>
<span class="sd">        plot_type = &quot;bar&quot;</span>
<span class="sd">        group_by = &quot;Diagnosis&quot;</span>
<span class="sd">        frac = &quot;group&quot;</span>
<span class="sd">        x_text_angle = 90</span>
<span class="sd">        swap = true</span>
<span class="sd">        position = &quot;stack&quot;</span>
<span class="sd">        ```</span>

<span class="sd">        ![Number of cells in each cluster by Diagnosis](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/stats/Number-of-cells-in-each-cluster-by-Diagnosis.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Number of cells in each cluster by Diagnosis (Circos Plot)</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.stats.&quot;Number of cells in each cluster by Diagnosis (Circos Plot)&quot;]</span>
<span class="sd">        plot_type = &quot;circos&quot;</span>
<span class="sd">        group_by = &quot;Diagnosis&quot;</span>
<span class="sd">        ```</span>

<span class="sd">        ![Number of cells in each cluster by Diagnosis (Circos Plot)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/stats/Number-of-cells-in-each-cluster-by-Diagnosis-Circos-Plot-.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Number of cells in each cluster by Diagnosis (Sankey Plot)</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.stats.&quot;Number of cells in each cluster by Diagnosis (Sankey Plot)&quot;]</span>
<span class="sd">        plot_type = &quot;sankey&quot;</span>
<span class="sd">        group_by = [&quot;seurat_clusters&quot;, &quot;Diagnosis&quot;]</span>
<span class="sd">        links_alpha = 0.6</span>
<span class="sd">        devpars = {width = 800}</span>
<span class="sd">        ```</span>

<span class="sd">        ![Number of cells in each cluster by Diagnosis (Sankey Plot)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/stats/Number-of-cells-in-each-cluster-by-Diagnosis-Sankey-Plot-.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Number of cells in each cluster by Sample (Spider Plot)</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.stats.&quot;Number of cells in each cluster by Sample (Spider Plot)&quot;]</span>
<span class="sd">        plot_type = &quot;spider&quot;</span>
<span class="sd">        group_by = &quot;Diagnosis&quot;</span>
<span class="sd">        palette = &quot;Set1&quot;</span>
<span class="sd">        ```</span>

<span class="sd">        ![Number of cells in each cluster by Sample (Spider Plot)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/stats/Number-of-cells-in-each-cluster-by-Sample-Spider-Plot-.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Number of genes detected in each cluster</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.ngenes.&quot;Number of genes detected in each cluster&quot;]</span>
<span class="sd">        plot_type = &quot;violin&quot;</span>
<span class="sd">        add_box = true</span>
<span class="sd">        add_point = true</span>
<span class="sd">        ```</span>

<span class="sd">        ![Number of genes detected in each cluster](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/ngenes/Number-of-genes-detected-in-each-cluster.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Feature Expression in Clusters (Violin Plots)</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.features_defaults]</span>
<span class="sd">        features = [&quot;CD3D&quot;, &quot;CD4&quot;, &quot;CD8A&quot;, &quot;MS4A1&quot;, &quot;CD14&quot;, &quot;LYZ&quot;, &quot;FCGR3A&quot;, &quot;NCAM1&quot;, &quot;KLRD1&quot;]</span>

<span class="sd">        [SeuratClusterStats.envs.features.&quot;Feature Expression in Clusters (Violin Plots)&quot;]</span>
<span class="sd">        plot_type = &quot;violin&quot;</span>
<span class="sd">        ident = &quot;seurat_clusters&quot;</span>
<span class="sd">        ```</span>

<span class="sd">        ![Feature Expression in Clusters (Violin Plots)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/features/Feature-Expression-in-Clusters-Violin-Plots-.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Feature Expression in Clusters (Ridge Plots)</span>

<span class="sd">        ```toml</span>
<span class="sd">        # Using the same features as above</span>
<span class="sd">        [SeuratClusterStats.envs.features.&quot;Feature Expression in Clusters (Ridge Plots)&quot;]</span>
<span class="sd">        plot_type = &quot;ridge&quot;</span>
<span class="sd">        ident = &quot;seurat_clusters&quot;</span>
<span class="sd">        flip = true</span>
<span class="sd">        ```</span>

<span class="sd">        ![Feature Expression in Clusters (Ridge Plots)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/features/Feature-Expression-in-Clusters-Ridge-Plots-.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Feature Expression in Clusters by Diagnosis</span>

<span class="sd">        ```toml</span>
<span class="sd">        # Using the same features as above</span>
<span class="sd">        [SeuratClusterStats.envs.features.&quot;Feature Expression in Clusters by Diagnosis&quot;]</span>
<span class="sd">        plot_type = &quot;violin&quot;</span>
<span class="sd">        group_by = &quot;Diagnosis&quot;</span>
<span class="sd">        ident = &quot;seurat_clusters&quot;</span>
<span class="sd">        comparisons = true</span>
<span class="sd">        sig_label = &quot;p.signif&quot;</span>
<span class="sd">        ```</span>

<span class="sd">        ![Feature Expression in Clusters by Diagnosis](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/features/Feature-Expression-in-Clusters-by-Diagnosis.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Feature Expression in Clusters (stacked)</span>

<span class="sd">        ```toml</span>
<span class="sd">        # Using the same features as above</span>
<span class="sd">        [SeuratClusterStats.envs.features.&quot;Feature Expression in Clusters (stacked)&quot;]</span>
<span class="sd">        plot_type = &quot;violin&quot;</span>
<span class="sd">        ident = &quot;seurat_clusters&quot;</span>
<span class="sd">        add_bg = true</span>
<span class="sd">        stack = true</span>
<span class="sd">        add_box = true</span>
<span class="sd">        ```</span>

<span class="sd">        ![Feature Expression in Clusters (stacked)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/features/Feature-Expression-in-Clusters-stacked-.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### CD4 Expression on UMAP</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.features.&quot;CD4 Expression on UMAP&quot;]</span>
<span class="sd">        plot_type = &quot;dim&quot;</span>
<span class="sd">        feature = &quot;CD4&quot;</span>
<span class="sd">        highlight = &quot;seurat_clusters == &#39;c1&#39;&quot;</span>
<span class="sd">        ```</span>

<span class="sd">        ![CD4 Expression on UMAP](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/features/CD4-Expression-on-UMAP.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Feature Expression in Clusters by Diagnosis (Heatmap)</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.features.&quot;Feature Expression in Clusters by Diagnosis (Heatmap)&quot;]</span>
<span class="sd">        # Grouped features</span>
<span class="sd">        features = {&quot;T cell markers&quot; = [&quot;CD3D&quot;, &quot;CD4&quot;, &quot;CD8A&quot;], &quot;B cell markers&quot; = [&quot;MS4A1&quot;], &quot;Monocyte markers&quot; = [&quot;CD14&quot;, &quot;LYZ&quot;, &quot;FCGR3A&quot;], &quot;NK cell markers&quot; = [&quot;NCAM1&quot;, &quot;KLRD1&quot;]}</span>
<span class="sd">        plot_type = &quot;heatmap&quot;</span>
<span class="sd">        ident = &quot;Diagnosis&quot;</span>
<span class="sd">        columns_split_by = &quot;seurat_clusters&quot;</span>
<span class="sd">        name = &quot;Expression&quot;</span>
<span class="sd">        devpars = {height = 560}</span>
<span class="sd">        ```</span>

<span class="sd">        ![Feature Expression in Clusters by Diagnosis (Heatmap)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/features/Feature-Expression-in-Clusters-by-Diagnosis-Heatmap-.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Feature Expression in Clusters by Diagnosis (Heatmap with annotations)</span>

<span class="sd">        ```toml</span>
<span class="sd">        # Using the default features</span>
<span class="sd">        [SeuratClusterStats.envs.features.&quot;Feature Expression in Clusters by Diagnosis (Heatmap with annotations)&quot;]</span>
<span class="sd">        ident = &quot;seurat_clusters&quot;</span>
<span class="sd">        cell_type = &quot;dot&quot;</span>
<span class="sd">        plot_type = &quot;heatmap&quot;</span>
<span class="sd">        name = &quot;Expression Level&quot;</span>
<span class="sd">        dot_size = &quot;nanmean&quot;</span>
<span class="sd">        dot_size_name = &quot;Percent Expressed&quot;</span>
<span class="sd">        add_bg = true</span>
<span class="sd">        rows_split_by = &quot;Diagnosis&quot;</span>
<span class="sd">        cluster_rows = false</span>
<span class="sd">        flip = true</span>
<span class="sd">        palette = &quot;YlOrRd&quot;</span>
<span class="sd">        column_annotation = [&quot;percent.mt&quot;, &quot;VDJ_Presence&quot;]</span>
<span class="sd">        column_annotation_type = {&quot;percent.mt&quot; = &quot;violin&quot;, VDJ_Presence = &quot;pie&quot;}</span>
<span class="sd">        column_annotation_params = {&quot;percent.mt&quot; = {show_legend = false}}</span>
<span class="sd">        devpars = {width = 1400, height = 900}</span>
<span class="sd">        ```</span>

<span class="sd">        ![Feature Expression in Clusters by Diagnosis (Heatmap with annotations)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/features/Feature-Expression-in-Clusters-by-Diagnosis-Heatmap-with-annotations-.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Dimensional reduction plot</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.features.&quot;Dimensional reduction plot&quot;]</span>
<span class="sd">        label = true</span>
<span class="sd">        ```</span>

<span class="sd">        ![Dimensional reduction plot](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/dimplots/Dimensional-reduction-plot.dim.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Dimensional reduction plot (with marks)</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.dimplots.&quot;Dimensional reduction plot (with marks)&quot;]</span>
<span class="sd">        add_mark = true</span>
<span class="sd">        mark_linetype = 2</span>
<span class="sd">        ```</span>

<span class="sd">        ![Dimensional reduction plot (with marks)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/dimplots/Dimensional-reduction-plot-with-marks-.dim.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Dimensional reduction plot (with hex bins)</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.dimplots.&quot;Dimensional reduction plot (with hex bins)&quot;]</span>
<span class="sd">        hex = true</span>
<span class="sd">        hex_bins = 50</span>
<span class="sd">        ```</span>

<span class="sd">        ![Dimensional reduction plot (with hex bins)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/dimplots/Dimensional-reduction-plot-with-hex-bins-.dim.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Dimensional reduction plot (with Diagnosis stats)</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.dimplots.&quot;Dimensional reduction plot (with Diagnosis stats)&quot;]</span>
<span class="sd">        stat_by = &quot;Diagnosis&quot;</span>
<span class="sd">        stat_plot_type = &quot;ring&quot;</span>
<span class="sd">        stat_plot_size = 0.15</span>
<span class="sd">        ```</span>

<span class="sd">        ![Dimensional reduction plot (with Diagnosis stats)](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/dimplots/Dimensional-reduction-plot-with-Diagnosis-stats-.dim.png){: width=&quot;80%&quot; }</span>

<span class="sd">        ### Dimensional reduction plot by Diagnosis</span>

<span class="sd">        ```toml</span>
<span class="sd">        [SeuratClusterStats.envs.dimplots.&quot;Dimensional reduction plot by Diagnosis&quot;]</span>
<span class="sd">        facet_by = &quot;Diagnosis&quot;</span>
<span class="sd">        highlight = true</span>
<span class="sd">        theme = &quot;theme_blank&quot;</span>
<span class="sd">        ```</span>

<span class="sd">        ![Dimensional reduction plot by Diagnosis](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/seuratclusterstats/SeuratClusterStats/sampleinfo.scRep.cluster_stats/dimplots/Dimensional-reduction-plot-by-Diagnosis.dim.png){: width=&quot;80%&quot; }</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object loaded by `SeuratClustering`</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory.</span>
<span class="sd">            Different types of plots will be saved in different subdirectories.</span>
<span class="sd">            For example, `clustree` plots will be saved in `clustrees` subdirectory.</span>
<span class="sd">            For each case in `envs.clustrees`, both the png and pdf files will be saved.</span>

<span class="sd">    Envs:</span>
<span class="sd">        mutaters (type=json): The mutaters to mutate the metadata to subset the cells.</span>
<span class="sd">            The mutaters will be applied in the order specified.</span>
<span class="sd">            You can also use the clone selectors to select the TCR clones/clusters.</span>
<span class="sd">            See &lt;https://pwwang.github.io/scplotter/reference/clone_selectors.html&gt;.</span>
<span class="sd">        cache (type=auto): Whether to cache the plots.</span>
<span class="sd">            Currently only plots for features are supported, since creating the those</span>
<span class="sd">            plots can be time consuming.</span>
<span class="sd">            If `True`, the plots will be cached in the job output directory, which will</span>
<span class="sd">            be not cleaned up when job is rerunning.</span>
<span class="sd">        clustrees_defaults (ns): The parameters for the clustree plots.</span>
<span class="sd">            - devpars (ns): The device parameters for the clustree plot.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - more_formats (type=list): The formats to save the plots other than `png`.</span>
<span class="sd">            - save_code (flag): Whether to save the code to reproduce the plot.</span>
<span class="sd">            - prefix (type=auto): string indicating columns containing clustering information.</span>
<span class="sd">                The trailing dot is not necessary and will be added automatically.</span>
<span class="sd">                When `TRUE`, clustrees will be plotted when there is `FindClusters` or</span>
<span class="sd">                `FindClusters.*` in the `obj@commands`.</span>
<span class="sd">                The latter is generated by `SeuratSubClustering`.</span>
<span class="sd">                This will be ignored when `envs.clustrees` is specified</span>
<span class="sd">                (the prefix of each case must be specified separately).</span>
<span class="sd">            - &lt;more&gt;: Other arguments passed to `scplotter::ClustreePlot`.</span>
<span class="sd">                See &lt;https://pwwang.github.io/scplotter/reference/ClustreePlot.html&gt;</span>
<span class="sd">        clustrees (type=json): The cases for clustree plots.</span>
<span class="sd">            Keys are the names of the plots and values are the dicts inherited from `env.clustrees_defaults` except `prefix`.</span>
<span class="sd">            There is no default case for `clustrees`.</span>
<span class="sd">        stats_defaults (ns): The default parameters for `stats`.</span>
<span class="sd">            This is to do some basic statistics on the clusters/cells. For more comprehensive analysis,</span>
<span class="sd">            see &lt;https://pwwang.github.io/scplotter/reference/CellStatPlot.html&gt;.</span>
<span class="sd">            The parameters from the cases can overwrite the default parameters.</span>
<span class="sd">            - subset: An expression to subset the cells, will be passed to `tidyrseurat::filter()`.</span>
<span class="sd">            - devpars (ns): The device parameters for the clustree plot.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - descr: The description of the plot, showing in the report.</span>
<span class="sd">            - more_formats (type=list): The formats to save the plots other than `png`.</span>
<span class="sd">            - save_code (flag): Whether to save the code to reproduce the plot.</span>
<span class="sd">            - save_data (flag): Whether to save the data used to generate the plot.</span>
<span class="sd">            - &lt;more&gt;: Other arguments passed to `scplotter::CellStatPlot`.</span>
<span class="sd">                See &lt;https://pwwang.github.io/scplotter/reference/CellStatPlot.html&gt;.</span>
<span class="sd">        stats (type=json): The number/fraction of cells to plot.</span>
<span class="sd">            Keys are the names of the plots and values are the dicts inherited from `env.stats_defaults`.</span>
<span class="sd">        ngenes_defaults (ns): The default parameters for `ngenes`.</span>
<span class="sd">            The default parameters to plot the number of genes expressed in each cell.</span>
<span class="sd">            - more_formats (type=list): The formats to save the plots other than `png`.</span>
<span class="sd">            - subset: An expression to subset the cells, will be passed to `tidyrseurat::filter()`.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">        ngenes (type=json): The number of genes expressed in each cell.</span>
<span class="sd">            Keys are the names of the plots and values are the dicts inherited from `env.ngenes_defaults`.</span>
<span class="sd">        features_defaults (ns): The default parameters for `features`.</span>
<span class="sd">            - features (type=auto): The features to plot.</span>
<span class="sd">                It can be either a string with comma separated features, a list of features, a file path with `file://` prefix with features</span>
<span class="sd">                (one per line), or an integer to use the top N features from `VariantFeatures(srtobj)`.</span>
<span class="sd">                It can also be a dict with the keys as the feature group names and the values as the features, which</span>
<span class="sd">                is used for heatmap to group the features.</span>
<span class="sd">            - order_by (type=auto): The order of the clusters to show on the plot.</span>
<span class="sd">                An expression passed to `dplyr::arrange()` on the grouped meta data frame (by `ident`).</span>
<span class="sd">                For example, you can order the clusters by the activation score of</span>
<span class="sd">                the cluster: `desc(mean(ActivationScore, na.rm = TRUE))`, suppose you have a column</span>
<span class="sd">                `ActivationScore` in the metadata.</span>
<span class="sd">                You may also specify the literal order of the clusters by a list of strings (at least two).</span>
<span class="sd">            - subset: An expression to subset the cells, will be passed to `tidyrseurat::filter()`.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - descr: The description of the plot, showing in the report.</span>
<span class="sd">            - more_formats (type=list): The formats to save the plots other than `png`.</span>
<span class="sd">            - save_code (flag): Whether to save the code to reproduce the plot.</span>
<span class="sd">            - save_data (flag): Whether to save the data used to generate the plot.</span>
<span class="sd">            - &lt;more&gt;: Other arguments passed to `scplotter::FeatureStatPlot`.</span>
<span class="sd">                See &lt;https://pwwang.github.io/scplotter/reference/FeatureStatPlot.html&gt;</span>
<span class="sd">        features (type=json): The plots for features, include gene expressions, and columns from metadata.</span>
<span class="sd">            Keys are the titles of the cases and values are the dicts inherited from `env.features_defaults`.</span>
<span class="sd">        dimplots_defaults (ns): The default parameters for `dimplots`.</span>
<span class="sd">            You can also set `dims` to `[1, 2, 3]` to make a 3D dimensional reduction plot.</span>
<span class="sd">            Note that the default `n.components` for `RunUMAP` is 2, so if you want to make a 3D UMAP plot,</span>
<span class="sd">            you need to set `n.components` to 3 in `RunUMAP` parameters in `SeuratSubClustering` or `SeuratClustering`.</span>
<span class="sd">            - group_by: The identity to use.</span>
<span class="sd">                If it is from subclustering (reduction `sub_umap_&lt;ident&gt;` exists), this reduction will be used if `reduction`</span>
<span class="sd">                is set to `dim` or `auto`.</span>
<span class="sd">            - split_by: The column name in metadata to split the cells into different plots.</span>
<span class="sd">            - subset: An expression to subset the cells, will be passed to `tidyrseurat::filter()`.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - reduction (choice): Which dimensionality reduction to use.</span>
<span class="sd">                - dim: Use `Seurat::DimPlot`.</span>
<span class="sd">                    First searches for `umap`, then `tsne`, then `pca`.</span>
<span class="sd">                    If `ident` is from subclustering, `sub_umap_&lt;ident&gt;` will be used.</span>
<span class="sd">                - auto: Same as `dim`</span>
<span class="sd">                - umap: Use `Seurat::UMAPPlot`.</span>
<span class="sd">                - tsne: Use `Seurat::TSNEPlot`.</span>
<span class="sd">                - pca: Use `Seurat::PCAPlot`.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://pwwang.github.io/scplotter/reference/CellDimPlot.html&gt;</span>
<span class="sd">        dimplots (type=json): The dimensional reduction plots.</span>
<span class="sd">            Keys are the titles of the plots and values are the dicts inherited from `env.dimplots_defaults`. It can also have other parameters from</span>
<span class="sd">            [`scplotter::CellDimPlot`](https://pwwang.github.io/scplotter/reference/CellDimPlot.html).</span>

<span class="sd">    Requires:</span>
<span class="sd">        r-seurat:</span>
<span class="sd">            - check: {{proc.lang}} -e &quot;library(Seurat)&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.srtobj | stem}}.cluster_stats&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;cache&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">tmpdir</span><span class="p">,</span>
        <span class="s2">&quot;clustrees_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;save_code&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;prefix&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;clustrees&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;stats_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;descr&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;save_code&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;save_data&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;stats&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Number of cells in each cluster (Bar Chart)&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;bar&quot;</span><span class="p">,</span>
                <span class="s2">&quot;x_text_angle&quot;</span><span class="p">:</span> <span class="mi">90</span><span class="p">,</span>
            <span class="p">},</span>
            <span class="s2">&quot;Number of cells in each cluster by Sample (Bar Chart)&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;bar&quot;</span><span class="p">,</span>
                <span class="s2">&quot;group_by&quot;</span><span class="p">:</span> <span class="s2">&quot;Sample&quot;</span><span class="p">,</span>
                <span class="s2">&quot;x_text_angle&quot;</span><span class="p">:</span> <span class="mi">90</span><span class="p">,</span>
            <span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;ngenes_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;height&quot;</span><span class="p">:</span> <span class="mi">800</span><span class="p">,</span> <span class="s2">&quot;width&quot;</span><span class="p">:</span> <span class="mi">1000</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;ngenes&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Number of genes expressed in each cluster&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="p">},</span>
        <span class="s2">&quot;features_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;features&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;order_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
            <span class="s2">&quot;descr&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;save_code&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;save_data&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;features&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;dimplots_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;group_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>  <span class="c1"># use default ident</span>
            <span class="s2">&quot;split_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;reduction&quot;</span><span class="p">:</span> <span class="s2">&quot;dim&quot;</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;dimplots&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Dimensional reduction plot&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
            <span class="p">},</span>
        <span class="p">},</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/SeuratClusterStats.R&quot;</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/common.svelte&quot;</span><span class="p">,</span>
        <span class="s2">&quot;report_paging&quot;</span><span class="p">:</span> <span class="mi">8</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">ModuleScoreCalculator</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.ModuleScoreCalculator"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.ModuleScoreCalculator" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.ModuleScoreCalculator">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the module scores for each cell</span>

<span class="sd">    The module scores are calculated by</span>
<span class="sd">    [`Seurat::AddModuleScore()`](https://satijalab.org/seurat/reference/addmodulescore)</span>
<span class="sd">    or [`Seurat::CellCycleScoring()`](https://satijalab.org/seurat/reference/cellcyclescoring)</span>
<span class="sd">    for cell cycle scores.</span>

<span class="sd">    The module scores are calculated as the average expression levels of each</span>
<span class="sd">    program on single cell level, subtracted by the aggregated expression of</span>
<span class="sd">    control feature sets. All analyzed features are binned based on averaged</span>
<span class="sd">    expression, and the control features are randomly selected from each bin.</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object loaded by `SeuratClustering`</span>

<span class="sd">    Output:</span>
<span class="sd">        rdsfile: The seurat object with module scores added to the metadata.</span>

<span class="sd">    Envs:</span>
<span class="sd">        defaults (ns): The default parameters for `modules`.</span>
<span class="sd">            - features: The features to calculate the scores. Multiple features</span>
<span class="sd">                should be separated by comma.</span>
<span class="sd">                You can also specify `cc.genes` or `cc.genes.updated.2019` to</span>
<span class="sd">                use the cell cycle genes to calculate cell cycle scores.</span>
<span class="sd">                If so, three columns will be added to the metadata, including</span>
<span class="sd">                `S.Score`, `G2M.Score` and `Phase`.</span>
<span class="sd">                Only one type of cell cycle scores can be calculated at a time.</span>
<span class="sd">            - nbin (type=int): Number of bins of aggregate expression levels</span>
<span class="sd">                for all analyzed features.</span>
<span class="sd">            - ctrl (type=int): Number of control features selected from</span>
<span class="sd">                the same bin per analyzed feature.</span>
<span class="sd">            - k (flag): Use feature clusters returned from `DoKMeans`.</span>
<span class="sd">            - assay: The assay to use.</span>
<span class="sd">            - seed (type=int): Set a random seed.</span>
<span class="sd">            - search (flag): Search for symbol synonyms for features in</span>
<span class="sd">                features that don&#39;t match features in object?</span>
<span class="sd">            - keep (flag): Keep the scores for each feature?</span>
<span class="sd">                Only works for non-cell cycle scores.</span>
<span class="sd">            - agg (choice): The aggregation function to use.</span>
<span class="sd">                Only works for non-cell cycle scores.</span>
<span class="sd">                - mean: The mean of the expression levels</span>
<span class="sd">                - median: The median of the expression levels</span>
<span class="sd">                - sum: The sum of the expression levels</span>
<span class="sd">                - max: The max of the expression levels</span>
<span class="sd">                - min: The min of the expression levels</span>
<span class="sd">                - var: The variance of the expression levels</span>
<span class="sd">                - sd: The standard deviation of the expression levels</span>
<span class="sd">        modules (type=json): The modules to calculate the scores.</span>
<span class="sd">            Keys are the names of the expression programs and values are the</span>
<span class="sd">            dicts inherited from `env.defaults`.</span>
<span class="sd">            Here are some examples -</span>
<span class="sd">            &gt;&gt;&gt; {</span>
<span class="sd">            &gt;&gt;&gt;     &quot;CellCycle&quot;: {&quot;features&quot;: &quot;cc.genes.updated.2019&quot;},</span>
<span class="sd">            &gt;&gt;&gt;     &quot;Exhaustion&quot;: {&quot;features&quot;: &quot;HAVCR2,ENTPD1,LAYN,LAG3&quot;},</span>
<span class="sd">            &gt;&gt;&gt;     &quot;Activation&quot;: {&quot;features&quot;: &quot;IFNG&quot;},</span>
<span class="sd">            &gt;&gt;&gt;     &quot;Proliferation&quot;: {&quot;features&quot;: &quot;STMN1,TUBB&quot;}</span>
<span class="sd">            &gt;&gt;&gt; }</span>

<span class="sd">            For `CellCycle`, the columns `S.Score`, `G2M.Score` and `Phase` will</span>
<span class="sd">            be added to the metadata. `S.Score` and `G2M.Score` are the cell cycle</span>
<span class="sd">            scores for each cell, and `Phase` is the cell cycle phase for each cell.</span>

<span class="sd">            You can also add Diffusion Components (DC) to the modules</span>
<span class="sd">            &gt;&gt;&gt; {&quot;DC&quot;: {&quot;features&quot;: 2, &quot;kind&quot;: &quot;diffmap&quot;}}</span>
<span class="sd">            will perform diffusion map as a reduction and add the first 2</span>
<span class="sd">            components as `DC_1` and `DC_2` to the metadata. `diffmap` is a shortcut</span>
<span class="sd">            for `diffusion_map`. Other key-value pairs will pass to</span>
<span class="sd">            [`destiny::DiffusionMap()`](https://www.rdocumentation.org/packages/destiny/versions/2.0.4/topics/DiffusionMap class).</span>
<span class="sd">            You can later plot the diffusion map by using</span>
<span class="sd">            `reduction = &quot;DC&quot;` in `env.dimplots` in `SeuratClusterStats`.</span>
<span class="sd">            This requires [`SingleCellExperiment`](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html)</span>
<span class="sd">            and [`destiny`](https://bioconductor.org/packages/release/bioc/html/destiny.html) R packages.</span>
<span class="sd">        post_mutaters (type=json): The mutaters to mutate the metadata after</span>
<span class="sd">            calculating the module scores.</span>
<span class="sd">            The mutaters will be applied in the order specified.</span>
<span class="sd">            This is useful when you want to create new scores based on the</span>
<span class="sd">            calculated module scores.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;rdsfile:file:{{in.srtobj | stem}}.qs&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;features&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;nbin&quot;</span><span class="p">:</span> <span class="mi">24</span><span class="p">,</span>
            <span class="s2">&quot;ctrl&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span>
            <span class="s2">&quot;k&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;assay&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;seed&quot;</span><span class="p">:</span> <span class="mi">8525</span><span class="p">,</span>
            <span class="s2">&quot;search&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;keep&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;agg&quot;</span><span class="p">:</span> <span class="s2">&quot;mean&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;modules&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="c1"># &quot;CellCycle&quot;: {&quot;features&quot;: &quot;cc.genes.updated.2019&quot;},</span>
            <span class="c1"># &quot;Exhaustion&quot;: {&quot;features&quot;: &quot;HAVCR2,ENTPD1,LAYN,LAG3&quot;},</span>
            <span class="c1"># &quot;Activation&quot;: {&quot;features&quot;: &quot;IFNG&quot;},</span>
            <span class="c1"># &quot;Proliferation&quot;: {&quot;features&quot;: &quot;STMN1,TUBB&quot;},</span>
        <span class="p">},</span>
        <span class="s2">&quot;post_mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/ModuleScoreCalculator.R&quot;</span>


<span class="nd">@mark</span><span class="p">(</span><span id="biopipen.ns.scrna.CellsDistribution"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.CellsDistribution" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.CellsDistribution">DOCS</a>
    <span class="n">deprecated</span><span class="o">=</span><span class="p">(</span>
        <span class="s2">&quot;[</span><span class="si">{proc.name}</span><span class="s2">] is deprecated, &quot;</span>
        <span class="s2">&quot;use [SeuratClusterStats] or [ClonalStats] instead.&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="k">class</span><span class="w"> </span><span class="nc">CellsDistribution</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Distribution of cells (i.e. in a TCR clone) from different groups</span>
<span class="sd">    for each cluster</span>

<span class="sd">    This generates a set of pie charts with proportion of cells in each cluster</span>
<span class="sd">    Rows are the cells identities (i.e. TCR clones or TCR clusters), columns</span>
<span class="sd">    are groups (i.e. clinic groups).</span>

<span class="sd">    Examples:</span>
<span class="sd">        ```toml</span>
<span class="sd">        [CellsDistribution.envs.mutaters]</span>
<span class="sd">        # Add Patient1_Tumor_Expanded column with CDR3.aa that</span>
<span class="sd">        # expands in Tumor of patient 1</span>
<span class="sd">        Patient1_Tumor_Expanded = &#39;&#39;&#39;</span>
<span class="sd">          expanded(., region, &quot;Tumor&quot;, subset = patient == &quot;Lung1&quot;, uniq = FALSE)</span>
<span class="sd">        &#39;&#39;&#39;</span>

<span class="sd">        [CellsDistribution.envs.cases.Patient1_Tumor_Expanded]</span>
<span class="sd">        cells_by = &quot;Patient1_Tumor_Expanded&quot;</span>
<span class="sd">        cells_orderby = &quot;desc(CloneSize)&quot;</span>
<span class="sd">        group_by = &quot;region&quot;</span>
<span class="sd">        group_order = [ &quot;Tumor&quot;, &quot;Normal&quot; ]</span>
<span class="sd">        ```</span>

<span class="sd">        ![CellsDistribution_example](https://pwwang.github.io/immunopipe/latest/processes/images/CellsDistribution_example.png)</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object in RDS format</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory.</span>
<span class="sd">            The results for each case will be saved in a subdirectory.</span>

<span class="sd">    Envs:</span>
<span class="sd">        mutaters (type=json): The mutaters to mutate the metadata</span>
<span class="sd">            Keys are the names of the mutaters and values are the R expressions</span>
<span class="sd">            passed by `dplyr::mutate()` to mutate the metadata.</span>

<span class="sd">        cluster_orderby: The order of the clusters to show on the plot.</span>
<span class="sd">            An expression passed to `dplyr::summarise()` on the grouped data frame (by `seurat_clusters`).</span>
<span class="sd">            The summary stat will be passed to `dplyr::arrange()` to order the clusters. It&#39;s applied on the whole meta.data before grouping and subsetting.</span>
<span class="sd">            For example, you can order the clusters by the activation score of</span>
<span class="sd">            the cluster: `desc(mean(ActivationScore, na.rm = TRUE))`, suppose you have a column</span>
<span class="sd">            `ActivationScore` in the metadata.</span>
<span class="sd">        group_by: The column name in metadata to group the cells for the columns of the plot.</span>
<span class="sd">        group_order (list): The order of the groups (columns) to show on the plot</span>
<span class="sd">        cells_by: The column name in metadata to group the cells for the rows of the plot.</span>
<span class="sd">            If your cell groups have overlapping cells, you can also use multiple columns, separated by comma (`,`).</span>
<span class="sd">            These columns will be concatenated to form the cell groups. For the overlapping cells, they will be</span>
<span class="sd">            counted multiple times for different groups. So make sure the cell group names in different columns</span>
<span class="sd">            are unique.</span>
<span class="sd">        cells_order (list): The order of the cells (rows) to show on the plot</span>
<span class="sd">        cells_orderby: An expression passed to `dplyr::arrange()` to order the cells (rows) of the plot.</span>
<span class="sd">            Only works when `cells-order` is not specified.</span>
<span class="sd">            The data frame passed to `dplyr::arrange()` is grouped by `cells_by` before ordering.</span>
<span class="sd">            You can have multiple expressions separated by semicolon (`;`). The expessions will be parsed by `rlang::parse_exprs()`.</span>
<span class="sd">            4 extra columns were added to the metadata for ordering the rows in the plot:</span>
<span class="sd">            * `CloneSize`: The size (number of cells) of clones (identified by `cells_by`)</span>
<span class="sd">            * `CloneGroupSize`: The clone size in each group (identified by `group_by`)</span>
<span class="sd">            * `CloneClusterSize`: The clone size in each cluster (identified by `seurat_clusters`)</span>
<span class="sd">            * `CloneGroupClusterSize`: The clone size in each group and cluster (identified by `group_by` and `seurat_clusters`)</span>
<span class="sd">        cells_n (type=int): The max number of groups to show for each cell group identity (row).</span>
<span class="sd">            Ignored if `cells_order` is specified.</span>
<span class="sd">        subset: An expression to subset the cells, will be passed to `dplyr::filter()` on metadata.</span>
<span class="sd">            This will be applied prior to `each`.</span>
<span class="sd">        descr: The description of the case, will be shown in the report.</span>
<span class="sd">        hm_devpars (ns): The device parameters for the heatmaps.</span>
<span class="sd">            - res (type=int): The resolution of the heatmaps.</span>
<span class="sd">            - height (type=int): The height of the heatmaps.</span>
<span class="sd">            - width (type=int): The width of the heatmaps.</span>
<span class="sd">        devpars (ns): The device parameters for the plots of pie charts.</span>
<span class="sd">            - res (type=int): The resolution of the plots</span>
<span class="sd">            - height (type=int): The height of the plots</span>
<span class="sd">            - width (type=int): The width of the plots</span>
<span class="sd">        each: The column name in metadata to separate the cells into different plots.</span>
<span class="sd">        prefix_each (flag): Whether to prefix the `each` column name to the</span>
<span class="sd">            value as the case/section name.</span>
<span class="sd">        section: The section to show in the report. This allows different cases to be put in the same section in report.</span>
<span class="sd">            Only works when `each` is not specified.</span>
<span class="sd">        overlap (list): Plot the overlap of cell groups (values of `cells_by`) in different cases</span>
<span class="sd">            under the same section.</span>
<span class="sd">            The section must have at least 2 cases, each case should have a single `cells_by` column.</span>
<span class="sd">        cases (type=json;order=99): If you have multiple cases, you can specify them here.</span>
<span class="sd">            Keys are the names of the cases and values are the options above except `mutaters`.</span>
<span class="sd">            If some options are not specified, the options in `envs` will be used.</span>
<span class="sd">            If no cases are specified, a default case will be used with case name `DEFAULT`.</span>

<span class="sd">    Requires:</span>
<span class="sd">        r-seurat:</span>
<span class="sd">            - check: {{proc.lang}} -e &quot;library(Seurat)&quot;</span>
<span class="sd">        r-dplyr:</span>
<span class="sd">            - check: {{proc.lang}} -e &quot;library(dplyr)&quot;</span>
<span class="sd">        r-tidyr:</span>
<span class="sd">            - check: {{proc.lang}} -e &quot;library(tidyr)&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.srtobj | stem}}.cells_distribution&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;cluster_orderby&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;group_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;group_order&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;cells_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;cells_order&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;cells_orderby&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;cells_n&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
        <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;descr&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;hm_devpars&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;each&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;prefix_each&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;section&quot;</span><span class="p">:</span> <span class="s2">&quot;DEFAULT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;overlap&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;cases&quot;</span><span class="p">:</span> <span class="p">{},</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/CellsDistribution.R&quot;</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/scrna/CellsDistribution.svelte&quot;</span><span class="p">,</span>
        <span class="s2">&quot;report_paging&quot;</span><span class="p">:</span> <span class="mi">8</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">SeuratMetadataMutater</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.SeuratMetadataMutater"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.SeuratMetadataMutater" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.SeuratMetadataMutater">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Mutate the metadata of the seurat object</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object loaded by SeuratPreparing</span>
<span class="sd">        metafile: Additional metadata</span>
<span class="sd">            A tab-delimited file with columns as meta columns and rows as</span>
<span class="sd">            cells.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The seurat object with the additional metadata</span>

<span class="sd">    Envs:</span>
<span class="sd">        mutaters (type=json): The mutaters to mutate the metadata.</span>
<span class="sd">            The key-value pairs will be passed the `dplyr::mutate()` to mutate the metadata.</span>
<span class="sd">        subset: An expression to subset the cells, will be passed to `dplyr::filter()`.</span>
<span class="sd">            This will be applied after mutating the metadata.</span>

<span class="sd">    Requires:</span>
<span class="sd">        r-seurat:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(Seurat)&quot;)</span>
<span class="sd">        r-tibble:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(tibble)&quot;)</span>
<span class="sd">        r-dplyr:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(dplyr)&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file, metafile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.srtobj | stem}}.qs&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/SeuratMetadataMutater.R&quot;</span>


<span class="nd">@mark</span><span class="p">(</span><span class="n">deprecated</span><span class="o">=</span><span class="s2">&quot;[</span><span class="si">{proc.name}</span><span class="s2">] is deprecated, use [SeuratClusterStats] instead.&quot;</span><span class="p">)</span><span id="biopipen.ns.scrna.DimPlots"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.DimPlots" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.DimPlots">DOCS</a>
<span class="k">class</span><span class="w"> </span><span class="nc">DimPlots</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Seurat - Dimensional reduction plots</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seruat object in RDS format</span>
<span class="sd">        configfile: A toml configuration file with &quot;cases&quot;</span>
<span class="sd">            If this is given, `envs.cases` will be overriden</span>
<span class="sd">        name: The name of the job, used in report</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory</span>

<span class="sd">    Envs:</span>
<span class="sd">        cases: The cases for the dim plots</span>
<span class="sd">            Keys are the names and values are the arguments to</span>
<span class="sd">            `Seurat::Dimplots`</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file, configfile:file, name:var&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.srtobj | stem}}.dimplots&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/DimPlots.R&quot;</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;cases&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Ident&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;group.by&quot;</span><span class="p">:</span> <span class="s2">&quot;ident&quot;</span><span class="p">}}}</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/scrna/DimPlots.svelte&quot;</span><span class="p">,</span>
        <span class="s2">&quot;report_toc&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">MarkersFinder</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.MarkersFinder"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.MarkersFinder" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.MarkersFinder">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find markers between different groups of cells</span>

<span class="sd">    When only `group_by` is specified as identity column in</span>
<span class="sd">    `envs.cases`, the markers will be found for all the clusters.</span>

<span class="sd">    You can also find the differentially expressed genes between</span>
<span class="sd">    any two groups of cells by setting `group_by` to a different</span>
<span class="sd">    column name in metadata. Follow `envs.cases` for more details.</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object loaded by `SeuratPreparing`</span>
<span class="sd">            If you have your `Seurat` object prepared by yourself, you can also</span>
<span class="sd">            use it here, but you should make sure that the object has been processed</span>
<span class="sd">            by `PrepSCTFindMarkers` if data is not normalized using `SCTransform`.</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory for the markers and plots</span>

<span class="sd">    Envs:</span>
<span class="sd">        ncores (type=int): Number of cores to use for parallel computing for some `Seurat` procedures.</span>
<span class="sd">            * Used in `future::plan(strategy = &quot;multicore&quot;, workers = &lt;ncores&gt;)` to parallelize some Seurat procedures.</span>
<span class="sd">            * See also: &lt;https://satijalab.org/seurat/articles/future_vignette.html&gt;</span>
<span class="sd">        mutaters (type=json): The mutaters to mutate the metadata.</span>
<span class="sd">            You can also use the clone selectors to select the TCR clones/clusters.</span>
<span class="sd">            See &lt;https://pwwang.github.io/scplotter/reference/clone_selectors.html&gt;.</span>
<span class="sd">        group_by: The column name in metadata to group the cells.</span>
<span class="sd">            If only `group_by` is specified, and `ident_1` and `ident_2` are</span>
<span class="sd">            not specified, markers will be found for all groups in this column</span>
<span class="sd">            in the manner of &quot;group vs rest&quot; comparison.</span>
<span class="sd">            `NA` group will be ignored.</span>
<span class="sd">            If `None`, `Seurat::Idents(srtobj)` will be used, which is usually</span>
<span class="sd">            `&quot;seurat_clusters&quot;` after unsupervised clustering.</span>
<span class="sd">        ident_1: The first group of cells to compare</span>
<span class="sd">            When this is empty, the comparisons will be expanded to each group v.s. the rest of the cells in `group_by`.</span>
<span class="sd">        ident_2: The second group of cells to compare</span>
<span class="sd">            If not provided, the rest of the cells are used for `ident_2`.</span>
<span class="sd">        each: The column name in metadata to separate the cells into different</span>
<span class="sd">            cases.</span>
<span class="sd">            When this is specified, the case will be expanded for each value of</span>
<span class="sd">            the column in metadata. For example, when you have `envs.cases.&quot;Cluster Markers&quot;.each = &quot;Sample&quot;`,</span>
<span class="sd">            then the case will be expanded as `envs.cases.&quot;Cluster Markers - Sample1&quot;`, `envs.cases.&quot;Cluster Markers - Sample2&quot;`, etc.</span>
<span class="sd">            You can specify `allmarker_plots` and `overlaps` to plot the markers for all cases in the same plot and plot the overlaps of the markers</span>
<span class="sd">            between different cases by values in this column.</span>
<span class="sd">        dbs (list): The dbs to do enrichment analysis for significant markers.</span>
<span class="sd">            You can use built-in dbs in `enrichit`, or provide your own gmt files.</span>
<span class="sd">            See also &lt;https://pwwang.github.io/enrichit/reference/FetchGMT.html&gt;.</span>
<span class="sd">            The built-in dbs include:</span>
<span class="sd">            * &quot;BioCarta&quot; or &quot;BioCarta_2016&quot;</span>
<span class="sd">            * &quot;GO_Biological_Process&quot; or &quot;GO_Biological_Process_2025&quot;</span>
<span class="sd">            * &quot;GO_Cellular_Component&quot; or &quot;GO_Cellular_Component_2025&quot;</span>
<span class="sd">            * &quot;GO_Molecular_Function&quot; or &quot;GO_Molecular_Function_2025&quot;</span>
<span class="sd">            * &quot;KEGG&quot;, &quot;KEGG_Human&quot;, &quot;KEGG_2021&quot;, or &quot;KEGG_2021_Human&quot;</span>
<span class="sd">            * &quot;Hallmark&quot;, &quot;MSigDB_Hallmark&quot;, or &quot;MSigDB_Hallmark_2020&quot;</span>
<span class="sd">            * &quot;Reactome&quot;, &quot;Reactome_Pathways&quot;, or &quot;Reactome_Pathways_2024&quot;</span>
<span class="sd">            * &quot;WikiPathways&quot;, &quot;WikiPathways_2024&quot;, &quot;WikiPathways_Human&quot;, or &quot;WikiPathways_2024_Human&quot;</span>
<span class="sd">            You can also fetch more dbs from &lt;https://maayanlab.cloud/Enrichr/#libraries&gt;.</span>
<span class="sd">        sigmarkers: An expression passed to `dplyr::filter()` to filter the</span>
<span class="sd">            significant markers for enrichment analysis.</span>
<span class="sd">            Available variables are `p_val`, `avg_log2FC`, `pct.1`, `pct.2` and</span>
<span class="sd">            `p_val_adj`. For example, `&quot;p_val_adj &lt; 0.05 &amp; abs(avg_log2FC) &gt; 1&quot;`</span>
<span class="sd">            to select markers with adjusted p-value &lt; 0.05 and absolute log2</span>
<span class="sd">            fold change &gt; 1.</span>
<span class="sd">        enrich_style (choice): The style of the enrichment analysis.</span>
<span class="sd">            The enrichment analysis will be done by `EnrichIt()` from [`enrichit`](https://pwwang.github.io/enrichit/).</span>
<span class="sd">            Two styles are available:</span>
<span class="sd">            - enrichr: `enrichr` style enrichment analysis (fisher&#39;s exact test will be used).</span>
<span class="sd">            - clusterprofiler: `clusterProfiler` style enrichment analysis (hypergeometric test will be used).</span>
<span class="sd">            - clusterProfiler: alias for `clusterprofiler`</span>
<span class="sd">        assay: The assay to use.</span>
<span class="sd">        error (flag): Error out if no/not enough markers are found or no pathways are enriched.</span>
<span class="sd">            If `False`, empty results will be returned.</span>
<span class="sd">        subset: An expression to subset the cells for each case.</span>
<span class="sd">        cache (type=auto): Where to cache the results.</span>
<span class="sd">            If `True`, cache to `outdir` of the job. If `False`, don&#39;t cache.</span>
<span class="sd">            Otherwise, specify the directory to cache to.</span>
<span class="sd">        rest (ns): Rest arguments for `Seurat::FindMarkers()`.</span>
<span class="sd">            Use `-` to replace `.` in the argument name. For example,</span>
<span class="sd">            use `min-pct` instead of `min.pct`.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/findmarkers&gt;</span>
<span class="sd">        allmarker_plots_defaults (ns): Default options for the plots for all markers when `ident_1` is not specified.</span>
<span class="sd">            - plot_type: The type of the plot.</span>
<span class="sd">                See &lt;https://pwwang.github.io/biopipen.utils.R/reference/VizDEGs.html&gt;.</span>
<span class="sd">                Available types are `violin`, `box`, `bar`, `ridge`, `dim`, `heatmap` and `dot`.</span>
<span class="sd">            - more_formats (type=list): The extra formats to save the plot in.</span>
<span class="sd">            - save_code (flag): Whether to save the code to generate the plot.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - &lt;more&gt;: Other arguments passed to [`biopipen.utils::VizDEGs()`](https://pwwang.github.io/biopipen.utils.R/reference/VizDEGs.html).</span>
<span class="sd">        allmarker_plots (type=json): All marker plot cases.</span>
<span class="sd">            The keys are the names of the cases and the values are the dicts inherited from `allmarker_plots_defaults`.</span>
<span class="sd">        allenrich_plots_defaults (ns): Default options for the plots to generate for the enrichment analysis.</span>
<span class="sd">            - plot_type: The type of the plot.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://pwwang.github.io/scplotter/reference/EnrichmentPlot.html&gt;.</span>
<span class="sd">        allenrich_plots (type=json): Cases of the plots to generate for the enrichment analysis.</span>
<span class="sd">            The keys are the names of the cases and the values are the dicts inherited from `allenrich_plots_defaults`.</span>
<span class="sd">            The cases under `envs.cases` can inherit this options.</span>
<span class="sd">        marker_plots_defaults (ns): Default options for the plots to generate for the markers.</span>
<span class="sd">            - plot_type: The type of the plot.</span>
<span class="sd">                See &lt;https://pwwang.github.io/biopipen.utils.R/reference/VizDEGs.html&gt;.</span>
<span class="sd">                Available types are `violin`, `box`, `bar`, `ridge`, `dim`, `heatmap` and `dot`.</span>
<span class="sd">                There are two additional types available - `volcano_pct` and `volcano_log2fc`.</span>
<span class="sd">            - more_formats (type=list): The extra formats to save the plot in.</span>
<span class="sd">            - save_code (flag): Whether to save the code to generate the plot.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - &lt;more&gt;: Other arguments passed to [`biopipen.utils::VizDEGs()`](https://pwwang.github.io/biopipen.utils.R/reference/VizDEGs.html).</span>
<span class="sd">                If `plot_type` is `volcano_pct` or `volcano_log2fc`, they will be passed to</span>
<span class="sd">                [`scplotter::VolcanoPlot()`](https://pwwang.github.io/plotthis/reference/VolcanoPlot.html).</span>
<span class="sd">        marker_plots (type=json): Cases of the plots to generate for the markers.</span>
<span class="sd">            Plot cases. The keys are the names of the cases and the values are the dicts inherited from `marker_plots_defaults`.</span>
<span class="sd">            The cases under `envs.cases` can inherit this options.</span>
<span class="sd">        enrich_plots_defaults (ns): Default options for the plots to generate for the enrichment analysis.</span>
<span class="sd">            - plot_type: The type of the plot.</span>
<span class="sd">                See &lt;https://pwwang.github.io/scplotter/reference/EnrichmentPlot.html&gt;.</span>
<span class="sd">                Available types are `bar`, `dot`, `lollipop`, `network`, `enrichmap` and `wordcloud`.</span>
<span class="sd">            - more_formats (type=list): The extra formats to save the plot in.</span>
<span class="sd">            - save_code (flag): Whether to save the code to generate the plot.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://pwwang.github.io/scplotter/reference/EnrichmentPlot.html&gt;.</span>
<span class="sd">        enrich_plots (type=json): Cases of the plots to generate for the enrichment analysis.</span>
<span class="sd">            The keys are the names of the cases and the values are the dicts inherited from `enrich_plots_defaults`.</span>
<span class="sd">            The cases under `envs.cases` can inherit this options.</span>
<span class="sd">        overlaps_defaults (ns): Default options for investigating the overlapping of significant markers between different cases or comparisons.</span>
<span class="sd">            This means either `ident_1` should be empty, so that they can be expanded to multiple comparisons.</span>
<span class="sd">            - sigmarkers: The expression to filter the significant markers for each case.</span>
<span class="sd">                If not provided, `envs.sigmarkers` will be used.</span>
<span class="sd">            - plot_type (choice): The type of the plot to generate for the overlaps.</span>
<span class="sd">                - venn: Use `plotthis::VennDiagram()`.</span>
<span class="sd">                - upset: Use `plotthis::UpsetPlot()`.</span>
<span class="sd">            - more_formats (type=list): The extra formats to save the plot in.</span>
<span class="sd">            - save_code (flag): Whether to save the code to generate the plot.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - &lt;more&gt;: More arguments pased to `plotthis::VennDiagram()`</span>
<span class="sd">                (&lt;https://pwwang.github.io/plotthis/reference/venndiagram1.html&gt;)</span>
<span class="sd">                or `plotthis::UpsetPlot()`</span>
<span class="sd">                (&lt;https://pwwang.github.io/plotthis/reference/upsetplot1.html&gt;)</span>
<span class="sd">        overlaps (type=json): Cases for investigating the overlapping of significant markers between different cases or comparisons.</span>
<span class="sd">            The keys are the names of the cases and the values are the dicts inherited from `overlaps_defaults`.</span>
<span class="sd">            There are two situations that we can perform overlaps:</span>
<span class="sd">            1. If `ident_1` is not specified, the overlaps can be performed between different comparisons.</span>
<span class="sd">            2. If `each` is specified, the overlaps can be performed between different cases, where in each case, `ident_1` must be specified.</span>
<span class="sd">        cases (type=json): If you have multiple cases for marker discovery, you can specify them</span>
<span class="sd">            here. The keys are the names of the cases and the values are the above options. If some options are</span>
<span class="sd">            not specified, the default values specified above (under `envs`) will be used.</span>
<span class="sd">            If no cases are specified, the default case will be added with the default values under `envs` with the name `Marker Discovery`.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.srtobj | stem0}}.markers&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;group_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;ident_1&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;ident_2&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;each&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;dbs&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;KEGG_2021_Human&quot;</span><span class="p">,</span> <span class="s2">&quot;MSigDB_Hallmark_2020&quot;</span><span class="p">],</span>
        <span class="s2">&quot;sigmarkers&quot;</span><span class="p">:</span> <span class="s2">&quot;p_val_adj &lt; 0.05&quot;</span><span class="p">,</span>
        <span class="s2">&quot;enrich_style&quot;</span><span class="p">:</span> <span class="s2">&quot;enrichr&quot;</span><span class="p">,</span>
        <span class="s2">&quot;assay&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;error&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;cache&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">tmpdir</span><span class="p">,</span>
        <span class="s2">&quot;rest&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;allmarker_plots_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;save_code&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;allmarker_plots&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;allenrich_plots_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;heatmap&quot;</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;allenrich_plots&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;marker_plots_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;save_code&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;marker_plots&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Volcano Plot (diff_pct)&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;volcano_pct&quot;</span><span class="p">},</span>
            <span class="s2">&quot;Volcano Plot (log2FC)&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;volcano_log2fc&quot;</span><span class="p">},</span>
            <span class="s2">&quot;Dot Plot&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;dot&quot;</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;enrich_plots_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;save_code&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;enrich_plots&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Bar Plot&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;bar&quot;</span><span class="p">,</span> <span class="s2">&quot;ncol&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;top_term&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;overlaps_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;sigmarkers&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;venn&quot;</span><span class="p">,</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;save_code&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;overlaps&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;cases&quot;</span><span class="p">:</span> <span class="p">{},</span>
    <span class="p">}</span>
    <span class="n">order</span> <span class="o">=</span> <span class="mi">5</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/MarkersFinder.R&quot;</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/scrna/MarkersFinder.svelte&quot;</span><span class="p">,</span>
        <span class="s2">&quot;report_paging&quot;</span><span class="p">:</span> <span class="mi">8</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">TopExpressingGenes</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.TopExpressingGenes"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.TopExpressingGenes" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.TopExpressingGenes">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find the top expressing genes in each cluster</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object in RDS or qs/qs2 format</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory for the tables and plots</span>

<span class="sd">    Envs:</span>
<span class="sd">        mutaters (type=json): The mutaters to mutate the metadata.</span>
<span class="sd">            You can also use the clone selectors to select the TCR clones/clusters.</span>
<span class="sd">            See &lt;https://pwwang.github.io/scplotter/reference/clone_selectors.html&gt;.</span>
<span class="sd">        ident: The group of cells to find the top expressing genes.</span>
<span class="sd">            The cells will be selected by the `group_by` column with this</span>
<span class="sd">            `ident` value in metadata.</span>
<span class="sd">            If not provided, the top expressing genes will be found for all</span>
<span class="sd">            groups of cells in the `group_by` column.</span>
<span class="sd">        group_by: The column name in metadata to group the cells.</span>
<span class="sd">        each: The column name in metadata to separate the cells into different</span>
<span class="sd">            cases.</span>
<span class="sd">        dbs (list): The dbs to do enrichment analysis for significant markers.</span>
<span class="sd">            You can use built-in dbs in `enrichit`, or provide your own gmt files.</span>
<span class="sd">            See also &lt;https://pwwang.github.io/enrichit/reference/FetchGMT.html&gt;.</span>
<span class="sd">            The built-in dbs include:</span>
<span class="sd">            * &quot;BioCarta&quot; or &quot;BioCarta_2016&quot;</span>
<span class="sd">            * &quot;GO_Biological_Process&quot; or &quot;GO_Biological_Process_2025&quot;</span>
<span class="sd">            * &quot;GO_Cellular_Component&quot; or &quot;GO_Cellular_Component_2025&quot;</span>
<span class="sd">            * &quot;GO_Molecular_Function&quot; or &quot;GO_Molecular_Function_2025&quot;</span>
<span class="sd">            * &quot;KEGG&quot;, &quot;KEGG_Human&quot;, &quot;KEGG_2021&quot;, or &quot;KEGG_2021_Human&quot;</span>
<span class="sd">            * &quot;Hallmark&quot;, &quot;MSigDB_Hallmark&quot;, or &quot;MSigDB_Hallmark_2020&quot;</span>
<span class="sd">            * &quot;Reactome&quot;, &quot;Reactome_Pathways&quot;, or &quot;Reactome_Pathways_2024&quot;</span>
<span class="sd">            * &quot;WikiPathways&quot;, &quot;WikiPathways_2024&quot;, &quot;WikiPathways_Human&quot;, or &quot;WikiPathways_2024_Human&quot;</span>
<span class="sd">            You can also fetch more dbs from &lt;https://maayanlab.cloud/Enrichr/#libraries&gt;.</span>
<span class="sd">        n (type=int): The number of top expressing genes to find.</span>
<span class="sd">        enrich_style (choice): The style of the enrichment analysis.</span>
<span class="sd">            The enrichment analysis will be done by `EnrichIt()` from [`enrichit`](https://pwwang.github.io/enrichit/).</span>
<span class="sd">            Two styles are available:</span>
<span class="sd">            - enrichr: `enrichr` style enrichment analysis (fisher&#39;s exact test will be used).</span>
<span class="sd">            - clusterprofiler: `clusterProfiler` style enrichment analysis (hypergeometric test will be used).</span>
<span class="sd">            - clusterProfiler: alias for `clusterprofiler`</span>
<span class="sd">        enrich_plots_defaults (ns): Default options for the plots to generate for the enrichment analysis.</span>
<span class="sd">            - plot_type: The type of the plot.</span>
<span class="sd">                See &lt;https://pwwang.github.io/scplotter/reference/EnrichmentPlot.html&gt;.</span>
<span class="sd">                Available types are `bar`, `dot`, `lollipop`, `network`, `enrichmap` and `wordcloud`.</span>
<span class="sd">            - more_formats (type=list): The extra formats to save the plot in.</span>
<span class="sd">            - save_code (flag): Whether to save the code to generate the plot.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://pwwang.github.io/scplotter/reference/EnrichmentPlot.htmll&gt;.</span>
<span class="sd">        enrich_plots (type=json): Cases of the plots to generate for the enrichment analysis.</span>
<span class="sd">            The keys are the names of the cases and the values are the dicts inherited from `enrich_plots_defaults`.</span>
<span class="sd">            The cases under `envs.cases` can inherit this options.</span>
<span class="sd">        subset: An expression to subset the cells for each case.</span>
<span class="sd">        cases (type=json): If you have multiple cases, you can specify them</span>
<span class="sd">            here. The keys are the names of the cases and the values are the</span>
<span class="sd">            above options except `mutaters`. If some options are</span>
<span class="sd">            not specified, the default values specified above will be used.</span>
<span class="sd">            If no cases are specified, the default case will be added with</span>
<span class="sd">            the default values under `envs` with the name `Top Expressing Genes`.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.srtobj | stem}}.top_expressing_genes&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/TopExpressingGenes.R&quot;</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;ident&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;group_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;each&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;dbs&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;KEGG_2021_Human&quot;</span><span class="p">,</span> <span class="s2">&quot;MSigDB_Hallmark_2020&quot;</span><span class="p">],</span>
        <span class="s2">&quot;n&quot;</span><span class="p">:</span> <span class="mi">250</span><span class="p">,</span>
        <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;enrich_style&quot;</span><span class="p">:</span> <span class="s2">&quot;enrichr&quot;</span><span class="p">,</span>
        <span class="s2">&quot;enrich_plots_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;save_code&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;enrich_plots&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Bar Plot&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;bar&quot;</span><span class="p">,</span> <span class="s2">&quot;ncol&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;top_term&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;cases&quot;</span><span class="p">:</span> <span class="p">{},</span>
    <span class="p">}</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/common.svelte&quot;</span><span class="p">,</span>
        <span class="s2">&quot;report_paging&quot;</span><span class="p">:</span> <span class="mi">8</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">ExprImputation</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.ExprImputation"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.ExprImputation" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.ExprImputation">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;This process imputes the dropout values in scRNA-seq data.</span>

<span class="sd">    It takes the Seurat object as input and outputs the Seurat object with</span>
<span class="sd">    imputed expression data.</span>

<span class="sd">    Reference:</span>
<span class="sd">    - [Linderman, George C., Jun Zhao, and Yuval Kluger. &quot;Zero-preserving imputation of scRNA-seq data using low-rank approximation.&quot; BioRxiv (2018): 397588.](https://www.nature.com/articles/s41467-021-27729-z)</span>
<span class="sd">    - [Li, Wei Vivian, and Jingyi Jessica Li. &quot;An accurate and robust imputation method scImpute for single-cell RNA-seq data.&quot; Nature communications 9.1 (2018): 997.](https://www.nature.com/articles/s41467-018-03405-7)</span>
<span class="sd">    - [Dijk, David van, et al. &quot;MAGIC: A diffusion-based imputation method reveals gene-gene interactions in single-cell RNA-sequencing data.&quot; BioRxiv (2017): 111591.](https://www.cell.com/cell/abstract/S0092-8674(18)30724-4)</span>

<span class="sd">    Input:</span>
<span class="sd">        infile: The input file in RDS/qs format of Seurat object</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output file in RDS format of Seurat object</span>
<span class="sd">            Note that with rmagic and alra, the original default assay will be</span>
<span class="sd">            renamed to `RAW` and the imputed RNA assay will be</span>
<span class="sd">            renamed to `RNA` and set as default assay.</span>

<span class="sd">    Envs:</span>
<span class="sd">        tool (choice): Either alra, scimpute or rmagic</span>
<span class="sd">            - alra: Use RunALRA() from Seurat</span>
<span class="sd">            - scimpute: Use scImpute() from scimpute</span>
<span class="sd">            - rmagic: Use magic() from Rmagic</span>
<span class="sd">        scimpute_args (ns): The arguments for scimpute</span>
<span class="sd">            - drop_thre (type=float): The dropout threshold</span>
<span class="sd">            - kcluster (type=int): Number of clusters to use</span>
<span class="sd">            - ncores (type=int): Number of cores to use</span>
<span class="sd">            - refgene: The reference gene file</span>
<span class="sd">        rmagic_args (ns): The arguments for rmagic</span>
<span class="sd">            - python: The python path where magic-impute is installed.</span>
<span class="sd">            - threshold (type=float): The threshold for magic imputation.</span>
<span class="sd">              Only the genes with dropout rates greater than this threshold (No. of</span>
<span class="sd">              cells with non-zero expression / total number of cells) will be imputed.</span>
<span class="sd">        alra_args (type=json): The arguments for `RunALRA()`</span>

<span class="sd">    Requires:</span>
<span class="sd">        r-scimpute:</span>
<span class="sd">            - if: {{proc.envs.tool == &quot;scimpute&quot;}}</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(scImpute)&quot;)</span>
<span class="sd">        r-rmagic:</span>
<span class="sd">            - if: {{proc.envs.tool == &quot;rmagic&quot;}}</span>
<span class="sd">            - check: |</span>
<span class="sd">                {{proc.lang}} &lt;(\</span>
<span class="sd">                    echo &quot;\</span>
<span class="sd">                        tryCatch(\</span>
<span class="sd">                            { setwd(dirname(Sys.getenv(&#39;CONDA_PREFIX&#39;))) }, \</span>
<span class="sd">                            error = function(e) NULL \</span>
<span class="sd">                        ); \</span>
<span class="sd">                        library(Rmagic)\</span>
<span class="sd">                    &quot;\</span>
<span class="sd">                )</span>
<span class="sd">        magic-impute:</span>
<span class="sd">            - if: {{proc.envs.tool == &quot;rmagic&quot;}}</span>
<span class="sd">            - check: {{proc.envs.rmagic_args.python}} -c &quot;import magic&quot;)</span>
<span class="sd">        r-dplyr:</span>
<span class="sd">            - if: {{proc.envs.tool == &quot;scimpute&quot;}}</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(dplyr)&quot;)</span>
<span class="sd">        r-seurat:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(Seurat)&quot;)</span>
<span class="sd">        r-seuratwrappers:</span>
<span class="sd">            - if: {{proc.envs.tool == &quot;alra&quot;}}</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(SeuratWrappers)&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.infile | stem}}.imputed.qs&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;tool&quot;</span><span class="p">:</span> <span class="s2">&quot;alra&quot;</span><span class="p">,</span>
        <span class="s2">&quot;rmagic_args&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;python&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">exe</span><span class="o">.</span><span class="n">magic_python</span><span class="p">,</span> <span class="s2">&quot;threshold&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">},</span>
        <span class="s2">&quot;scimpute_args&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;drop_thre&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span>
            <span class="s2">&quot;kcluster&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
            <span class="s2">&quot;refgene&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">refgene</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;alra_args&quot;</span><span class="p">:</span> <span class="p">{},</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/ExprImputation.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">SCImpute</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.SCImpute"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.SCImpute" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.SCImpute">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Impute the dropout values in scRNA-seq data.</span>

<span class="sd">    Deprecated. Use `ExprImputation` instead.</span>

<span class="sd">    Input:</span>
<span class="sd">        infile: The input file for imputation</span>
<span class="sd">            Either a SeuratObject or a matrix of count/TPM</span>
<span class="sd">        groupfile: The file to subset the matrix or label the cells</span>
<span class="sd">            Could be an output from ImmunarchFilter</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output matrix</span>

<span class="sd">    Envs:</span>
<span class="sd">        infmt: The input format.</span>
<span class="sd">            Either `seurat` or `matrix`</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infile:file, groupfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;outfile:file:{{in.infile | stem | replace: &#39;.seurat&#39;, &#39;&#39;}}.&quot;</span> <span class="s2">&quot;{{envs.outfmt}}&quot;</span>
    <span class="p">]</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;infmt&quot;</span><span class="p">:</span> <span class="s2">&quot;seurat&quot;</span><span class="p">,</span>  <span class="c1"># or matrix</span>
        <span class="s2">&quot;outfmt&quot;</span><span class="p">:</span> <span class="s2">&quot;txt&quot;</span><span class="p">,</span>  <span class="c1"># or csv, rds</span>
        <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;count&quot;</span><span class="p">,</span>  <span class="c1"># or TPM</span>
        <span class="s2">&quot;drop_thre&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span>
        <span class="s2">&quot;kcluster&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>  <span class="c1"># or Kcluster</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;refgene&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">refgene</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/SCImpute.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">SeuratFilter</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.SeuratFilter"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.SeuratFilter" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.SeuratFilter">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Filtering cells from a seurat object</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object in RDS</span>
<span class="sd">        filters: The filters to apply. Could be a file or string in TOML, or</span>
<span class="sd">            a python dictionary, with following keys:</span>
<span class="sd">            - mutaters: Create new columns in the metadata</span>
<span class="sd">            - filter: A R expression that will pass to</span>
<span class="sd">              `subset(sobj, subset = ...)` to filter the cells</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The filtered seurat object in RDS</span>

<span class="sd">    Envs:</span>
<span class="sd">        invert: Invert the selection?</span>

<span class="sd">    Requires:</span>
<span class="sd">        r-seurat:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(&#39;Seurat&#39;)&quot;)</span>
<span class="sd">        r-dplyr:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(&#39;dplyr&#39;)&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file, filters:var&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.srtobj | stem}}.filtered.RDS&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;invert&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/SeuratFilter.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">SeuratSubset</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.SeuratSubset"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.SeuratSubset" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.SeuratSubset">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Subset a seurat object into multiple seruat objects</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object in RDS</span>
<span class="sd">        subsets: The subsettings to apply. Could be a file or string in TOML, or</span>
<span class="sd">            a python dictionary, with following keys:</span>
<span class="sd">            - &lt;name&gt;: Name of the case</span>
<span class="sd">                mutaters: Create new columns in the metadata</span>
<span class="sd">                subset: A R expression that will pass to</span>
<span class="sd">                    `subset(sobj, subset = ...)`</span>
<span class="sd">                groupby: The column to group by, each value will be a case</span>
<span class="sd">                    If groupby is given, subset will be ignored, each value</span>
<span class="sd">                    of the groupby column will be a case</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory with the subset seurat objects</span>

<span class="sd">    Envs:</span>
<span class="sd">        ignore_nas: Ignore NA values?</span>

<span class="sd">    Requires:</span>
<span class="sd">        r-seurat:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(&#39;Seurat&#39;)&quot;)</span>
<span class="sd">        r-dplyr:</span>
<span class="sd">            - check: {{proc.lang}} &lt;(echo &quot;library(&#39;dplyr&#39;)&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file, subsets:var&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.srtobj | stem}}.subsets&quot;</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ignore_nas&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/SeuratSubset.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">SeuratSplit</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.SeuratSplit"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.SeuratSplit" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.SeuratSplit">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Split a seurat object into multiple seruat objects</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object in RDS</span>
<span class="sd">        by: The metadata column to split by</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory with the subset seurat objects</span>

<span class="sd">    Envs:</span>
<span class="sd">        by: The metadata column to split by</span>
<span class="sd">            Ignored if `by` is given in the input</span>
<span class="sd">        recell: Rename the cell ids using the `by` column</span>
<span class="sd">            A string of R function taking the original cell ids and `by`</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file, by:var&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.srtobj | stem}}.subsets&quot;</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;recell&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>  <span class="c1"># &quot;function(cellid, by) {}&quot;,</span>
    <span class="p">}</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/SeuratSplit.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">Subset10X</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.Subset10X"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.Subset10X" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.Subset10X">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Subset 10X data, mostly used for testing</span>

<span class="sd">    Requires r-matrix to load matrix.mtx.gz</span>

<span class="sd">    Input:</span>
<span class="sd">        indir: The input directory</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory</span>

<span class="sd">    Envs:</span>
<span class="sd">        seed: The seed for random number generator</span>
<span class="sd">        nfeats: The number of features to keep.</span>
<span class="sd">            If &lt;=1 then it will be the percentage of features to keep</span>
<span class="sd">        ncells: The number of cells to keep.</span>
<span class="sd">            If &lt;=1 then it will be the percentage of cells to keep</span>
<span class="sd">        feats_to_keep: The features/genes to keep.</span>
<span class="sd">            The final features list will be `feats_to_keep` + `nfeats`</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;indir:dir&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.indir | stem}}&quot;</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;seed&quot;</span><span class="p">:</span> <span class="mi">8525</span><span class="p">,</span>
        <span class="s2">&quot;nfeats&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">,</span>
        <span class="s2">&quot;ncells&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">,</span>
        <span class="s2">&quot;feats_to_keep&quot;</span><span class="p">:</span> <span class="p">[],</span>
    <span class="p">}</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/Subset10X.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">SeuratTo10X</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.SeuratTo10X"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.SeuratTo10X" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.SeuratTo10X">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Write a Seurat object to 10X format</span>

<span class="sd">    using `write10xCounts` from `DropletUtils`</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object in RDS</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory.</span>
<span class="sd">            When `envs.split_by` is specified, the subdirectories will be</span>
<span class="sd">            created for each distinct value of the column.</span>
<span class="sd">            Otherwise, the matrices will be written to the output directory.</span>

<span class="sd">    Envs:</span>
<span class="sd">        version: The version of 10X format</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.srtobj | stem}}&quot;</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;version&quot;</span><span class="p">:</span> <span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="s2">&quot;split_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/SeuratTo10X.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">ScFGSEA</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.ScFGSEA"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.ScFGSEA" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.ScFGSEA">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Gene set enrichment analysis for cells in different groups using `fgsea`</span>

<span class="sd">    This process allows us to do Gene Set Enrichment Analysis (GSEA) on the expression data,</span>
<span class="sd">    but based on variaties of grouping, including the from the meta data and the</span>
<span class="sd">    scTCR-seq data as well.</span>

<span class="sd">    The GSEA is done using the</span>
<span class="sd">    [fgsea](https://bioconductor.org/packages/release/bioc/html/fgsea.html) package,</span>
<span class="sd">    which allows to quickly and accurately calculate arbitrarily low GSEA P-values</span>
<span class="sd">    for a collection of gene sets.</span>
<span class="sd">    The fgsea package is based on the fast algorithm for preranked GSEA described in</span>
<span class="sd">    [Subramanian et al. 2005](https://www.pnas.org/content/102/43/15545).</span>

<span class="sd">    For each case, the process will generate a table with the enrichment scores for</span>
<span class="sd">    each gene set, and GSEA plots for the top gene sets.</span>

<span class="sd">    Examples:</span>
<span class="sd">        ### The summary and GSEA plots</span>

<span class="sd">        ![GSEA summary](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/scfgsea/ScFGSEA/sampleinfo.fgsea/seurat_clusters/c1/summary.png){: width=&quot;80%&quot;}</span>

<span class="sd">        ![GSEA plot](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/scfgsea/ScFGSEA/sampleinfo.fgsea/seurat_clusters/c1/pathways.png){: width=&quot;80%&quot;}</span>

<span class="sd">        ### Summary plot for all subsets or idents</span>

<span class="sd">        If you use `each` to separate the cells into different subsets, this is useful to</span>
<span class="sd">        make a summary plot for all subsets. Or if you don&#39;t specify `ident_1`, the summary plot for all idents in `group_by` will be generated.</span>

<span class="sd">        ```toml</span>
<span class="sd">        [ScFGSEA.envs]</span>
<span class="sd">        group_by = &quot;Diagnosis&quot;</span>
<span class="sd">        ident_1 = &quot;Colitis&quot;</span>
<span class="sd">        ident_2 = &quot;Control&quot;</span>
<span class="sd">        each = &quot;seurat_clusters&quot;</span>

<span class="sd">        [ScFGSEA.envs.alleach_plots.Heatmap]</span>
<span class="sd">        plot_type = &quot;heatmap&quot;</span>
<span class="sd">        group_by = &quot;Diagnosis&quot;</span>
<span class="sd">        ```</span>

<span class="sd">        ![GSEA summary for all subsets](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/scfgsea/ScFGSEA/sampleinfo.fgsea/GSEA-all-seurat_clusters-/all.Heatmap.png){: width=&quot;80%&quot;}</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object in RDS format</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory for the results and plots</span>

<span class="sd">    Envs:</span>
<span class="sd">        ncores (type=int): Number of cores for parallelization</span>
<span class="sd">            Passed to `nproc` of `fgseaMultilevel()`.</span>
<span class="sd">        cache (type=auto): Where to cache the results.</span>
<span class="sd">            If `True`, cache to `outdir` of the job. If `False`, don&#39;t cache.</span>
<span class="sd">            Otherwise, specify the directory to cache to.</span>
<span class="sd">        mutaters (type=json): The mutaters to mutate the metadata.</span>
<span class="sd">            The key-value pairs will be passed the `dplyr::mutate()` to mutate the metadata.</span>
<span class="sd">            You can also use the clone selectors to select the TCR clones/clusters.</span>
<span class="sd">            See &lt;https://pwwang.github.io/scplotter/reference/clone_selectors.html&gt;.</span>

<span class="sd">        group_by: The column name in metadata to group the cells.</span>
<span class="sd">        ident_1: The first group of cells to compare</span>
<span class="sd">        ident_2: The second group of cells to compare, if not provided, the rest of the cells that are not `NA`s in `group_by` column are used for `ident_2`.</span>
<span class="sd">        assay: The assay to use. If not provided, the default assay will be used.</span>
<span class="sd">        each: The column name in metadata to separate the cells into different subsets to do the analysis.</span>
<span class="sd">        subset: An expression to subset the cells.</span>
<span class="sd">        gmtfile: The pathways in GMT format, with the gene names/ids in the same format as the seurat object.</span>
<span class="sd">            You can use built-in dbs in `enrichit`, or provide your own gmt files.</span>
<span class="sd">            See also &lt;https://pwwang.github.io/enrichit/reference/FetchGMT.html&gt;.</span>
<span class="sd">            The built-in dbs include:</span>
<span class="sd">            * &quot;BioCarta&quot; or &quot;BioCarta_2016&quot;</span>
<span class="sd">            * &quot;GO_Biological_Process&quot; or &quot;GO_Biological_Process_2025&quot;</span>
<span class="sd">            * &quot;GO_Cellular_Component&quot; or &quot;GO_Cellular_Component_2025&quot;</span>
<span class="sd">            * &quot;GO_Molecular_Function&quot; or &quot;GO_Molecular_Function_2025&quot;</span>
<span class="sd">            * &quot;KEGG&quot;, &quot;KEGG_Human&quot;, &quot;KEGG_2021&quot;, or &quot;KEGG_2021_Human&quot;</span>
<span class="sd">            * &quot;Hallmark&quot;, &quot;MSigDB_Hallmark&quot;, or &quot;MSigDB_Hallmark_2020&quot;</span>
<span class="sd">            * &quot;Reactome&quot;, &quot;Reactome_Pathways&quot;, or &quot;Reactome_Pathways_2024&quot;</span>
<span class="sd">            * &quot;WikiPathways&quot;, &quot;WikiPathways_2024&quot;, &quot;WikiPathways_Human&quot;, or &quot;WikiPathways_2024_Human&quot;</span>
<span class="sd">            You can also fetch more dbs from &lt;https://maayanlab.cloud/Enrichr/#libraries&gt;.</span>
<span class="sd">        method (choice): The method to do the preranking.</span>
<span class="sd">            - signal_to_noise: Signal to noise.</span>
<span class="sd">                The larger the differences of the means (scaled by the standard deviations);</span>
<span class="sd">                that is, the more distinct the gene expression is in each phenotype and the more the gene</span>
<span class="sd">                acts as a &quot;class marker&quot;.</span>
<span class="sd">            - s2n: Alias of signal_to_noise.</span>
<span class="sd">            - abs_signal_to_noise: The absolute value of signal_to_noise.</span>
<span class="sd">            - abs_s2n: Alias of abs_signal_to_noise.</span>
<span class="sd">            - t_test: T test.</span>
<span class="sd">                Uses the difference of means scaled by the standard deviation and number of samples.</span>
<span class="sd">            - ratio_of_classes: Also referred to as fold change.</span>
<span class="sd">                Uses the ratio of class means to calculate fold change for natural scale data.</span>
<span class="sd">            - diff_of_classes: Difference of class means.</span>
<span class="sd">                Uses the difference of class means to calculate fold change for nature scale data</span>
<span class="sd">            - log2_ratio_of_classes: Log2 ratio of class means.</span>
<span class="sd">                Uses the log2 ratio of class means to calculate fold change for natural scale data.</span>
<span class="sd">                This is the recommended statistic for calculating fold change for log scale data.</span>
<span class="sd">        top (type=auto): Do gsea table and enrich plot for top N pathways.</span>
<span class="sd">            If it is &lt; 1, will apply it to `padj`, selecting pathways with `padj` &lt; `top`.</span>
<span class="sd">        eps (type=float): This parameter sets the boundary for calculating the p value.</span>
<span class="sd">            See &lt;https://rdrr.io/bioc/fgsea/man/fgseaMultilevel.html&gt;</span>
<span class="sd">        alleach_plots_defaults (ns): Default options for the plots to generate for all pathways.</span>
<span class="sd">            - plot_type: The type of the plot, currently either dot or heatmap (default)</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://pwwang.github.io/biopipen.utils.R/reference/VizGSEA.html&gt;.</span>
<span class="sd">        alleach_plots (type=json): Cases of the plots to generate for all pathways.</span>
<span class="sd">            The keys are the names of the cases and the values are the dicts inherited from `alleach_plots_defaults`.</span>
<span class="sd">        minsize (type=int): Minimal size of a gene set to test. All pathways below the threshold are excluded.</span>
<span class="sd">        maxsize (type=int): Maximal size of a gene set to test. All pathways above the threshold are excluded.</span>
<span class="sd">        rest (type=json;order=98): Rest arguments for [`fgsea()`](https://rdrr.io/bioc/fgsea/man/fgsea.html)</span>
<span class="sd">            See also &lt;https://rdrr.io/bioc/fgsea/man/fgseaMultilevel.html&gt;</span>
<span class="sd">        cases (type=json;order=99): If you have multiple cases, you can specify them here.</span>
<span class="sd">            The keys are the names of the cases and the values are the above options except `mutaters`.</span>
<span class="sd">            If some options are not specified, the default values specified above will be used.</span>
<span class="sd">            If no cases are specified, the default case will be added with the name `GSEA`.</span>

<span class="sd">    Requires:</span>
<span class="sd">        bioconductor-fgsea:</span>
<span class="sd">            - check: {{proc.lang}} -e &quot;library(fgsea)&quot;</span>
<span class="sd">        r-seurat:</span>
<span class="sd">            - check: {{proc.lang}} -e &quot;library(seurat)&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{(in.casefile or in.srtobj) | stem0}}.fgsea&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;cache&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">tmpdir</span><span class="p">,</span>
        <span class="s2">&quot;assay&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;group_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;ident_1&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;ident_2&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;each&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;gmtfile&quot;</span><span class="p">:</span> <span class="s2">&quot;KEGG_2021_Human&quot;</span><span class="p">,</span>
        <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;s2n&quot;</span><span class="p">,</span>
        <span class="s2">&quot;top&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span>
        <span class="s2">&quot;minsize&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
        <span class="s2">&quot;maxsize&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span>
        <span class="s2">&quot;eps&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s2">&quot;alleach_plots_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;heatmap&quot;</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;alleach_plots&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;rest&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;cases&quot;</span><span class="p">:</span> <span class="p">{},</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/ScFGSEA.R&quot;</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/common.svelte&quot;</span><span class="p">,</span>
        <span class="s2">&quot;report_paging&quot;</span><span class="p">:</span> <span class="mi">8</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">CellTypeAnnotation</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.CellTypeAnnotation"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.CellTypeAnnotation" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.CellTypeAnnotation">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Annotate the cell clusters. Currently, four ways are supported:</span>

<span class="sd">    1. Pass the cell type annotation directly</span>
<span class="sd">    2. Use [`ScType`](https://github.com/IanevskiAleksandr/sc-type)</span>
<span class="sd">    3. Use [`scCATCH`](https://github.com/ZJUFanLab/scCATCH)</span>
<span class="sd">    4. Use [`hitype`](https://github.com/pwwang/hitype)</span>
<span class="sd">    5. Use [`celltypist`](https://github.com/Teichlab/celltypist)</span>

<span class="sd">    The annotated cell types will replace the original identity column in the metadata,</span>
<span class="sd">    so that the downstream processes will use the annotated cell types.</span>

<span class="sd">    /// Note</span>

<span class="sd">    When cell types are annotated, the original identity column (e.g. `seurat_clusters`) will be renamed</span>
<span class="sd">    to `envs.backup_col` (e.g. `seurat_clusters_id`), and the new identity column will be added.</span>

<span class="sd">    ///</span>

<span class="sd">    If you are using `ScType`, `scCATCH`, or `hitype`, a text file containing the mapping from</span>
<span class="sd">    the original identity to the new cell types will be generated and saved to</span>
<span class="sd">    `cluster2celltype.tsv` under `&lt;workdir&gt;/&lt;pipline_name&gt;/CellTypeAnnotation/0/output/`.</span>

<span class="sd">    Examples:</span>
<span class="sd">        ```toml</span>
<span class="sd">        [CellTypeAnnotation.envs]</span>
<span class="sd">        tool = &quot;direct&quot;</span>
<span class="sd">        cell_types = [&quot;CellType1&quot;, &quot;CellType2&quot;, &quot;-&quot;, &quot;CellType4&quot;]</span>
<span class="sd">        ```</span>

<span class="sd">        The cell types will be assigned as:</span>

<span class="sd">        ```</span>
<span class="sd">        0 -&gt; CellType1</span>
<span class="sd">        1 -&gt; CellType2</span>
<span class="sd">        2 -&gt; 2</span>
<span class="sd">        3 -&gt; CellType4</span>
<span class="sd">        ```</span>

<span class="sd">    Input:</span>
<span class="sd">        sobjfile: The single-cell object in RDS/qs/qs2/h5ad format.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The rds/qs/qs2/h5ad file of seurat object with cell type annotated.</span>
<span class="sd">            A text file containing the mapping from the old identity to the new cell types</span>
<span class="sd">            will be generated and saved to `cluster2celltype.tsv` under the job output directory.</span>
<span class="sd">            Note that if `envs.ident` is specified, the output Seurat object will have</span>
<span class="sd">            the identity set to the specified column in metadata.</span>

<span class="sd">    Envs:</span>
<span class="sd">        tool (choice): The tool to use for cell type annotation.</span>
<span class="sd">            - sctype: Use `scType` to annotate cell types.</span>
<span class="sd">                See &lt;https://github.com/IanevskiAleksandr/sc-type&gt;</span>
<span class="sd">            - hitype: Use `hitype` to annotate cell types.</span>
<span class="sd">                See &lt;https://github.com/pwwang/hitype&gt;</span>
<span class="sd">            - sccatch: Use `scCATCH` to annotate cell types.</span>
<span class="sd">                See &lt;https://github.com/ZJUFanLab/scCATCH&gt;</span>
<span class="sd">            - celltypist: Use `celltypist` to annotate cell types.</span>
<span class="sd">                See &lt;https://github.com/Teichlab/celltypist&gt;</span>
<span class="sd">            - direct: Directly assign cell types</span>
<span class="sd">        sctype_tissue: The tissue to use for `sctype`.</span>
<span class="sd">            Avaiable tissues should be the first column (`tissueType`) of `sctype_db`.</span>
<span class="sd">            If not specified, all rows in `sctype_db` will be used.</span>
<span class="sd">        sctype_db: The database to use for sctype.</span>
<span class="sd">            Check examples at &lt;https://github.com/IanevskiAleksandr/sc-type/blob/master/ScTypeDB_full.xlsx&gt;</span>
<span class="sd">        ident: The column name in metadata to use as the clusters.</span>
<span class="sd">            If not specified, the identity column will be used when input is rds/qs/qs2 (supposing we have a Seurat object).</span>
<span class="sd">            If input data is h5ad, this is required to run cluster-based annotation tools.</span>
<span class="sd">            For `celltypist`, this is a shortcut to set `over_clustering` in `celltypist_args`.</span>
<span class="sd">        backup_col: The backup column name to store the original identities.</span>
<span class="sd">            If not specified, the original identity column will not be stored.</span>
<span class="sd">            If `envs.newcol` is specified, this will be ignored.</span>
<span class="sd">        hitype_tissue: The tissue to use for `hitype`.</span>
<span class="sd">            Avaiable tissues should be the first column (`tissueType`) of `hitype_db`.</span>
<span class="sd">            If not specified, all rows in `hitype_db` will be used.</span>
<span class="sd">        hitype_db: The database to use for hitype.</span>
<span class="sd">            Compatible with `sctype_db`.</span>
<span class="sd">            See also &lt;https://pwwang.github.io/hitype/articles/prepare-gene-sets.html&gt;</span>
<span class="sd">            You can also use built-in databases, including `hitypedb_short`, `hitypedb_full`, and `hitypedb_pbmc3k`.</span>
<span class="sd">        cell_types (type=auto): The cell types to use for direct annotation.</span>
<span class="sd">            If given as a list (array), you can use `&quot;-&quot;` or `&quot;&quot;` as the placeholder for the clusters that</span>
<span class="sd">            you want to keep the original cell types. If the length of `cell_types` is shorter than the number of</span>
<span class="sd">            clusters, the remaining clusters will be kept as the original cell types.</span>
<span class="sd">            You can also use `NA` to remove the clusters from downstream analysis. This</span>
<span class="sd">            only works when `envs.newcol` is not specified.</span>
<span class="sd">            If given as a dict (map), the keys are the original cluster names and the values are the new cell types.</span>

<span class="sd">            /// Note</span>
<span class="sd">            If `tool` is `direct` and `cell_types` is not specified or an empty list,</span>
<span class="sd">            the original cell types will be kept and nothing will be changed.</span>
<span class="sd">            ///</span>

<span class="sd">        more_cell_types (type=json): The additional cell type annotations to add to the metadata.</span>
<span class="sd">            The keys are the new column names and the values are the cell types lists.</span>
<span class="sd">            The cell type lists work the same as `cell_types` above.</span>
<span class="sd">            This is useful when you want to keep multiple annotations of cell types.</span>

<span class="sd">        sccatch_args (ns): The arguments for `scCATCH::findmarkergene()` if `tool` is `sccatch`.</span>
<span class="sd">            - species: The specie of cells.</span>
<span class="sd">            - cancer: If the sample is from cancer tissue, then the cancer type may be defined.</span>
<span class="sd">            - tissue: Tissue origin of cells must be defined.</span>
<span class="sd">            - marker: The marker genes for cell type identification.</span>
<span class="sd">            - if_use_custom_marker (flag): Whether to use custom marker genes. If `True`, no `species`, `cancer`, and `tissue` are needed.</span>
<span class="sd">            - &lt;more&gt;: Other arguments for [`scCATCH::findmarkergene()`](https://rdrr.io/cran/scCATCH/man/findmarkergene.html).</span>
<span class="sd">                You can pass an RDS file to `sccatch_args.marker` to work as custom marker. If so,</span>
<span class="sd">                `if_use_custom_marker` will be set to `TRUE` automatically.</span>
<span class="sd">        celltypist_args (ns): The arguments for `celltypist::celltypist()` if `tool` is `celltypist`.</span>
<span class="sd">            - model: The path to model file.</span>
<span class="sd">            - python: The python path where celltypist is installed.</span>
<span class="sd">            - majority_voting: When true, it refines cell identities within local subclusters after an over-clustering approach</span>
<span class="sd">                at the cost of increased runtime.</span>
<span class="sd">            - over_clustering (type=auto): The column name in metadata to use as clusters for majority voting.</span>
<span class="sd">                Set to `False` to disable over-clustering.</span>
<span class="sd">                When `in.sobjfile` is rds/qs/qs2 (supposing we have a Seurat object), the default ident is used by default.</span>
<span class="sd">                Otherwise, it is False by default.</span>
<span class="sd">            - assay: When converting a Seurat object to AnnData, the assay to use.</span>
<span class="sd">                If input is h5seurat, this defaults to RNA.</span>
<span class="sd">                If input is Seurat object in RDS, this defaults to the default assay.</span>
<span class="sd">        merge (flag): Whether to merge the clusters with the same cell types.</span>
<span class="sd">            Otherwise, a suffix will be added to the cell types (ie. `.1`, `.2`, etc).</span>
<span class="sd">        newcol: The new column name to store the cell types.</span>
<span class="sd">            If not specified, the identity column will be overwritten.</span>
<span class="sd">            If specified, the original identity column will be kept and `Idents` will be kept as the original identity.</span>
<span class="sd">        outtype (choice): The output file type. Currently only works for `celltypist`.</span>
<span class="sd">            An RDS file will be generated for other tools.</span>
<span class="sd">            - input: Use the same file type as the input.</span>
<span class="sd">            - rds: Use RDS file.</span>
<span class="sd">            - qs: Use qs2 file.</span>
<span class="sd">            - qs2: Use qs2 file.</span>
<span class="sd">            - h5ad: Use AnnData file.</span>

<span class="sd">    Requires:</span>
<span class="sd">        r-HGNChelper:</span>
<span class="sd">            - if: {{proc.envs.tool == &#39;sctype&#39;}}</span>
<span class="sd">            - check: {{proc.lang}} -e &quot;library(HGNChelper)&quot;</span>
<span class="sd">        r-seurat:</span>
<span class="sd">            - if: {{proc.envs.tool == &#39;sctype&#39;}}</span>
<span class="sd">            - check: {{proc.lang}} -e &quot;library(Seurat)&quot;</span>
<span class="sd">        r-dplyr:</span>
<span class="sd">            - if: {{proc.envs.tool == &#39;sctype&#39;}}</span>
<span class="sd">            - check: {{proc.lang}} -e &quot;library(dplyr)&quot;</span>
<span class="sd">        r-openxlsx:</span>
<span class="sd">            - if: {{proc.envs.tool == &#39;sctype&#39;}}</span>
<span class="sd">            - check: {{proc.lang}} -e &quot;library(openxlsx)&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;sobjfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;outfile:file:&quot;</span>
        <span class="s2">&quot;{{in.sobjfile | stem}}.annotated.&quot;</span>
        <span class="s2">&quot;{{- ext0(in.sobjfile) if envs.outtype == &#39;input&#39; else envs.outtype -}}&quot;</span>
    <span class="p">)</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;tool&quot;</span><span class="p">:</span> <span class="s2">&quot;hitype&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sctype_tissue&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;sctype_db&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">sctype_db</span><span class="p">,</span>
        <span class="s2">&quot;ident&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;backup_col&quot;</span><span class="p">:</span> <span class="s2">&quot;seurat_clusters_id&quot;</span><span class="p">,</span>
        <span class="s2">&quot;cell_types&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;more_cell_types&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;sccatch_args&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;cancer&quot;</span><span class="p">:</span> <span class="s2">&quot;Normal&quot;</span><span class="p">,</span>
            <span class="s2">&quot;tissue&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;marker&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;if_use_custom_marker&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;hitype_tissue&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;hitype_db&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;celltypist_args&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;model&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;python&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span><span class="p">,</span>
            <span class="s2">&quot;majority_voting&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
            <span class="s2">&quot;over_clustering&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;assay&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;merge&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;newcol&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;outtype&quot;</span><span class="p">:</span> <span class="s2">&quot;input&quot;</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/CellTypeAnnotation.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">SeuratMap2Ref</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.SeuratMap2Ref"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.SeuratMap2Ref" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.SeuratMap2Ref">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Map the seurat object to reference</span>

<span class="sd">    See: &lt;https://satijalab.org/seurat/articles/integration_mapping.html&gt;</span>
<span class="sd">    and &lt;https://satijalab.org/seurat/articles/multimodal_reference_mapping.html&gt;</span>

<span class="sd">    Input:</span>
<span class="sd">        sobjfile: The seurat object</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The rds file of seurat object with cell type annotated.</span>
<span class="sd">            Note that the reduction name will be `ref.umap` for the mapping.</span>
<span class="sd">            To visualize the mapping, you should use `ref.umap` as the reduction name.</span>

<span class="sd">    Envs:</span>
<span class="sd">        ncores (type=int;order=-100): Number of cores to use.</span>
<span class="sd">            When `split_by` is used, this will be the number of cores for each object to map to the reference.</span>
<span class="sd">            When `split_by` is not used, this is used in `future::plan(strategy = &quot;multicore&quot;, workers = &lt;ncores&gt;)`</span>
<span class="sd">            to parallelize some Seurat procedures.</span>
<span class="sd">            See also: &lt;https://satijalab.org/seurat/archive/v3.0/future_vignette.html&gt;</span>
<span class="sd">        mutaters (type=json): The mutaters to mutate the metadata.</span>
<span class="sd">            This is helpful when we want to create new columns for `split_by`.</span>
<span class="sd">        use: A column name of metadata from the reference</span>
<span class="sd">            (e.g. `celltype.l1`, `celltype.l2`) to transfer to the query as the</span>
<span class="sd">            cell types (ident) for downstream analysis. This field is required.</span>
<span class="sd">            If you want to transfer multiple columns, you can use</span>
<span class="sd">            `envs.MapQuery.refdata`.</span>
<span class="sd">        ident: The name of the ident for query transferred from `envs.use` of the reference.</span>
<span class="sd">        ref: The reference seurat object file.</span>
<span class="sd">            Either an RDS file or a h5seurat file that can be loaded by</span>
<span class="sd">            `Seurat::LoadH5Seurat()`.</span>
<span class="sd">            The file type is determined by the extension. `.rds` or `.RDS` for</span>
<span class="sd">            RDS file, `.h5seurat` or `.h5` for h5seurat file.</span>
<span class="sd">        refnorm (choice): Normalization method the reference used. The same method will be used for the query.</span>
<span class="sd">            - LogNormalize: Using [`NormalizeData`](https://satijalab.org/seurat/reference/normalizedata).</span>
<span class="sd">            - SCTransform: Using [`SCTransform`](https://satijalab.org/seurat/reference/sctransform).</span>
<span class="sd">            - SCT: Alias of SCTransform.</span>
<span class="sd">            - auto: Automatically detect the normalization method.</span>
<span class="sd">                If the default assay of reference is `SCT`, then `SCTransform` will be used.</span>
<span class="sd">        split_by: The column name in metadata to split the query into multiple objects.</span>
<span class="sd">            This helps when the original query is too large to process.</span>
<span class="sd">        skip_if_normalized: Skip normalization if the query is already normalized.</span>
<span class="sd">            Since the object is supposed to be generated by `SeuratPreparing`, it is already normalized.</span>
<span class="sd">            However, a different normalization method may be used.</span>
<span class="sd">            If the reference is normalized by the same method as the query, the normalization can be skipped.</span>
<span class="sd">            Otherwise, the normalization cannot be skipped.</span>
<span class="sd">            The normalization method used for the query set is determined by the default assay.</span>
<span class="sd">            If `SCT`, then `SCTransform` is used; otherwise, `NormalizeData` is used.</span>
<span class="sd">            You can set this to `False` to force re-normalization (with or without the arguments previously used).</span>
<span class="sd">        SCTransform (ns): Arguments for [`SCTransform()`](https://satijalab.org/seurat/reference/sctransform)</span>
<span class="sd">            - do-correct-umi (flag): Place corrected UMI matrix in assay counts layer?</span>
<span class="sd">            - do-scale (flag): Whether to scale residuals to have unit variance?</span>
<span class="sd">            - do-center (flag): Whether to center residuals to have mean zero?</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/sctransform&gt;.</span>
<span class="sd">                Note that the hyphen (`-`) will be transformed into `.` for the keys.</span>
<span class="sd">        NormalizeData (ns): Arguments for [`NormalizeData()`](https://satijalab.org/seurat/reference/normalizedata)</span>
<span class="sd">            - normalization-method: Normalization method.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/normalizedata&gt;.</span>
<span class="sd">                Note that the hyphen (`-`) will be transformed into `.` for the keys.</span>
<span class="sd">        FindTransferAnchors (ns): Arguments for [`FindTransferAnchors()`](https://satijalab.org/seurat/reference/findtransferanchors)</span>
<span class="sd">            - normalization-method (choice): Name of normalization method used.</span>
<span class="sd">                - LogNormalize: Log-normalize the data matrix</span>
<span class="sd">                - SCT: Scale data using the SCTransform method</span>
<span class="sd">                - auto: Automatically detect the normalization method.</span>
<span class="sd">                    See `envs.refnorm`.</span>
<span class="sd">            - reference-reduction: Name of dimensional reduction to use from the reference if running the pcaproject workflow.</span>
<span class="sd">                Optionally enables reuse of precomputed reference dimensional reduction.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/findtransferanchors&gt;.</span>
<span class="sd">                Note that the hyphen (`-`) will be transformed into `.` for the keys.</span>
<span class="sd">        MapQuery (ns): Arguments for [`MapQuery()`](https://satijalab.org/seurat/reference/mapquery)</span>
<span class="sd">            - reference-reduction: Name of reduction to use from the reference for neighbor finding</span>
<span class="sd">            - reduction-model: `DimReduc` object that contains the umap model.</span>
<span class="sd">            - refdata (type=json): Extra data to transfer from the reference to the query.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://satijalab.org/seurat/reference/mapquery&gt;.</span>
<span class="sd">                Note that the hyphen (`-`) will be transformed into `.` for the keys.</span>
<span class="sd">        cache (type=auto): Whether to cache the information at different steps.</span>
<span class="sd">            If `True`, the seurat object will be cached in the job output directory, which will be not cleaned up when job is rerunning.</span>
<span class="sd">            The cached seurat object will be saved as `&lt;signature&gt;.&lt;kind&gt;.RDS` file, where `&lt;signature&gt;` is the signature determined by</span>
<span class="sd">            the input and envs of the process.</span>
<span class="sd">            See &lt;https://github.com/satijalab/seurat/issues/7849&gt;, &lt;https://github.com/satijalab/seurat/issues/5358&gt; and</span>
<span class="sd">            &lt;https://github.com/satijalab/seurat/issues/6748&gt; for more details also about reproducibility issues.</span>
<span class="sd">            To not use the cached seurat object, you can either set `cache` to `False` or delete the cached file at</span>
<span class="sd">            `&lt;signature&gt;.RDS` in the cache directory.</span>
<span class="sd">        plots (type=json): The plots to generate.</span>
<span class="sd">            The keys are the names of the plots and the values are the arguments for the plot.</span>
<span class="sd">            The arguments will be passed to `biopipen.utils::VizSeuratMap2Ref()` to generate the plots.</span>
<span class="sd">            The plots will be saved to the output directory.</span>
<span class="sd">            See &lt;https://pwwang.github.io/biopipen.utils.R/reference/VizSeuratMap2Ref.html&gt;.</span>

<span class="sd">    Requires:</span>
<span class="sd">        r-seurat:</span>
<span class="sd">            - check: {{proc.lang}} -e &quot;library(Seurat)&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;sobjfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.sobjfile | stem}}.qs&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs_depth</span> <span class="o">=</span> <span class="mi">3</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;use&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;ident&quot;</span><span class="p">:</span> <span class="s2">&quot;seurat_clusters&quot;</span><span class="p">,</span>
        <span class="s2">&quot;mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;ref&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;refnorm&quot;</span><span class="p">:</span> <span class="s2">&quot;auto&quot;</span><span class="p">,</span>
        <span class="s2">&quot;split_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;skip_if_normalized&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;SCTransform&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;do-correct-umi&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;do-scale&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;do-center&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;NormalizeData&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;normalization-method&quot;</span><span class="p">:</span> <span class="s2">&quot;LogNormalize&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;FindTransferAnchors&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="c1"># &quot;reference-reduction&quot;: &quot;spca&quot;,</span>
        <span class="p">},</span>
        <span class="s2">&quot;MapQuery&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="c1"># &quot;reference-reduction&quot;: &quot;spca&quot;,</span>
            <span class="c1"># &quot;reduction-model&quot;: &quot;wnn.umap&quot;,</span>
            <span class="s2">&quot;refdata&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="c1"># &quot;celltype-l1&quot;: &quot;celltype.l1&quot;,</span>
                <span class="c1"># &quot;celltype-l2&quot;: &quot;celltype.l2&quot;,</span>
                <span class="c1"># &quot;predicted_ADT&quot;: &quot;ADT&quot;,</span>
            <span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;cache&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">tmpdir</span><span class="p">,</span>
        <span class="s2">&quot;plots&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Mapped Identity&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;features&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{ident}</span><span class="s2">:</span><span class="si">{use}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="p">},</span>
            <span class="s2">&quot;Mapping Score&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;features&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{ident}</span><span class="s2">.score&quot;</span><span class="p">,</span>
            <span class="p">},</span>
        <span class="p">},</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/SeuratMap2Ref.R&quot;</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/common.svelte&quot;</span><span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">RadarPlots</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.RadarPlots"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.RadarPlots" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.RadarPlots">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Radar plots for cell proportion in different clusters.</span>

<span class="sd">    This process generates the radar plots for the clusters of T cells.</span>
<span class="sd">    It explores the proportion of cells in different groups (e.g. Tumor vs Blood)</span>
<span class="sd">    in different T-cell clusters.</span>

<span class="sd">    Examples:</span>
<span class="sd">        Let&#39;s say we have a metadata like this:</span>

<span class="sd">        | Cell | Source | Timepoint | seurat_clusters |</span>
<span class="sd">        | ---- | ------ | --------- | --------------- |</span>
<span class="sd">        | A    | Blood  | Pre       | 0               |</span>
<span class="sd">        | B    | Blood  | Pre       | 0               |</span>
<span class="sd">        | C    | Blood  | Post      | 1               |</span>
<span class="sd">        | D    | Blood  | Post      | 1               |</span>
<span class="sd">        | E    | Tumor  | Pre       | 2               |</span>
<span class="sd">        | F    | Tumor  | Pre       | 2               |</span>
<span class="sd">        | G    | Tumor  | Post      | 3               |</span>
<span class="sd">        | H    | Tumor  | Post      | 3               |</span>

<span class="sd">        With configurations:</span>

<span class="sd">        ```toml</span>
<span class="sd">        [RadarPlots.envs]</span>
<span class="sd">        by = &quot;Source&quot;</span>
<span class="sd">        ```</span>

<span class="sd">        Then we will have a radar plots like this:</span>

<span class="sd">        ![Radar plots](https://pwwang.github.io/immunopipe/latest/processes/images/RadarPlots-default.png)</span>

<span class="sd">        We can use `each` to separate the cells into different cases:</span>

<span class="sd">        ```toml</span>
<span class="sd">        [RadarPlots.envs]</span>
<span class="sd">        by = &quot;Source&quot;</span>
<span class="sd">        each = &quot;Timepoint&quot;</span>
<span class="sd">        ```</span>

<span class="sd">        Then we will have two radar plots, one for `Pre` and one for `Post`:</span>

<span class="sd">        ![Radar plots](https://pwwang.github.io/immunopipe/latest/processes/images/RadarPlots-each.png)</span>

<span class="sd">        Using `cluster_order` to change the order of the clusters and show only the first 3 clusters:</span>

<span class="sd">        ```toml</span>
<span class="sd">        [RadarPlots.envs]</span>
<span class="sd">        by = &quot;Source&quot;</span>
<span class="sd">        cluster_order = [&quot;2&quot;, &quot;0&quot;, &quot;1&quot;]</span>
<span class="sd">        breaks = [0, 50, 100]  # also change the breaks</span>
<span class="sd">        ```</span>

<span class="sd">        ![Radar plots cluster_order](https://pwwang.github.io/immunopipe/latest/processes/images/RadarPlots-cluster_order.png)</span>


<span class="sd">        /// Attention</span>
<span class="sd">        All the plots used in the examples are just for demonstration purpose. The real plots will have different appearance.</span>
<span class="sd">        ///</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object in RDS or qs/qs2 format</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory for the plots</span>

<span class="sd">    Envs:</span>
<span class="sd">        mutaters (type=json): Mutaters to mutate the metadata of the</span>
<span class="sd">            seurat object. Keys are the column names and values are the</span>
<span class="sd">            expressions to mutate the columns. These new columns will be</span>
<span class="sd">            used to define your cases.</span>
<span class="sd">        by: Which column to use to separate the cells in different groups.</span>
<span class="sd">            `NA`s will be ignored. For example, If you have a column named `Source`</span>
<span class="sd">            that marks the source of the cells, and you want to separate the cells</span>
<span class="sd">            into `Tumor` and `Blood` groups, you can set `by` to `Source`.</span>
<span class="sd">            The there will be two curves in the radar plot, one for `Tumor` and</span>
<span class="sd">            one for `Blood`.</span>
<span class="sd">        each: A column with values to separate all cells in different cases</span>
<span class="sd">            When specified, the case will be expanded to multiple cases for</span>
<span class="sd">            each value in the column.</span>
<span class="sd">            If specified, `section` will be ignored, and the case name will</span>
<span class="sd">            be used as the section name.</span>
<span class="sd">        prefix_each (flag): Whether to prefix the `each` column name to the values as the</span>
<span class="sd">            case/section name.</span>
<span class="sd">        breakdown: An additional column with groups to break down the cells</span>
<span class="sd">            distribution in each cluster. For example, if you want to see the</span>
<span class="sd">            distribution of the cells in each cluster in different samples. In</span>
<span class="sd">            this case, you should have multiple values in each `by`. These values</span>
<span class="sd">            won&#39;t be plotted in the radar plot, but a barplot will be generated</span>
<span class="sd">            with the mean value of each group and the error bar.</span>
<span class="sd">        test (choice): The test to use to calculate the p values.</span>
<span class="sd">            If there are more than 2 groups in `by`, the p values will be calculated</span>
<span class="sd">            pairwise group by group. Only works when `breakdown` is specified and</span>
<span class="sd">            `by` has 2 groups or more.</span>
<span class="sd">            - wilcox: Wilcoxon rank sum test</span>
<span class="sd">            - t: T test</span>
<span class="sd">            - none: No test will be performed</span>
<span class="sd">        order (list): The order of the values in `by`. You can also limit</span>
<span class="sd">            (filter) the values we have in `by`. For example, if column `Source`</span>
<span class="sd">            has values `Tumor`, `Blood`, `Spleen`, and you only want to plot</span>
<span class="sd">            `Tumor` and `Blood`, you can set `order` to `[&quot;Tumor&quot;, &quot;Blood&quot;]`.</span>
<span class="sd">            This will also have `Tumor` as the first item in the legend and `Blood`</span>
<span class="sd">            as the second item.</span>
<span class="sd">        colors: The colors for the groups in `by`. If not specified,</span>
<span class="sd">            the default colors will be used.</span>
<span class="sd">            Multiple colors can be separated by comma (`,`).</span>
<span class="sd">            You can specify `biopipen` to use the `biopipen` palette.</span>
<span class="sd">        ident: The column name of the cluster information.</span>
<span class="sd">        cluster_order (list): The order of the clusters.</span>
<span class="sd">            You may also use it to filter the clusters. If not given,</span>
<span class="sd">            all clusters will be used.</span>
<span class="sd">            If the cluster names are integers, use them directly for the order,</span>
<span class="sd">            even though a prefix `Cluster` is added on the plot.</span>
<span class="sd">        breaks (list;itype=int): breaks of the radar plots, from 0 to 100.</span>
<span class="sd">            If not given, the breaks will be calculated automatically.</span>
<span class="sd">        direction (choice): Direction to calculate the percentages.</span>
<span class="sd">            - inter-cluster: the percentage of the cells in all groups</span>
<span class="sd">                in each cluster (percentage adds up to 1 for each cluster).</span>
<span class="sd">            - intra-cluster: the percentage of the cells in all clusters.</span>
<span class="sd">                (percentage adds up to 1 for each group).</span>
<span class="sd">        section: If you want to put multiple cases into a same section</span>
<span class="sd">            in the report, you can set this option to the name of the section.</span>
<span class="sd">            Only used in the report.</span>
<span class="sd">        subset: The subset of the cells to do the analysis.</span>
<span class="sd">        bar_devpars (ns): The parameters for `png()` for the barplot</span>
<span class="sd">            - res (type=int): The resolution of the plot</span>
<span class="sd">            - height (type=int): The height of the plot</span>
<span class="sd">            - width (type=int): The width of the plot</span>
<span class="sd">        devpars (ns): The parameters for `png()`</span>
<span class="sd">            - res (type=int): The resolution of the plot</span>
<span class="sd">            - height (type=int): The height of the plot</span>
<span class="sd">            - width (type=int): The width of the plot</span>
<span class="sd">        cases (type=json): The cases for the multiple radar plots.</span>
<span class="sd">            Keys are the names of the cases and values are the arguments for</span>
<span class="sd">            the plots (`each`, `by`, `order`, `breaks`, `direction`,</span>
<span class="sd">            `ident`, `cluster_order` and `devpars`).</span>
<span class="sd">            If not cases are given, a default case will be used, with the</span>
<span class="sd">            key `DEFAULT`.</span>
<span class="sd">            The keys must be valid string as part of the file name.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.srtobj | stem}}.radar_plots&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/RadarPlots.R&quot;</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;each&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;prefix_each&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;order&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;colors&quot;</span><span class="p">:</span> <span class="s2">&quot;biopipen&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ident&quot;</span><span class="p">:</span> <span class="s2">&quot;seurat_clusters&quot;</span><span class="p">,</span>
        <span class="s2">&quot;cluster_order&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;breakdown&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;test&quot;</span><span class="p">:</span> <span class="s2">&quot;wilcox&quot;</span><span class="p">,</span>
        <span class="s2">&quot;breaks&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;direction&quot;</span><span class="p">:</span> <span class="s2">&quot;intra-cluster&quot;</span><span class="p">,</span>
        <span class="s2">&quot;section&quot;</span><span class="p">:</span> <span class="s2">&quot;DEFAULT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;bar_devpars&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span>
            <span class="s2">&quot;width&quot;</span><span class="p">:</span> <span class="mi">1200</span><span class="p">,</span>
            <span class="s2">&quot;height&quot;</span><span class="p">:</span> <span class="mi">800</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span>
            <span class="s2">&quot;width&quot;</span><span class="p">:</span> <span class="mi">1200</span><span class="p">,</span>
            <span class="s2">&quot;height&quot;</span><span class="p">:</span> <span class="mi">1000</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;cases&quot;</span><span class="p">:</span> <span class="p">{},</span>
    <span class="p">}</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/scrna/RadarPlots.svelte&quot;</span><span class="p">,</span>
    <span class="p">}</span>


<span class="nd">@mark</span><span class="p">(</span><span class="n">deprecated</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span id="biopipen.ns.scrna.MetaMarkers"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.MetaMarkers" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.MetaMarkers">DOCS</a>
<span class="k">class</span><span class="w"> </span><span class="nc">MetaMarkers</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find markers between three or more groups of cells, using one-way ANOVA</span>
<span class="sd">    or Kruskal-Wallis test.</span>

<span class="sd">    Sometimes, you may want to find the markers for cells from more than 2 groups.</span>
<span class="sd">    In this case, you can use this process to find the markers for the groups and</span>
<span class="sd">    do enrichment analysis for the markers. Each marker is examined using either</span>
<span class="sd">    one-way ANOVA or Kruskal-Wallis test.</span>
<span class="sd">    The p values are adjusted using the specified method. The significant markers</span>
<span class="sd">    are then used for enrichment analysis using</span>
<span class="sd">    [enrichr](https://maayanlab.cloud/Enrichr/) api.</span>

<span class="sd">    Other than the markers and the enrichment analysis as outputs, this process also</span>
<span class="sd">    generates violin plots for the top 10 markers.</span>

<span class="sd">    Input:</span>
<span class="sd">        srtobj: The seurat object loaded by `SeuratPreparing`</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory for the markers</span>

<span class="sd">    Envs:</span>
<span class="sd">        ncores (type=int): Number of cores to use to parallelize for genes</span>
<span class="sd">        mutaters (type=json): The mutaters to mutate the metadata</span>
<span class="sd">            The key-value pairs will be passed the `dplyr::mutate()` to mutate the metadata.</span>

<span class="sd">        group-by: The column name in metadata to group the cells.</span>
<span class="sd">            If only `group-by` is specified, and `idents` are</span>
<span class="sd">            not specified, markers will be found for all groups in this column.</span>
<span class="sd">            `NA` group will be ignored.</span>
<span class="sd">        idents: The groups of cells to compare, values should be in the `group-by` column.</span>
<span class="sd">        each: The column name in metadata to separate the cells into different cases.</span>
<span class="sd">        prefix_each (flag): Whether to add the `each` value as prefix to the case name.</span>
<span class="sd">        dbs (list): The dbs to do enrichment analysis for significant markers.</span>
<span class="sd">            You can use built-in dbs in `enrichit`, or provide your own gmt files.</span>
<span class="sd">            See also &lt;https://pwwang.github.io/enrichit/reference/FetchGMT.html&gt;.</span>
<span class="sd">            The built-in dbs include:</span>
<span class="sd">            * &quot;BioCarta&quot; or &quot;BioCarta_2016&quot;</span>
<span class="sd">            * &quot;GO_Biological_Process&quot; or &quot;GO_Biological_Process_2025&quot;</span>
<span class="sd">            * &quot;GO_Cellular_Component&quot; or &quot;GO_Cellular_Component_2025&quot;</span>
<span class="sd">            * &quot;GO_Molecular_Function&quot; or &quot;GO_Molecular_Function_2025&quot;</span>
<span class="sd">            * &quot;KEGG&quot;, &quot;KEGG_Human&quot;, &quot;KEGG_2021&quot;, or &quot;KEGG_2021_Human&quot;</span>
<span class="sd">            * &quot;Hallmark&quot;, &quot;MSigDB_Hallmark&quot;, or &quot;MSigDB_Hallmark_2020&quot;</span>
<span class="sd">            * &quot;Reactome&quot;, &quot;Reactome_Pathways&quot;, or &quot;Reactome_Pathways_2024&quot;</span>
<span class="sd">            * &quot;WikiPathways&quot;, &quot;WikiPathways_2024&quot;, &quot;WikiPathways_Human&quot;, or &quot;WikiPathways_2024_Human&quot;</span>
<span class="sd">            You can also fetch more dbs from &lt;https://maayanlab.cloud/Enrichr/#libraries&gt;.</span>
<span class="sd">        subset: The subset of the cells to do the analysis.</span>
<span class="sd">            An expression passed to `dplyr::filter()`.</span>
<span class="sd">        p_adjust (choice): The method to adjust the p values, which can be used to filter the significant markers.</span>
<span class="sd">            See also &lt;https://rdrr.io/r/stats/p.adjust.html&gt;</span>
<span class="sd">            - holm: Holm-Bonferroni method</span>
<span class="sd">            - hochberg: Hochberg method</span>
<span class="sd">            - hommel: Hommel method</span>
<span class="sd">            - bonferroni: Bonferroni method</span>
<span class="sd">            - BH: Benjamini-Hochberg method</span>
<span class="sd">            - BY: Benjamini-Yekutieli method</span>
<span class="sd">            - fdr: FDR method of Benjamini-Hochberg</span>
<span class="sd">            - none: No adjustment</span>
<span class="sd">        sigmarkers: An expression passed to `dplyr::filter()` to filter the</span>
<span class="sd">            significant markers for enrichment analysis. The default is `p.value &lt; 0.05`.</span>
<span class="sd">            If `method = &#39;anova&#39;`, the variables that can be used for filtering are:</span>
<span class="sd">            `sumsq`, `meansq`, `statistic`, `p.value` and `p_adjust`.</span>
<span class="sd">            If `method = &#39;kruskal&#39;`, the variables that can be used for filtering are:</span>
<span class="sd">            `statistic`, `p.value` and `p_adjust`.</span>
<span class="sd">        section: The section name for the report.</span>
<span class="sd">            Worked only when `each` is not specified.</span>
<span class="sd">            Otherwise, the section name will be constructed from `each` and `group-by`.</span>
<span class="sd">            If `DEFAULT`, and it&#39;s the only section, it not included in the case/section names.</span>
<span class="sd">        method (choice): The method for the test.</span>
<span class="sd">            - anova: One-way ANOVA</span>
<span class="sd">            - kruskal: Kruskal-Wallis test</span>
<span class="sd">        cases (type=json): If you have multiple cases, you can specify them</span>
<span class="sd">            here. The keys are the names of the cases and the values are the</span>
<span class="sd">            above options except `ncores` and `mutaters`. If some options are</span>
<span class="sd">            not specified, the default values specified above will be used.</span>
<span class="sd">            If no cases are specified, the default case will be added with</span>
<span class="sd">            the default values under `envs` with the name `DEFAULT`.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;srtobj:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.srtobj | stem}}.meta_markers&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/MetaMarkers.R&quot;</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;group-by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;idents&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;each&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;prefix_each&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;p_adjust&quot;</span><span class="p">:</span> <span class="s2">&quot;BH&quot;</span><span class="p">,</span>
        <span class="s2">&quot;dbs&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;KEGG_2021_Human&quot;</span><span class="p">,</span> <span class="s2">&quot;MSigDB_Hallmark_2020&quot;</span><span class="p">],</span>
        <span class="s2">&quot;sigmarkers&quot;</span><span class="p">:</span> <span class="s2">&quot;p_adjust &lt; 0.05&quot;</span><span class="p">,</span>
        <span class="s2">&quot;section&quot;</span><span class="p">:</span> <span class="s2">&quot;DEFAULT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;anova&quot;</span><span class="p">,</span>
        <span class="s2">&quot;cases&quot;</span><span class="p">:</span> <span class="p">{},</span>
    <span class="p">}</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/scrna/MetaMarkers.svelte&quot;</span><span class="p">,</span>
        <span class="s2">&quot;report_paging&quot;</span><span class="p">:</span> <span class="mi">8</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">Seurat2AnnData</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.Seurat2AnnData"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.Seurat2AnnData" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.Seurat2AnnData">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert seurat object to AnnData</span>

<span class="sd">    Input:</span>
<span class="sd">        sobjfile: The seurat object file, in RDS or qs/qs2 format</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The AnnData file</span>

<span class="sd">    Envs:</span>
<span class="sd">        assay: The assay to use for AnnData.</span>
<span class="sd">            If not specified, the default assay will be used.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;sobjfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.sobjfile | stem}}.h5ad&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/Seurat2AnnData.R&quot;</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;assay&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">AnnData2Seurat</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.AnnData2Seurat"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.AnnData2Seurat" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.AnnData2Seurat">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert AnnData to seurat object</span>

<span class="sd">    Input:</span>
<span class="sd">        adfile: The AnnData .h5ad file</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The seurat object file in RDS or qs/qs2 format</span>

<span class="sd">    Envs:</span>
<span class="sd">        assay: The assay to use to convert to seurat object.</span>
<span class="sd">        ident: The column name in `adfile.obs` to use as the identity</span>
<span class="sd">            for the seurat object.</span>
<span class="sd">            If not specified, no identity will be set.</span>
<span class="sd">        dotplot_check (type=auto): Whether to do a check with a dot plot.</span>
<span class="sd">            (`scplotter::FeatureStatPlot(plot_type = &quot;dot&quot;, ..)` will be used)</span>
<span class="sd">            to see if the conversion is successful.</span>
<span class="sd">            Set to `False` to disable the check.</span>
<span class="sd">            If `True`, top 10 variable genes will be used for the check.</span>
<span class="sd">            You can give a list of genes or a string of genes with comma (`,`) separated</span>
<span class="sd">            to use for the check.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;adfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.adfile | stem}}.qs&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;assay&quot;</span><span class="p">:</span> <span class="s2">&quot;RNA&quot;</span><span class="p">,</span> <span class="s2">&quot;ident&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;dotplot_check&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/AnnData2Seurat.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">ScSimulation</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.ScSimulation"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.ScSimulation" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.ScSimulation">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Simulate single-cell data using splatter.</span>

<span class="sd">    See &lt;https://www.bioconductor.org/packages/devel/bioc/vignettes/splatter/inst/doc/splatter.html#2_Quickstart&gt;</span>

<span class="sd">    Input:</span>
<span class="sd">        seed: The seed for the simulation</span>
<span class="sd">            You could also use string as the seed, and the seed will be</span>
<span class="sd">            generated by `digest::digest2int()`.</span>
<span class="sd">            So this could also work as a unique identifier for the simulation (ie. Sample ID).</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output Seurat object/SingleCellExperiment in qs/qs2 format</span>

<span class="sd">    Envs:</span>
<span class="sd">        ngenes (type=int): The number of genes to simulate</span>
<span class="sd">        ncells (type=int): The number of cells to simulate</span>
<span class="sd">        nspikes (type=int): The number of spike-ins to simulate</span>
<span class="sd">            When `ngenes`, `ncells`, and `nspikes` are not specified, the default</span>
<span class="sd">            params from `mockSCE()` will be used. By default, `ngenes = 2000`,</span>
<span class="sd">            `ncells = 200`, and `nspikes = 100`.</span>
<span class="sd">        outtype (choice): The output file type.</span>
<span class="sd">            - seurat: Seurat object</span>
<span class="sd">            - singlecellexperiment: SingleCellExperiment object</span>
<span class="sd">            - sce: alias for `singlecellexperiment`</span>
<span class="sd">        method (choice): which simulation method to use. Options are:</span>
<span class="sd">            - single: produces a single population</span>
<span class="sd">            - groups: produces distinct groups (eg. cell types), or</span>
<span class="sd">            - paths: selects cells from continuous trajectories (eg. differentiation processes)</span>
<span class="sd">        params (ns): Other parameters for simulation.</span>
<span class="sd">            The parameters are initialized `splitEstimate(mockSCE())` and then</span>
<span class="sd">            updated with the given parameters.</span>
<span class="sd">            See &lt;https://rdrr.io/bioc/splatter/man/SplatParams.html&gt;.</span>
<span class="sd">            Hyphens (`-`) will be transformed into dots (`.`) for the keys.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;seed:var&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:simulatied_{{in.seed}}.RDS&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;ngenes&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;ncells&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;nspikes&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;outtype&quot;</span><span class="p">:</span> <span class="s2">&quot;seurat&quot;</span><span class="p">,</span>
        <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;single&quot;</span><span class="p">,</span>
        <span class="s2">&quot;params&quot;</span><span class="p">:</span> <span class="p">{},</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/ScSimulation.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">CellCellCommunication</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.CellCellCommunication"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.CellCellCommunication" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.CellCellCommunication">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Cell-cell communication inference</span>

<span class="sd">    This is implemented based on [LIANA](https://liana-py.readthedocs.io/en/latest/index.html),</span>
<span class="sd">    which is a Python package for cell-cell communication inference and provides a list of existing</span>
<span class="sd">    methods including [CellPhoneDB](https://github.com/ventolab/CellphoneDB),</span>
<span class="sd">    [Connectome](https://github.com/msraredon/Connectome/), log2FC,</span>
<span class="sd">    [NATMI](https://github.com/forrest-lab/NATMI),</span>
<span class="sd">    [SingleCellSignalR](https://github.com/SCA-IRCM/SingleCellSignalR), Rank_Aggregate, Geometric Mean,</span>
<span class="sd">    [scSeqComm](https://gitlab.com/sysbiobig/scseqcomm), and [CellChat](https://github.com/jinworks/CellChat).</span>

<span class="sd">    You can also try `python -c &#39;import liana; liana.mt.show_methods()&#39;` to see the methods available.</span>

<span class="sd">    Note that this process does not do any visualization. You can use `CellCellCommunicationPlots`</span>
<span class="sd">    to visualize the results.</span>

<span class="sd">    Reference:</span>
<span class="sd">    - [Review](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9184522/).</span>
<span class="sd">    - [LIANA](https://www.biorxiv.org/content/10.1101/2023.08.19.553863v1).</span>

<span class="sd">    Input:</span>
<span class="sd">        sobjfile: The seurat object file in RDS or h5seurat format or AnnData file.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output file with the &#39;liana_res&#39; data frame.</span>
<span class="sd">            Stats are provided for both ligand and receptor entities, more specifically: ligand and receptor are</span>
<span class="sd">            the two entities that potentially interact. As a reminder, CCC events are not limited to secreted signalling,</span>
<span class="sd">            but we refer to them as ligand and receptor for simplicity.</span>
<span class="sd">            Also, in the case of heteromeric complexes, the ligand and receptor columns represent the subunit with minimum</span>
<span class="sd">            expression, while *_complex corresponds to the actual complex, with subunits being separated by _.</span>
<span class="sd">            source and target columns represent the source/sender and target/receiver cell identity for each interaction, respectively</span>
<span class="sd">            * `*_props`: represents the proportion of cells that express the entity.</span>
<span class="sd">                By default, any interactions in which either entity is not expressed in above 10%% of cells per cell type</span>
<span class="sd">                is considered as a false positive, under the assumption that since CCC occurs between cell types, a sufficient</span>
<span class="sd">                proportion of cells within should express the genes.</span>
<span class="sd">            * `*_means`: entity expression mean per cell type.</span>
<span class="sd">            * `lr_means`: mean ligand-receptor expression, as a measure of ligand-receptor interaction magnitude.</span>
<span class="sd">            * `cellphone_pvals`: permutation-based p-values, as a measure of interaction specificity.</span>

<span class="sd">            A typical output will look like this:</span>

<span class="sd">            | ligand | ligand_complex | ligand_props | ligand_trimean | mat_max | receptor | receptor_complex | receptor_props | receptor_trimean | source | target | lr_probs | cellchat_pvals | mag_score | spec_score |</span>
<span class="sd">            |--------|---------------|--------------|----------------|---------|----------|------------------|----------------|------------------|--------|--------|----------|----------------|-----------|------------|</span>
<span class="sd">            | VIM | VIM | 1.00 | 0.36 | 8.73 | CD44 | CD44 | 0.77 | 0.16 | c7 | c3 | 0.10 | 0.00 | 0.10 | 0.00 |</span>
<span class="sd">            | MIF | MIF | 0.97 | 0.22 | 8.73 | CXCR4 | CD74_CXCR4 | 0.87 | 0.26 | c5 | c6 | 0.10 | 0.00 | 0.10 | 0.00 |</span>
<span class="sd">            | HLA-B | HLA-B | 1.00 | 0.44 | 8.73 | KLRD1 | KLRD1 | 0.73 | 0.13 | c9 | c2 | 0.10 | 0.00 | 0.10 | 0.00 |</span>
<span class="sd">            | HMGB1 | HMGB1 | 0.99 | 0.26 | 8.73 | CXCR4 | CXCR4 | 0.81 | 0.21 | c2 | c7 | 0.10 | 0.00 | 0.10 | 0.00 |</span>
<span class="sd">            | CD48 | CD48 | 0.94 | 0.20 | 8.73 | CD2 | CD2 | 0.99 | 0.28 | c7 | c8 | 0.10 | 0.00 | 0.10 | 0.00 |</span>
<span class="sd">            | HLA-C | HLA-C | 1.00 | 0.38 | 8.73 | CD8B | CD8B | 0.73 | 0.15 | c1 | c9 | 0.10 | 0.00 | 0.10 | 0.00 |</span>
<span class="sd">            | LGALS1 | LGALS1 | 0.95 | 0.17 | 8.73 | CD69 | CD69 | 0.99 | 0.34 | c10 | c5 | 0.10 | 0.00 | 0.10 | 0.00 |</span>

<span class="sd">    Envs:</span>
<span class="sd">        method (choice): The method to use for cell-cell communication inference.</span>
<span class="sd">            - CellPhoneDB: Use CellPhoneDB method.</span>
<span class="sd">                Magnitude Score: lr_means; Specificity Score: cellphone_pvals.</span>
<span class="sd">            - Connectome: Use Connectome method.</span>
<span class="sd">            - log2FC: Use log2FC method.</span>
<span class="sd">            - NATMI: Use NATMI method.</span>
<span class="sd">            - SingleCellSignalR: Use SingleCellSignalR method.</span>
<span class="sd">            - Rank_Aggregate: Use Rank_Aggregate method.</span>
<span class="sd">            - Geometric_Mean: Use Geometric Mean method.</span>
<span class="sd">            - scSeqComm: Use scSeqComm method.</span>
<span class="sd">            - CellChat: Use CellChat method.</span>
<span class="sd">            - cellphonedb: alias for `CellPhoneDB`</span>
<span class="sd">            - connectome: alias for `Connectome`</span>
<span class="sd">            - log2fc: alias for `log2FC`</span>
<span class="sd">            - natmi: alias for `NATMI`</span>
<span class="sd">            - singlesignaler: alias for `SingleCellSignalR`</span>
<span class="sd">            - rank_aggregate: alias for `Rank_Aggregate`</span>
<span class="sd">            - geometric_mean: alias for `Geometric_Mean`</span>
<span class="sd">            - scseqcomm: alias for `scSeqComm`</span>
<span class="sd">            - cellchat: alias for `CellChat`</span>
<span class="sd">        subset: An expression in string to subset the cells.</span>
<span class="sd">            When a `.rds` or `.h5seurat` file is provided for `in.sobjfile`, you can provide an expression in `R`,</span>
<span class="sd">            which will be passed to `base::subset()` in `R` to subset the cells.</span>
<span class="sd">            But you can always pass an expression in `python` to subset the cells.</span>
<span class="sd">            See &lt;https://anndata.readthedocs.io/en/latest/tutorials/notebooks/getting-started.html#subsetting-using-metadata&gt;.</span>
<span class="sd">            You should use `adata` to refer to the AnnData object. For example, `adata.obs.groups == &quot;g1&quot;` will subset the cells</span>
<span class="sd">            with `groups` equal to `g1`.</span>
<span class="sd">        subset_using: The method to subset the cells.</span>
<span class="sd">            - auto: Automatically detect the method to use.</span>
<span class="sd">                Note that this is not always accurate. We simply check if `[` is in the expression.</span>
<span class="sd">                If so, we use `python` to subset the cells; otherwise, we use `R`.</span>
<span class="sd">            - python: Use python to subset the cells.</span>
<span class="sd">            - r: Use R to subset the cells.</span>
<span class="sd">        split_by: The column name in metadata to split the cells to run the method separately.</span>
<span class="sd">            The results will be combined together with this column in the final output.</span>
<span class="sd">        assay: The assay to use for the analysis.</span>
<span class="sd">            Only works for Seurat object.</span>
<span class="sd">        seed (type=int): The seed for the random number generator.</span>
<span class="sd">        ncores (type=int): The number of cores to use.</span>
<span class="sd">        groupby: The column name in metadata to group the cells.</span>
<span class="sd">            Typically, this column should be the cluster id.</span>
<span class="sd">            If provided input is a Seurat object, the default identity will be used by default.</span>
<span class="sd">            Otherwise, it is recommended to provide this parameter.</span>
<span class="sd">            &quot;seurat_clusters&quot; will be used with a warning if the input is in AnnData format and</span>
<span class="sd">            this parameter is not provided.</span>
<span class="sd">        group_by: alias for `groupby`</span>
<span class="sd">        species (choice): The species of the cells.</span>
<span class="sd">            - human: Human cells, the &#39;consensus&#39; resource will be used.</span>
<span class="sd">            - mouse: Mouse cells, the &#39;mouseconsensus&#39; resource will be used.</span>
<span class="sd">        expr_prop (type=float): Minimum expression proportion for the ligands and</span>
<span class="sd">            receptors (+ their subunits) in the corresponding cell identities. Set to 0</span>
<span class="sd">            to return unfiltered results.</span>
<span class="sd">        min_cells (type=int): Minimum cells (per cell identity if grouped by `groupby`)</span>
<span class="sd">            to be considered for downstream analysis.</span>
<span class="sd">        n_perms (type=int): Number of permutations for the permutation test.</span>
<span class="sd">            Relevant only for permutation-based methods (e.g., `CellPhoneDB`).</span>
<span class="sd">            If `0` is passed, no permutation testing is performed.</span>
<span class="sd">        rscript: The path to the Rscript executable used to convert RDS file to AnnData.</span>
<span class="sd">            if `in.sobjfile` is an RDS file, it will be converted to AnnData file (h5ad).</span>
<span class="sd">            You need `Seurat`, `SeuratDisk` and `digest` installed.</span>
<span class="sd">        &lt;more&gt;: Other arguments for the method.</span>
<span class="sd">            The arguments are passed to the method directly.</span>
<span class="sd">            See the method documentation for more details and also</span>
<span class="sd">            `help(liana.mt.&lt;method&gt;.__call__)` in Python.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;sobjfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.sobjfile | stem}}-ccc.txt&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;cellchat&quot;</span><span class="p">,</span>
        <span class="s2">&quot;assay&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;seed&quot;</span><span class="p">:</span> <span class="mi">1337</span><span class="p">,</span>
        <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;subset_using&quot;</span><span class="p">:</span> <span class="s2">&quot;auto&quot;</span><span class="p">,</span>
        <span class="s2">&quot;split_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;groupby&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;group_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="s2">&quot;human&quot;</span><span class="p">,</span>
        <span class="s2">&quot;expr_prop&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">,</span>
        <span class="s2">&quot;min_cells&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span>
        <span class="s2">&quot;n_perms&quot;</span><span class="p">:</span> <span class="mi">1000</span><span class="p">,</span>
        <span class="s2">&quot;rscript&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/CellCellCommunication.py&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">CellCellCommunicationPlots</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.CellCellCommunicationPlots"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.CellCellCommunicationPlots" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.CellCellCommunicationPlots">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Visualization for cell-cell communication inference.</span>

<span class="sd">    Examples:</span>
<span class="sd">        ### Network Plot</span>

<span class="sd">        ```toml</span>
<span class="sd">        [CellCellCommunicationPlots.envs.cases.&quot;Cell-Cell Communication Network&quot;]</span>
<span class="sd">        plot_type = &quot;network&quot;</span>
<span class="sd">        legend-position = &quot;none&quot;</span>
<span class="sd">        theme = &quot;theme_blank&quot;</span>
<span class="sd">        theme_args = {add_coord = false}</span>
<span class="sd">        ```</span>

<span class="sd">        ![Network Plot](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/cccplots/CellCellCommunicationPlots/sampleinfo.scRep-ccc_plots/Cell-Cell-Communication-Network.png){: width=&quot;80%&quot;}</span>

<span class="sd">        ### Circos Plot</span>

<span class="sd">        ![Circos Plot](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/cccplots/CellCellCommunicationPlots/sampleinfo.scRep-ccc_plots/Cell-Cell-Communication-Circos-Plot.png){: width=&quot;80%&quot;}</span>

<span class="sd">        ### Heatmap Plot</span>

<span class="sd">        ![Heatmap Plot](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/cccplots/CellCellCommunicationPlots/sampleinfo.scRep-ccc_plots/Cell-Cell-Communication-Heatmap.png){: width=&quot;80%&quot;}</span>

<span class="sd">        ### Cell-Cell Communication Interaction (Box Plot)</span>

<span class="sd">        ```toml</span>
<span class="sd">        [CellCellCommunicationPlots.envs.cases.&quot;Cell-Cell Communication Interaction (Box Plot)&quot;]</span>
<span class="sd">        plot_type = &quot;box&quot;</span>
<span class="sd">        x_text_angle = 90</span>
<span class="sd">        method = &quot;interaction&quot;</span>
<span class="sd">        ```</span>

<span class="sd">        ![Box Plot](https://raw.githubusercontent.com/pwwang/immunopipe/tests-output/cccplots/CellCellCommunicationPlots/sampleinfo.scRep-ccc_plots/Cell-Cell-Communication-Interaction-Box-Plot-.png){: width=&quot;80%&quot;}</span>

<span class="sd">    Input:</span>
<span class="sd">        cccfile: The output file from `CellCellCommunication`</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory for the plots.</span>

<span class="sd">    Envs:</span>
<span class="sd">        subset: An expression to pass to `dplyr::filter()` to subset the ccc data.</span>
<span class="sd">        magnitude: The column name in the data to use as the magnitude of the</span>
<span class="sd">            communication. By default, the second last column will be used.</span>
<span class="sd">            See `li.mt.show_methods()` for the available methods in LIANA. or</span>
<span class="sd">            &lt;https://liana-py.readthedocs.io/en/latest/notebooks/basic_usage.html#Tileplot&gt;</span>
<span class="sd">        specificity: The column name in the data to use as the specificity of the communication.</span>
<span class="sd">            By default, the last column will be used. If the method doesn&#39;t have a specificity, set it to None.</span>
<span class="sd">        devpars (ns): The parameters for the plot.</span>
<span class="sd">            - res (type=int): The resolution of the plot</span>
<span class="sd">            - height (type=int): The height of the plot</span>
<span class="sd">            - width (type=int): The width of the plot</span>
<span class="sd">        more_formats (type=list): The additional formats to save the plots.</span>
<span class="sd">        descr: The description of the plot.</span>
<span class="sd">        cases (type=json): The cases for the plots.</span>
<span class="sd">            The keys are the names of the cases and the values are the arguments for</span>
<span class="sd">            the plots. The arguments include the ones inherited from `envs`.</span>
<span class="sd">            You can have a special `plot_type` `&quot;table&quot;` to generate a table for the</span>
<span class="sd">            ccc data to save as a text file and show in the report.</span>
<span class="sd">            If no cases are given, a default case will be used, with the</span>
<span class="sd">            key `Cell-Cell Communication`.</span>
<span class="sd">        &lt;more&gt;: Other arguments passed to</span>
<span class="sd">            [scplotter::CCCPlot](https://pwwang.github.io/scplotter/reference/CCCPlot.html)</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;cccfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.cccfile | stem}}_plots&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;magnitude&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;specificity&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
        <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;descr&quot;</span><span class="p">:</span> <span class="s2">&quot;Cell-cell communication plot&quot;</span><span class="p">,</span>
        <span class="s2">&quot;cases&quot;</span><span class="p">:</span> <span class="p">{},</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/CellCellCommunicationPlots.R&quot;</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/common.svelte&quot;</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">ScVelo</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.ScVelo"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.ScVelo" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.ScVelo">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Velocity analysis for single-cell RNA-seq data</span>

<span class="sd">    This process is implemented based on the Python package `scvelo` (v0.3.3).</span>
<span class="sd">    Note that it doesn&#39;t work with `numpy&gt;=2`.</span>

<span class="sd">    Input:</span>
<span class="sd">        sobjfile: The seurat object file in RDS or h5seurat format or AnnData file.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output object with the velocity embeddings and information.</span>
<span class="sd">            In either RDS, h5seurat or h5ad format, depending on the `envs.outtype`.</span>
<span class="sd">            There will be also plots generated in the output directory</span>
<span class="sd">            (parent directory of `outfile`).</span>
<span class="sd">            Note that these plots will not be used in the report, but can be used as</span>
<span class="sd">            supplementary information for the velocity analysis.</span>
<span class="sd">            To visualize the velocity embeddings, you can use the `SeuratClusterStats`</span>
<span class="sd">            process with `v_reduction` provided to one of the `envs.dimplots`.</span>

<span class="sd">    Envs:</span>
<span class="sd">        ncores (type=int): Number of cores to use.</span>
<span class="sd">        group_by: The column name in metadata to group the cells.</span>
<span class="sd">            Typically, this column should be the cluster id.</span>
<span class="sd">            If provided input is a Seurat object, the default identity will be used by</span>
<span class="sd">            default. Otherwise, it is recommended to provide this parameter.</span>
<span class="sd">            &quot;seurat_clusters&quot; will be used with a warning if the input is in AnnData</span>
<span class="sd">            format and this parameter is not provided.</span>
<span class="sd">        mode (type=list): The mode to use for the velocity analysis.</span>
<span class="sd">            It should be a subset of `[&#39;deterministic&#39;, &#39;stochastic&#39;, &#39;dynamical&#39;]`,</span>
<span class="sd">            meaning that we can perform the velocity analysis in multiple modes.</span>
<span class="sd">        fitting_by (choice): The mode to use for fitting the velocities.</span>
<span class="sd">            - stochastic: Stochastic mode</span>
<span class="sd">            - deterministic: Deterministic mode</span>
<span class="sd">        min_shared_counts (type=int): Minimum number of counts</span>
<span class="sd">            (both unspliced and spliced) required for a gene.</span>
<span class="sd">        n_neighbors (type=int): The number of neighbors to use for the velocity graph.</span>
<span class="sd">        n_pcs (type=int): The number of PCs to use for the velocity graph.</span>
<span class="sd">        denoise (flag): Whether to denoise the data.</span>
<span class="sd">        denoise_topn (type=int): Number of genes with highest likelihood selected to</span>
<span class="sd">            infer velocity directions.</span>
<span class="sd">        kinetics (flag): Whether to compute the RNA velocity kinetics.</span>
<span class="sd">        kinetics_topn (type=int): Number of genes with highest likelihood selected to</span>
<span class="sd">            infer velocity directions.</span>
<span class="sd">        calculate_velocity_genes (flag): Whether to calculate the velocity genes.</span>
<span class="sd">        top_n (type=int): The number of top features to plot.</span>
<span class="sd">        rscript: The path to the Rscript executable used to convert RDS file to AnnData.</span>
<span class="sd">            if `in.sobjfile` is an RDS file, it will be converted to AnnData file</span>
<span class="sd">            (h5ad). You need `Seurat`, `SeuratDisk` and `digest` installed.</span>
<span class="sd">        outtype (choice): The output file type.</span>
<span class="sd">            - &lt;input&gt;: The same as the input file type.</span>
<span class="sd">            - h5seurat: h5seurat file</span>
<span class="sd">            - h5ad: h5ad file</span>
<span class="sd">            - qs: qs/qs2 file</span>
<span class="sd">            - qs2: qs2 file</span>
<span class="sd">            - rds: RDS file</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;sobjfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;outfile:file:{{in.sobjfile | stem}}-scvelo.&quot;</span>
        <span class="s2">&quot;{{ext0(in.sobjfile) if envs.outtype == &#39;&lt;input&gt;&#39; else envs.outtype}}&quot;</span>
    <span class="p">)</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;group_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;mode&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;deterministic&quot;</span><span class="p">,</span> <span class="s2">&quot;stochastic&quot;</span><span class="p">,</span> <span class="s2">&quot;dynamical&quot;</span><span class="p">],</span>
        <span class="s2">&quot;fitting_by&quot;</span><span class="p">:</span> <span class="s2">&quot;stochastic&quot;</span><span class="p">,</span>
        <span class="s2">&quot;min_shared_counts&quot;</span><span class="p">:</span> <span class="mi">30</span><span class="p">,</span>
        <span class="s2">&quot;n_neighbors&quot;</span><span class="p">:</span> <span class="mi">30</span><span class="p">,</span>
        <span class="s2">&quot;n_pcs&quot;</span><span class="p">:</span> <span class="mi">30</span><span class="p">,</span>
        <span class="s2">&quot;denoise&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;denoise_topn&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
        <span class="s2">&quot;kinetics&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;kinetics_topn&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span>
        <span class="s2">&quot;calculate_velocity_genes&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;top_n&quot;</span><span class="p">:</span> <span class="mi">6</span><span class="p">,</span>
        <span class="s2">&quot;rscript&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span><span class="p">,</span>
        <span class="s2">&quot;outtype&quot;</span><span class="p">:</span> <span class="s2">&quot;&lt;input&gt;&quot;</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/ScVelo.py&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">Slingshot</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.Slingshot"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.Slingshot" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.Slingshot">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Trajectory inference using Slingshot</span>

<span class="sd">    This process is implemented based on the R package `slingshot`.</span>

<span class="sd">    Input:</span>
<span class="sd">        sobjfile: The seurat object file in RDS or qs format.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output object with the trajectory information.</span>
<span class="sd">            The lineages are stored in the metadata of the seurat object at</span>
<span class="sd">            columns `LineageX`, where X is the lineage number. The `BranchID`</span>
<span class="sd">            column contains the branch id for each cell.</span>
<span class="sd">            One can use</span>
<span class="sd">            `scplotter::CellDimPlot(object, lineages = c(&quot;Lineage1&quot;, &quot;Lineage2&quot;, ...))`</span>
<span class="sd">            to visualize the trajectories.</span>

<span class="sd">    Envs:</span>
<span class="sd">        group_by: The column name in metadata to group the cells.</span>
<span class="sd">            Typically, this column should be the cluster id.</span>
<span class="sd">            Default is the default identity of the seurat object.</span>
<span class="sd">        reduction: The nonlinear reduction to use for the trajectory analysis.</span>
<span class="sd">        dims (type=auto): The dimensions to use for the analysis.</span>
<span class="sd">            A list or a string with comma separated values.</span>
<span class="sd">            Consecutive numbers can be specified with a colon (`:`) or a dash (`-`).</span>
<span class="sd">        start: The starting group for the Slingshot analysis.</span>
<span class="sd">        end: The ending group for the Slingshot analysis.</span>
<span class="sd">        prefix: The prefix to add to the column names of the resulting pseudotime variable.</span>
<span class="sd">        reverse (flag): Logical value indicating whether to reverse the pseudotime variable.</span>
<span class="sd">        align_start (flag): Whether to align the starting pseudotime values at the maximum pseudotime.</span>
<span class="sd">        seed (type=int): The seed for the random number generator.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;sobjfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.sobjfile | stem}}.qs&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;group_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;reduction&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;dims&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
        <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;prefix&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;reverse&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;align_start&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;seed&quot;</span><span class="p">:</span> <span class="mi">8525</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/Slingshot.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">LoomTo10X</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.LoomTo10X"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.LoomTo10X" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.LoomTo10X">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert Loom file to 10X format</span>

<span class="sd">    Input:</span>
<span class="sd">        loomfile: The Loom file</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory for the 10X format files,</span>
<span class="sd">            including the `matrix.mtx.gz`, `barcodes.tsv.gz` and `features.tsv.gz`</span>
<span class="sd">            files.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;loomfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.loomfile | stem}}.10X&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/LoomTo10X.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">PseudoBulkDEG</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.PseudoBulkDEG"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.PseudoBulkDEG" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.PseudoBulkDEG">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Pseduo-bulk differential gene expression analysis</span>

<span class="sd">    This process performs differential gene expression analysis, instead of</span>
<span class="sd">    on single-cell level, on the pseudo-bulk data, aggregated from the single-cell data.</span>

<span class="sd">    Input:</span>
<span class="sd">        sobjfile: The seurat object file in RDS or qs/qs2 format.</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output containing the results of the differential gene expression</span>
<span class="sd">            analysis.</span>

<span class="sd">    Envs:</span>
<span class="sd">        ncores (type=int): Number of cores to use for parallelization.</span>
<span class="sd">        mutaters (type=json): Mutaters to mutate the metadata of the</span>
<span class="sd">            seurat object. Keys are the new column names and values are the</span>
<span class="sd">            expressions to mutate the columns. These new columns can be</span>
<span class="sd">            used to define your cases.</span>
<span class="sd">            You can also use the clone selectors to select the TCR clones/clusters.</span>
<span class="sd">            See &lt;https://pwwang.github.io/scplotter/reference/clone_selectors.html&gt;.</span>
<span class="sd">        each: The column name in metadata to separate the cells into different cases.</span>
<span class="sd">            When specified, the case will be expanded to multiple cases for</span>
<span class="sd">            each value in the column.</span>
<span class="sd">        cache (type=auto): Where to cache the results.</span>
<span class="sd">            If `True`, cache to `outdir` of the job. If `False`, don&#39;t cache.</span>
<span class="sd">            Otherwise, specify the directory to cache to.</span>
<span class="sd">        subset: An expression in string to subset the cells.</span>
<span class="sd">        aggregate_by: The column names in metadata to aggregate the cells.</span>
<span class="sd">        layer: The layer to pull and aggregate the data.</span>
<span class="sd">        assay: The assay to pull and aggregate the data.</span>
<span class="sd">        error (flag): Error out if no/not enough markers are found or no pathways are enriched.</span>
<span class="sd">            If `False`, empty results will be returned.</span>
<span class="sd">        group_by: The column name in metadata to group the cells.</span>
<span class="sd">        ident_1: The first identity to compare.</span>
<span class="sd">        ident_2: The second identity to compare.</span>
<span class="sd">            If not specified, the rest of the identities will be compared with `ident_1`.</span>
<span class="sd">        paired_by: The column name in metadata to mark the paired samples.</span>
<span class="sd">            For example, subject. If specified, the paired test will be performed.</span>
<span class="sd">        dbs (list): The databases to use for enrichment analysis.</span>
<span class="sd">            You can use built-in dbs in `enrichit`, or provide your own gmt files.</span>
<span class="sd">            See also &lt;https://pwwang.github.io/enrichit/reference/FetchGMT.html&gt;.</span>
<span class="sd">            The built-in dbs include:</span>
<span class="sd">            * &quot;BioCarta&quot; or &quot;BioCarta_2016&quot;</span>
<span class="sd">            * &quot;GO_Biological_Process&quot; or &quot;GO_Biological_Process_2025&quot;</span>
<span class="sd">            * &quot;GO_Cellular_Component&quot; or &quot;GO_Cellular_Component_2025&quot;</span>
<span class="sd">            * &quot;GO_Molecular_Function&quot; or &quot;GO_Molecular_Function_2025&quot;</span>
<span class="sd">            * &quot;KEGG&quot;, &quot;KEGG_Human&quot;, &quot;KEGG_2021&quot;, or &quot;KEGG_2021_Human&quot;</span>
<span class="sd">            * &quot;Hallmark&quot;, &quot;MSigDB_Hallmark&quot;, or &quot;MSigDB_Hallmark_2020&quot;</span>
<span class="sd">            * &quot;Reactome&quot;, &quot;Reactome_Pathways&quot;, or &quot;Reactome_Pathways_2024&quot;</span>
<span class="sd">            * &quot;WikiPathways&quot;, &quot;WikiPathways_2024&quot;, &quot;WikiPathways_Human&quot;, or &quot;WikiPathways_2024_Human&quot;</span>
<span class="sd">            You can also fetch more dbs from &lt;https://maayanlab.cloud/Enrichr/#libraries&gt;.</span>
<span class="sd">        sigmarkers: An expression passed to `dplyr::filter()` to filter the</span>
<span class="sd">            significant markers for enrichment analysis.</span>
<span class="sd">            The default is `p_val_adj &lt; 0.05`.</span>
<span class="sd">            If `tool = &#39;DESeq2&#39;`, the variables that can be used for filtering</span>
<span class="sd">            are: `baseMean`, `log2FC`, `lfcSE`, `stat`, `p_val`, `p_val_adj`.</span>
<span class="sd">            If `tool = &#39;edgeR&#39;`, the variables that can be used for filtering</span>
<span class="sd">            are: `logCPM`, `log2FC`, `LR`, `p_val`, `p_val_adj`.</span>
<span class="sd">        enrich_style (choice): The style of the enrichment analysis.</span>
<span class="sd">            - enrichr: Use `enrichr`-style for the enrichment analysis.</span>
<span class="sd">            - clusterProfiler: Use `clusterProfiler`-style for the enrichment analysis.</span>
<span class="sd">        allmarker_plots_defaults (ns): Default options for the plots for all markers when `ident_1` is not specified.</span>
<span class="sd">            - plot_type: The type of the plot.</span>
<span class="sd">                See &lt;https://pwwang.github.io/scplotter/reference/FeatureStatPlot.html&gt;.</span>
<span class="sd">                Available types are `violin`, `box`, `bar`, `ridge`, `dim`, `heatmap` and `dot`.</span>
<span class="sd">            - more_formats (type=list): The extra formats to save the plot in.</span>
<span class="sd">            - save_code (flag): Whether to save the code to generate the plot.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - order_by: an expression to order the markers, passed by `dplyr::arrange()`.</span>
<span class="sd">            - genes: The number of top genes to show or an expression passed to `dplyr::filter()` to filter the genes.</span>
<span class="sd">            - &lt;more&gt;: Other arguments passed to [`scplotter::FeatureStatPlot()`](https://pwwang.github.io/scplotter/reference/FeatureStatPlot.html).</span>
<span class="sd">        allmarker_plots (type=json): All marker plot cases.</span>
<span class="sd">            The keys are the names of the cases and the values are the dicts inherited from `allmarker_plots_defaults`.</span>
<span class="sd">        allenrich_plots_defaults (ns): Default options for the plots to generate for the enrichment analysis.</span>
<span class="sd">            - plot_type: The type of the plot.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://pwwang.github.io/scplotter/reference/EnrichmentPlot.html&gt;.</span>
<span class="sd">        allenrich_plots (type=json): Cases of the plots to generate for the enrichment analysis.</span>
<span class="sd">            The keys are the names of the cases and the values are the dicts inherited from `allenrich_plots_defaults`.</span>
<span class="sd">            The cases under `envs.cases` can inherit this options.</span>
<span class="sd">        marker_plots_defaults (ns): Default options for the plots to generate for the markers.</span>
<span class="sd">            - plot_type: The type of the plot.</span>
<span class="sd">                See &lt;https://pwwang.github.io/scplotter/reference/FeatureStatPlot.html&gt;.</span>
<span class="sd">                Available types are `violin`, `box`, `bar`, `ridge`, `dim`, `heatmap` and `dot`.</span>
<span class="sd">                There are two additional types available - `volcano_pct` and `volcano_log2fc`.</span>
<span class="sd">            - more_formats (type=list): The extra formats to save the plot in.</span>
<span class="sd">            - save_code (flag): Whether to save the code to generate the plot.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - order_by: an expression to order the markers, passed by `dplyr::arrange()`.</span>
<span class="sd">            - genes: The number of top genes to show or an expression passed to `dplyr::filter()` to filter the genes.</span>
<span class="sd">            - &lt;more&gt;: Other arguments passed to [`scplotter::FeatureStatPlot()`](https://pwwang.github.io/scplotter/reference/FeatureStatPlot.html).</span>
<span class="sd">                If `plot_type` is `volcano_pct` or `volcano_log2fc`, they will be passed to</span>
<span class="sd">                [`scplotter::VolcanoPlot()`](https://pwwang.github.io/plotthis/reference/VolcanoPlot.html).</span>
<span class="sd">        marker_plots (type=json): Cases of the plots to generate for the markers.</span>
<span class="sd">            Plot cases. The keys are the names of the cases and the values are the dicts inherited from `marker_plots_defaults`.</span>
<span class="sd">            The cases under `envs.cases` can inherit this options.</span>
<span class="sd">        enrich_plots_defaults (ns): Default options for the plots to generate for the enrichment analysis.</span>
<span class="sd">            - plot_type: The type of the plot.</span>
<span class="sd">                See &lt;https://pwwang.github.io/scplotter/reference/EnrichmentPlot.html&gt;.</span>
<span class="sd">                Available types are `bar`, `dot`, `lollipop`, `network`, `enrichmap` and `wordcloud`.</span>
<span class="sd">            - more_formats (type=list): The extra formats to save the plot in.</span>
<span class="sd">            - save_code (flag): Whether to save the code to generate the plot.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - &lt;more&gt;: See &lt;https://pwwang.github.io/scplotter/reference/EnrichmentPlot.htmll&gt;.</span>
<span class="sd">        enrich_plots (type=json): Cases of the plots to generate for the enrichment analysis.</span>
<span class="sd">            The keys are the names of the cases and the values are the dicts inherited from `enrich_plots_defaults`.</span>
<span class="sd">            The cases under `envs.cases` can inherit this options.</span>
<span class="sd">        overlaps_defaults (ns): Default options for investigating the overlapping of significant markers between different cases or comparisons.</span>
<span class="sd">            This means either `ident_1` should be empty, so that they can be expanded to multiple comparisons.</span>
<span class="sd">            - sigmarkers: The expression to filter the significant markers for each case.</span>
<span class="sd">                If not provided, `envs.sigmarkers` will be used.</span>
<span class="sd">            - plot_type (choice): The type of the plot to generate for the overlaps.</span>
<span class="sd">                - venn: Use `plotthis::VennDiagram()`.</span>
<span class="sd">                - upset: Use `plotthis::UpsetPlot()`.</span>
<span class="sd">            - more_formats (type=list): The extra formats to save the plot in.</span>
<span class="sd">            - save_code (flag): Whether to save the code to generate the plot.</span>
<span class="sd">            - devpars (ns): The device parameters for the plots.</span>
<span class="sd">                - res (type=int): The resolution of the plots.</span>
<span class="sd">                - height (type=int): The height of the plots.</span>
<span class="sd">                - width (type=int): The width of the plots.</span>
<span class="sd">            - &lt;more&gt;: More arguments pased to `plotthis::VennDiagram()`</span>
<span class="sd">                (&lt;https://pwwang.github.io/plotthis/reference/venndiagram1.html&gt;)</span>
<span class="sd">                or `plotthis::UpsetPlot()`</span>
<span class="sd">                (&lt;https://pwwang.github.io/plotthis/reference/upsetplot1.html&gt;)</span>
<span class="sd">        overlaps (type=json): Cases for investigating the overlapping of significant markers between different cases or comparisons.</span>
<span class="sd">            The keys are the names of the cases and the values are the dicts inherited from `overlaps_defaults`.</span>
<span class="sd">            There are two situations that we can perform overlaps:</span>
<span class="sd">            1. If `ident_1` is not specified, the overlaps can be performed between different comparisons.</span>
<span class="sd">            2. If `each` is specified, the overlaps can be performed between different cases, where in each case, `ident_1` must be specified.</span>
<span class="sd">        tool (choice): The method to use for the differential expression analysis.</span>
<span class="sd">            - DESeq2: Use DESeq2 for the analysis.</span>
<span class="sd">            - edgeR: Use edgeR for the analysis.</span>
<span class="sd">        plots_defaults (ns): The default parameters for the plots.</span>
<span class="sd">            - &lt;more&gt;: Parameters passed to `biopipen.utils::VizBulkDEGs()`.</span>
<span class="sd">                See: &lt;https://pwwang.github.io/biopipen.utils.R/reference/VizBulkDEGs.html&gt;</span>
<span class="sd">        plots (type=json): The parameters for the plots.</span>
<span class="sd">            The keys are the names of the plots and the values are the parameters</span>
<span class="sd">            for the plots. The parameters will override the defaults in `plots_defaults`.</span>
<span class="sd">            If not specified, no plots will be generated.</span>
<span class="sd">        cases (type=json): The cases for the analysis.</span>
<span class="sd">            The keys are the names of the cases and the values are the arguments for</span>
<span class="sd">            the analysis. The arguments include the ones inherited from `envs`.</span>
<span class="sd">            If no cases are specified, a default case will be added with</span>
<span class="sd">            the name `DEG Analysis` and the default values specified above.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;sobjfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.sobjfile | stem}}.pseudobulk_deg&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/PseudoBulkDEG.R&quot;</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;mutaters&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;cache&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">tmpdir</span><span class="p">,</span>
        <span class="s2">&quot;each&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;aggregate_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;layer&quot;</span><span class="p">:</span> <span class="s2">&quot;counts&quot;</span><span class="p">,</span>
        <span class="s2">&quot;assay&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;error&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;group_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;ident_1&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;ident_2&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;paired_by&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;tool&quot;</span><span class="p">:</span> <span class="s2">&quot;DESeq2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;dbs&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;KEGG_2021_Human&quot;</span><span class="p">,</span> <span class="s2">&quot;MSigDB_Hallmark_2020&quot;</span><span class="p">],</span>
        <span class="s2">&quot;sigmarkers&quot;</span><span class="p">:</span> <span class="s2">&quot;p_val_adj &lt; 0.05&quot;</span><span class="p">,</span>
        <span class="s2">&quot;enrich_style&quot;</span><span class="p">:</span> <span class="s2">&quot;enrichr&quot;</span><span class="p">,</span>
        <span class="s2">&quot;allmarker_plots_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;save_code&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
            <span class="s2">&quot;order_by&quot;</span><span class="p">:</span> <span class="s2">&quot;desc(abs(log2FC))&quot;</span><span class="p">,</span>
            <span class="s2">&quot;genes&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;allmarker_plots&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;allenrich_plots_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;heatmap&quot;</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;allenrich_plots&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;marker_plots_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;save_code&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
            <span class="s2">&quot;order_by&quot;</span><span class="p">:</span> <span class="s2">&quot;desc(abs(log2FC))&quot;</span><span class="p">,</span>
            <span class="s2">&quot;genes&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;marker_plots&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Volcano Plot&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;volcano&quot;</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;enrich_plots_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;save_code&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;enrich_plots&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Bar Plot&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;bar&quot;</span><span class="p">,</span> <span class="s2">&quot;ncol&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;top_term&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;overlaps_defaults&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;sigmarkers&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s2">&quot;plot_type&quot;</span><span class="p">:</span> <span class="s2">&quot;venn&quot;</span><span class="p">,</span>
            <span class="s2">&quot;more_formats&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;save_code&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;devpars&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;res&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="s2">&quot;overlaps&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;cases&quot;</span><span class="p">:</span> <span class="p">{},</span>
    <span class="p">}</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/common.svelte&quot;</span><span class="p">,</span>
        <span class="s2">&quot;report_paging&quot;</span><span class="p">:</span> <span class="mi">8</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">CellSNPLite</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.CellSNPLite"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.CellSNPLite" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.CellSNPLite">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Genotyping bi-allelic SNPs on single cells using cellsnp-lite.</span>

<span class="sd">    The output from cellsnp-lite can be directly used for downstream analysis such as -</span>

<span class="sd">    * Donor deconvolution in multiplexed single-cell RNA-seq data (e.g., with vireo).</span>
<span class="sd">    * Allele-specific CNV analysis in single-cell or spatial transcriptomics data (e.g., with Numbat, XClone, or CalicoST).</span>
<span class="sd">    * Clonal substructure discovery using single cell mitochondrial variants (e.g., with MQuad).</span>

<span class="sd">    Here we only support model `1a`/`2a` in cellsnp-lite, which is designed for a single bam file as input.</span>
<span class="sd">    For model `1b`/`2b`, which is designed for multiple bam files as input (e.g., one per cell), you can still</span>
<span class="sd">    run with this process, but only one bam file is allowed.</span>

<span class="sd">    See &lt;https://github.com/single-cell-genetics/cellsnp-lite&gt; for more details about cellsnp-lite.</span>

<span class="sd">    Input:</span>
<span class="sd">        crdir: The cellranger output directory or the directory containing</span>
<span class="sd">            the bam file and barcode file.</span>
<span class="sd">            It should contain the `outs/possorted_genome_bam.bam` file and</span>
<span class="sd">            the `outs/filtered_feature_bc_matrix/barcodes.tsv.gz` file.</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory for cellsnp-lite results.</span>

<span class="sd">    Envs:</span>
<span class="sd">        ncores (type=int): The number of cores to use.</span>
<span class="sd">            Will pass to `-p` option in cellsnp-lite.</span>
<span class="sd">        regionsVCF: A vcf file listing all candidate SNPs, for fetch each variants.</span>
<span class="sd">        genotype (flag): Whether to perform genotyping.</span>
<span class="sd">            If `False`, only the allele counts will be computed.</span>
<span class="sd">        gzip (flag): Whether to gzip the output files.</span>
<span class="sd">        &lt;more&gt;: Other arguments passed to cellsnp-lite.</span>
<span class="sd">            See &lt;https://cellsnp-lite.readthedocs.io/en/latest/main/manual.html#full-parameters&gt; for more details.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>

    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;crdir:dir&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        outdir:dir:</span>
<span class="s2">        {</span><span class="si">%- i</span><span class="s2">f basename(in.crdir) == &#39;outs&#39; -%}</span>
<span class="s2">            {{in.crdir | dirname | basename}}</span>
<span class="s2">        {</span><span class="si">%- e</span><span class="s2">lse -%}</span>
<span class="s2">            {{in.crdir | basename}}</span>
<span class="s2">        {</span><span class="si">%- e</span><span class="s2">ndif -%}</span>
<span class="s2">        .cellsnp</span>
<span class="s2">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;cellsnp_lite&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">exe</span><span class="o">.</span><span class="n">cellsnp_lite</span><span class="p">,</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;regionsVCF&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;genotype&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;gzip&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/CellSNPLite.py&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">MQuad</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.MQuad"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.MQuad" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.MQuad">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Clonal substructure discovery using single cell mitochondrial variants with MQuad.</span>

<span class="sd">    MQuad uses a Mixture Model for Mitochondrial Mutation detection in single-cell omics data.</span>

<span class="sd">    MQuad is a tool that detects mitochondrial mutations that are informative for clonal substructure inference. It uses a binomial mixture model to assess the heteroplasmy of mtDNA variants among background noise.</span>

<span class="sd">    Input:</span>
<span class="sd">        cellsnpout: The output directory from `CellSNPLite` process, which should contain</span>
<span class="sd">            AD and DP sparse matrices (.mtx) or the vcf file.</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory for MQuad results.</span>

<span class="sd">    Envs:</span>
<span class="sd">        ncores (type=int): The number of cores to use.</span>
<span class="sd">            It will be passed to `--nproc` option in MQuad.</span>
<span class="sd">        seed (type=int): The seed for the random number generator.</span>
<span class="sd">            It will be passed to `--randSeed` option in MQuad.</span>
<span class="sd">        &lt;more&gt;: Other arguments passed to MQuad.</span>
<span class="sd">            See &lt;https://github.com/single-cell-genetics/MQuad/blob/main/mquad/mquad_CLI.py&gt; for more details.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;cellsnpout:dir&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.cellsnpout | stem}}.mquad&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;mquad&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">exe</span><span class="o">.</span><span class="n">mquad</span><span class="p">,</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;seed&quot;</span><span class="p">:</span> <span class="mi">8525</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/MQuad.py&quot;</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/scrna/MQuad.svelte&quot;</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">MQuadMerge</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.MQuadMerge"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.MQuadMerge" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.MQuadMerge">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Merge multiple MQuad results for multiple samples.</span>

<span class="sd">    We will merge the passed_ad.mtx, passed_dp.mtx and passed_variant_names.txt files</span>
<span class="sd">    from multiple samples</span>

<span class="sd">    Input:</span>
<span class="sd">        mquaddirs: The output directories from `MQuad` process for multiple samples.</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory for merged MQuad results.</span>
<span class="sd">            This can be later used as input to `VireoSNP` process.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;mquaddirs:dirs&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.mquaddirs | first | stem | append: &#39;-etc&#39;}}.mquadmerged&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/MQuadMerge.py&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">VireoSNP</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.scrna.VireoSNP"></span><a class="mkapi-docs-link" title="biopipen.ns.scrna.VireoSNP" href="../../biopipen.ns.scrna/#biopipen.ns.scrna.VireoSNP">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Demultiplexing of single-cell RNA-seq data using vireoSNP.</span>

<span class="sd">    VireoSNP is a Bayesian method for demultiplexing pooled single-cell RNA-seq data</span>
<span class="sd">    using natural genetic variations (SNPs) without requiring genotype reference.</span>

<span class="sd">    Refers to &lt;https://github.com/single-cell-genetics/vireo/blob/master/examples/vireoSNP_clones.ipynb&gt; for more details.</span>

<span class="sd">    Input:</span>
<span class="sd">        cellsnpout: The output directory from `CellSNPLite` process, which should</span>
<span class="sd">            contain AD and DP sparse matrices (.mtx) or the vcf file.</span>
<span class="sd">            To investigate the clonal substructure using mitochondrial variants, run</span>
<span class="sd">            `MQuad` first to select the informative variants, and then use the</span>
<span class="sd">            filtered vcf file from MQuad as input to this process.</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The output directory for vireoSNP results.</span>

<span class="sd">    Envs:</span>
<span class="sd">        seed (type=int): The seed for the random number generator.</span>
<span class="sd">        ncores (type=int): The number of cores to use for model fitting for</span>
<span class="sd">            different number of clones.</span>
<span class="sd">        n_init (type=int): The number of random initializations to perform.</span>
<span class="sd">        n_clones (type=auto): The number of clones in the pooled single-cell RNA-seq data.</span>
<span class="sd">            (Refered as `n_donor` in vireoSNP documentation.)</span>
<span class="sd">            If a 2-element list of provided, we will try to estimate the best number of clones</span>
<span class="sd">            between the two values using the elbow plot method.</span>
<span class="sd">        min_iter (type=int): The minimum number of iterations to perform.</span>
<span class="sd">        max_iter (type=int): The maximum number of iterations to perform.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;cellsnpout:dir&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:{{in.cellsnpout | stem}}.vireoSNP&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;seed&quot;</span><span class="p">:</span> <span class="mi">8525</span><span class="p">,</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;n_init&quot;</span><span class="p">:</span> <span class="mi">50</span><span class="p">,</span>
        <span class="s2">&quot;n_clones&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">10</span><span class="p">],</span>
        <span class="s2">&quot;min_iter&quot;</span><span class="p">:</span> <span class="mi">30</span><span class="p">,</span>
        <span class="s2">&quot;max_iter&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/scrna/VireoSNP.py&quot;</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/scrna/VireoSNP.svelte&quot;</span><span class="p">,</span>
    <span class="p">}</span>
</code></pre></div>
</div>
</div>
              
            </div>
          </div><footer>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  <div class="footer-notes">
    <a href="https://github.com/pwwang/biopipen" title="Open repository in GitHub">Github Repo</a>
    <div class="footer-sep"></div>Built with <a href="https://www.mkdocs.org/">MkDocs</a> using theme <a href="https://github.com/pwwang/mkdocs-rtd">mkdocs-rtd</a>.
  </div>
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
  </span>
</div>
    <script src="../../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../../..";</script>
    <script src="../../../js/theme_extra.js"></script>
    <script src="../../../js/theme.js"></script>
    <script src="../../../js/version-select.js"></script>
      <script src="../../../js/mkapi.js"></script>
      <script src="../../../extra.js"></script>
      <script src="../../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
