"""
Procs for plink 1.x
"""
from pyppl import Proc, Box
from . import params, rimport, bashimport

"""
@name:
	pPlinkMiss
@description:
	Find samples and snps with missing calls, calculate the call rates and plot them.
@input:
	`indir:dir`: The input directory containing .bed/.bim/.fam files
@output:
	`outdir:dir`: The output directory. Default: `{{i.indir | fn}}.miss`
		- `.imiss`: The miss calls for samples
		- `.lmiss`: The miss calls for snps
		- `.samplecr.fail`: The samples fail sample call rate cutoff (`args.samplecr`)
		- `.snpcr.fail`: The SNPs fail snp call rate cutoff (`args.snpcr`)
@args:
	`plink`: The path to plink.
	`samplecr`: The sample call rate cutoff. Default: `.95`
	`snpcr`: The SNP call rate cutoff. Default: `.95`
	`plot`: Whether plot the distribution of the call rates? Default: `True`
	`devpars`: The device parameters for the plot. Default: `Box(res=300, width=2000, height=2000)`
"""
pPlinkMiss               = Proc(desc = 'Find samples and snps with missing calls')
pPlinkMiss.input         = 'indir:dir'
pPlinkMiss.output        = 'outdir:dir:{{i.indir | fn}}.miss'
pPlinkMiss.args.plink    = params.plink.value
pPlinkMiss.args.samplecr = .95
pPlinkMiss.args.snpcr    = .95
pPlinkMiss.args.plot     = True
pPlinkMiss.args.devpars  = Box(res=300, width=2000, height=2000)
pPlinkMiss.envs.rimport  = rimport
pPlinkMiss.lang          = params.Rscript.value
pPlinkMiss.script        = "file:scripts/plink/pPlinkMiss.r"

"""
@name:
	pPlinkHet
@description:
	Calculate the heterozygosity of each sample.
@input:
	`indir:dir`: The input directory containing .bed/.bim/.fam files
@output:
	`outdir:dir`: The output directory. Default: `{{i.indir | fn}}.het`
		- `.het`: The heterozygosity file generated by `plink`.
		- `.het.fail`: The samples fail sample heterozygosity cutoff (`args.cutoff`)
@args:
	`plink`: The path to plink.
	`cutoff`: The sample heterozygosity cutoff. Default: `3` (mean-3SD ~ mean+3SD)
	`plot`: Whether plot the distribution of the heterozygosity? Default: `True`
	`devpars`: The device parameters for the plot. Default: `Box(res=300, width=2000, height=2000)`
"""
pPlinkHet              = Proc(desc = 'Calculate the heterozygosity of each sample')
pPlinkHet.input        = 'indir:dir'
pPlinkHet.output       = 'outdir:dir:{{i.indir | fn}}.het'
pPlinkHet.args.plink   = params.plink.value
pPlinkHet.args.cutoff  = 3
pPlinkHet.args.plot    = True
pPlinkHet.args.devpars = Box(res=300, width=2000, height=2000)
pPlinkHet.envs.rimport = rimport
pPlinkHet.lang         = params.Rscript.value
pPlinkHet.script       = "file:scripts/plink/pPlinkHet.r"

"""
@name:
	pPlinkHWE
@description:
	Hardy-Weinberg Equilibrium report and filtering.
@input:
	`indir:dir`: The input directory containing .bed/.bim/.fam files
@output:
	`outdir:dir`: The output directory. Default: `{{i.indir | fn}}.hwe`
		- `.hwe`: The HWE report by `plink`
		- `.hardy.fail`: The SNPs fail HWE test
@args:
	`plink`: The path to plink.
	`cutoff`: The HWE p-value cutoff. Default: `1e-5`
	`plot`: Whether plot the distribution of the HWE p-values? Default: `True`
	`devpars`: The device parameters for the plot. Default: `Box(res=300, width=2000, height=2000)`
"""
pPlinkHWE = Proc(desc = "Hardy-Weinberg Equilibrium report and filtering.")
pPlinkHWE.input        = 'indir:dir'
pPlinkHWE.output       = 'outdir:dir:{{i.indir | fn}}.hwe'
pPlinkHWE.args.plink   = params.plink.value
pPlinkHWE.args.cutoff  = 1e-5
pPlinkHWE.args.plot    = True
pPlinkHWE.args.devpars = Box(res=300, width=2000, height=2000)
pPlinkHWE.envs.rimport = rimport
pPlinkHWE.lang         = params.Rscript.value
pPlinkHWE.script       = "file:scripts/plink/pPlinkHWE.r"

"""
@name:
	pPlinkIBD
@description:
	Estimate the degree of recent shared ancestry individual pairs,
	the identity by descent (IBD)
@input:
	`indir:dir`: The input directory containing .bed/.bim/.fam files
@output:
	`outdir:dir`: The output directory. Default: `{{i.indir | fn}}.ibd`
		- `.genome`: The original genome report from `plink`
		- `.ibd.png`: The heatmap of PI_HAT
@args:
	`plink`: The path to plink.
	`indep`: To give a concrete example: the command above that specifies 50 5 0.2 would a) consider a window of 50 SNPs, b) calculate LD between each pair of SNPs in the window, b) remove one of a pair of SNPs if the LD is greater than 0.5, c) shift the window 5 SNPs forward and repeat the procedure.
	`pihat`: The PI_HAT cutoff. Default: 0.1875 (see: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5007749/)
	`plot` : Whether plot the PI_HAT heatmap? Default: `True`
	`devpars`: The device parameters for the plot. Default: `Box(res=300, width=2200, height=1600)`
"""
pPlinkIBD              = Proc(desc = "Estimate the identity by descent (IBD)")
pPlinkIBD.input        = 'indir:dir'
pPlinkIBD.output       = 'outdir:dir:{{i.indir | fn}}.ibd'
pPlinkIBD.args.plink   = params.plink.value
pPlinkIBD.args.highld  = params.highld.value
pPlinkIBD.args.indep   = [50, 5, .2]
pPlinkIBD.args.pihat   = 0.1875 # ref: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5007749/
pPlinkIBD.args.plot    = True
pPlinkIBD.args.devpars = Box(res=300, width=2200, height=1600)
pPlinkIBD.envs.rimport = rimport
pPlinkIBD.lang         = params.Rscript.value
pPlinkIBD.script       = "file:scripts/plink/pPlinkIBD.r"

"""
@name:
	pPlinkRemove
@description:
	Remove failed samples and/or SNPs
	The samples/SNPs to be removed should be generated by one of:
	`pPlinkHet`, `pPlinkHWE`, `pPlinkIBD` or `pPlinkMiss`
@input:
	`indir:dir`: The input directory containing .bed/.bim/.fam files
	`pdir:dir` : The output directory from one of the processes listed in description
		- It could also be the `.fail` file generated by those processes
@output:
	`outdir:dir`: The output directory containing the `.bed/.bim/.fam` after filtering.
@args:
	`plink`: The path to plink.
"""
pPlinkRemove              = Proc(desc = "Remove failed samples and SNPs")
pPlinkRemove.input        = 'indir:dir, pdir:dir'
pPlinkRemove.output       = 'outdir:dir:{{i.indir | fn}}'
pPlinkRemove.args.plink   = params.plink.value
pPlinkRemove.envs.rimport = rimport
pPlinkRemove.lang         = params.Rscript.value
pPlinkRemove.script       = "file:scripts/plink/pPlinkRemove.r"

"""
@name:
	pPlink2Vcf
@description:
	Convert plink binaries into VCF file.
@input:
	`indir:dir`: The input directory containing .bed/.bim/.fam files
@output:
	`outfile:file`: The output vcf file.
@args:
	`plink`: The path to plink.
	`gz`   : Whether bgzip the output vcf file. Default: `False`
	`samid`: What to use as sample ID. Default: `both`
		- `both`: use `<FID>_<IID>` as sample id
		- `fid` : use `<FID>` as sample id
		- `iid` : use `<IID>` as sample id
"""
pPlink2Vcf             = Proc(desc = "Convert plink binary files to VCF file.")
pPlink2Vcf.input       = 'indir:dir'
pPlink2Vcf.output      = 'outfile:file:{{i.indir | bn}}.vcf{% if args.gz %}.gz{% endif %}'
pPlink2Vcf.args.plink  = params.plink.value
pPlink2Vcf.args.gz     = False
pPlink2Vcf.args.samid  = 'both' # fid, iid
pPlink2Vcf.args.chroms = {"23": "X", "24": "Y", "25": "XY", "26": "M"}
pPlink2Vcf.lang        = params.python.value
pPlink2Vcf.script      = "file:scripts/plink/pPlink2Vcf.py"

"""
@name:
	pPlink2GTMat
@description:
	Convert plink binaries into genotype matrix.
@input:
	`indir:dir`: The input directory containing .bed/.bim/.fam files
@output:
	`outfile:file`: The output genotype matrix file.
@args:
	`plink`: The path to plink.
	`samid`: What to use as sample ID. Default: `both`
		- `both`: use `<FID>_<IID>` as sample id
		- `fid` : use `<FID>` as sample id
		- `iid` : use `<IID>` as sample id
"""
pPlink2GTMat             = Proc(desc = "Convert plink binary files to genotype matrix")
pPlink2GTMat.input       = 'indir:dir'
pPlink2GTMat.output      = 'outfile:file:{{i.indir | bn}}.gtmat.txt'
pPlink2GTMat.args.plink  = params.plink.value
pPlink2GTMat.args.samid  = 'both' # fid, iid
pPlink2GTMat.args.chroms = {"23": "X", "24": "Y", "25": "XY", "26": "M"}
pPlink2GTMat.args.nors   = "NOVEL"
pPlink2GTMat.lang        = params.python.value
pPlink2GTMat.script      = "file:scripts/plink/pPlink2GTMat.py"
