<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../img/favicon.ico" />
    <title>biopipen.ns.protein - biopipen</title>
    <link rel="stylesheet" href="../../../css/theme.css" />
    <link rel="stylesheet" href="../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="//use.fontawesome.com/releases/v5.8.1/css/all.css" rel="stylesheet" />
        <link href="//use.fontawesome.com/releases/v5.8.1/css/v4-shims.css" rel="stylesheet" />
        <link href="../../../css/mkapi-common.css" rel="stylesheet" />
        <link href="../../../style.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "biopipen.ns.protein";
        var mkdocs_page_input_path = "api/source/biopipen.ns.protein.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.configure({ ignoreUnescapedHTML: true }); hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../..">
          <img src="../../../img/logo.png" class="logo" alt="Logo"/>
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="nav-scrollable-wrapper">
      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Pipelines</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../../pipelines/cnvkit_pipeline/">CNVkit Pipeline</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../../pipelines/cellranger_pipeline/">CellRanger Pipeline</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">API</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.utils</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils/">biopipen.utils</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.gene/">biopipen.utils.gene</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.reference/">biopipen.utils.reference</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.misc/">biopipen.utils.misc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.vcf/">biopipen.utils.vcf</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.core</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core/">biopipen.core</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.proc/">biopipen.core.proc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.filters/">biopipen.core.filters</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.testing/">biopipen.core.testing</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.config/">biopipen.core.config</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.defaults/">biopipen.core.defaults</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.ns</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns/">biopipen.ns</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.gene/">biopipen.ns.gene</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cellranger/">biopipen.ns.cellranger</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.scrna_metabolic_landscape/">biopipen.ns.scrna_metabolic_landscape</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.protein/">biopipen.ns.protein</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.delim/">biopipen.ns.delim</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.rnaseq/">biopipen.ns.rnaseq</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.plot/">biopipen.ns.plot</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.tcgamaf/">biopipen.ns.tcgamaf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.regulatory/">biopipen.ns.regulatory</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.stats/">biopipen.ns.stats</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cellranger_pipeline/">biopipen.ns.cellranger_pipeline</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.bed/">biopipen.ns.bed</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnv/">biopipen.ns.cnv</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.snp/">biopipen.ns.snp</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.web/">biopipen.ns.web</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.misc/">biopipen.ns.misc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnvkit/">biopipen.ns.cnvkit</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnvkit_pipeline/">biopipen.ns.cnvkit_pipeline</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.tcr/">biopipen.ns.tcr</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.scrna/">biopipen.ns.scrna</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.vcf/">biopipen.ns.vcf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.gsea/">biopipen.ns.gsea</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.bam/">biopipen.ns.bam</a>
                </li>
    </ul>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../../CHANGELOG/">Change log</a>
                </li>
              </ul>
      </div>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../..">biopipen</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../.." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">biopipen.ns.protein</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/pwwang/biopipen/edit/master/docs/api/source/biopipen.ns.protein.md">Edit on pwwang/biopipen</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="mkapi-node">
<h1 id="" class="mkapi-object mkapi-object-code plain">
<span class="mkapi-object-kind mkapi-object-kind-code">SOURCE CODE</span>
<span class="mkapi-object-prefix">biopipen.ns.</span><span class="mkapi-object-name">protein</span>
<span id="biopipen.ns.protein"></span><a class="mkapi-docs-link" href="../../biopipen.ns.protein">DOCS</a>
</h1>

<div class="mkapi-code">
<div class="codehilite"><pre><span></span><code><span class="sd">&quot;&quot;&quot;Protein-related processes.&quot;&quot;&quot;</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">..core.proc</span><span class="w"> </span><span class="kn">import</span> <span class="n">Proc</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">..core.config</span><span class="w"> </span><span class="kn">import</span> <span class="n">config</span>


<span class="k">class</span><span class="w"> </span><span class="nc">Prodigy</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.protein.Prodigy"></span><a class="mkapi-docs-link" title="biopipen.ns.protein.Prodigy" href="../../biopipen.ns.protein/#biopipen.ns.protein.Prodigy">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Prediction of binding affinity of protein-protein complexes based on</span>
<span class="sd">    intermolecular contacts using Prodigy.</span>

<span class="sd">    See &lt;https://rascar.science.uu.nl/prodigy/&gt; and</span>
<span class="sd">    &lt;https://github.com/haddocking/prodigy&gt;.</span>

<span class="sd">    `prodigy-prot` must be installed under the given python of `proc.lang`.</span>

<span class="sd">    Input:</span>
<span class="sd">        infile: The structure file in PDB or mmCIF format.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output file generated by Prodigy.</span>
<span class="sd">        outdir: The output directory containing all output files.</span>

<span class="sd">    Envs:</span>
<span class="sd">        distance_cutoff (type=float): The distance cutoff to calculate intermolecular</span>
<span class="sd">            contacts.</span>
<span class="sd">        acc_threshold (type=float): The accessibility threshold for BSA analysis.</span>
<span class="sd">        temperature (type=float): The temperature (C) for Kd prediction.</span>
<span class="sd">        contact_list (flag): Whether to generate contact list.</span>
<span class="sd">        pymol_selection (flag): Whether output a script to highlight the interface</span>
<span class="sd">            residues in PyMOL.</span>
<span class="sd">        selection (list): The selection of the chains to analyze.</span>
<span class="sd">            `[&#39;A&#39;, &#39;B&#39;]` will analyze chains A and B.</span>
<span class="sd">            `[&#39;A,B&#39;, &#39;C&#39;]` will analyze chain A and C; and B and C.</span>
<span class="sd">            `[&#39;A&#39;, &#39;B&#39;, &#39;C&#39;]` will analyze all combinations of A, B, and C.</span>
<span class="sd">        outtype (choice): Set the format of the output file (`out.outfile`).</span>
<span class="sd">            All three files will be generated. This option only determines which</span>
<span class="sd">            is assigned to `out.outfile`.</span>
<span class="sd">            - raw: The raw output file from prodigy.</span>
<span class="sd">            - json: The output file in JSON format.</span>
<span class="sd">            - tsv: The output file in CSV format.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;outfile:file:{{in.infile | stem}}_prodigy/&quot;</span>
        <span class="s2">&quot;{{in.infile | stem}}.{{envs.outtype if envs.outtype != &#39;raw&#39; else &#39;out&#39;}}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;outdir:dir:{{in.infile | stem}}_prodigy&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;distance_cutoff&quot;</span><span class="p">:</span> <span class="mf">5.5</span><span class="p">,</span>
        <span class="s2">&quot;acc_threshold&quot;</span><span class="p">:</span> <span class="mf">0.05</span><span class="p">,</span>
        <span class="s2">&quot;temperature&quot;</span><span class="p">:</span> <span class="mf">25.0</span><span class="p">,</span>
        <span class="s2">&quot;contact_list&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;pymol_selection&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;selection&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;outtype&quot;</span><span class="p">:</span> <span class="s2">&quot;json&quot;</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/protein/Prodigy.py&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">ProdigySummary</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.protein.ProdigySummary"></span><a class="mkapi-docs-link" title="biopipen.ns.protein.ProdigySummary" href="../../biopipen.ns.protein/#biopipen.ns.protein.ProdigySummary">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Summary of the output from `Prodigy`.</span>

<span class="sd">    Input:</span>
<span class="sd">        infiles: The output json file generated by `Prodigy`.</span>

<span class="sd">    Output:</span>
<span class="sd">        outdir: The directory of summary files generated by `ProdigySummary`.</span>

<span class="sd">    Envs:</span>
<span class="sd">        group (type=auto): The group of the samples for boxplots.</span>
<span class="sd">            If `None`, don&#39;t do boxplots.</span>
<span class="sd">            It can be a dict of group names and sample names, e.g.</span>
<span class="sd">            `{&quot;group1&quot;: [&quot;sample1&quot;, &quot;sample2&quot;], &quot;group2&quot;: [&quot;sample3&quot;]}`</span>
<span class="sd">            or a file containing the group information, with the first column</span>
<span class="sd">            being the sample names and the second column being the group names.</span>
<span class="sd">            The file should be tab-delimited with no header.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infiles:files&quot;</span>
    <span class="n">input_data</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">ch</span><span class="p">:</span> <span class="p">[[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">odir</span><span class="si">}</span><span class="s2">/_prodigy.tsv&quot;</span> <span class="k">for</span> <span class="n">odir</span> <span class="ow">in</span> <span class="n">ch</span><span class="o">.</span><span class="n">outdir</span><span class="p">]]</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outdir:dir:prodigy_summary&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/protein/ProdigySummary.R&quot;</span>
    <span class="n">plugin_opts</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="s2">&quot;file://../reports/protein/ProdigySummary.svelte&quot;</span><span class="p">}</span>


<span class="k">class</span><span class="w"> </span><span class="nc">MMCIF2PDB</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.protein.MMCIF2PDB"></span><a class="mkapi-docs-link" title="biopipen.ns.protein.MMCIF2PDB" href="../../biopipen.ns.protein/#biopipen.ns.protein.MMCIF2PDB">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert mmCIF or PDBx file to PDB file.</span>

<span class="sd">    Using [BeEM](https://github.com/kad-ecoli/BeEM)</span>

<span class="sd">    Input:</span>
<span class="sd">        infile: The input mmCIF or PDBx file.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output PDB file.</span>
<span class="sd">            The &quot;outfmt&quot; set to 3 to always output a single PDB file.</span>

<span class="sd">    Envs:</span>
<span class="sd">        tool (choice): The tool to use for conversion.</span>
<span class="sd">            - maxit: Use MAXIT.</span>
<span class="sd">            - beem: Use BeEM.</span>
<span class="sd">        maxit: The path to the MAXIT executable.</span>
<span class="sd">        beem: The path to the BeEM executable.</span>
<span class="sd">        &lt;more&gt;: Other options for MAXIT/BeEM.</span>
<span class="sd">            For BeEM, &quot;outfmt&quot; will not be used as it is set to 3.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.infile | stem}}.pdb&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;tool&quot;</span><span class="p">:</span> <span class="s2">&quot;maxit&quot;</span><span class="p">,</span>
        <span class="s2">&quot;maxit&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">exe</span><span class="o">.</span><span class="n">maxit</span><span class="p">,</span>
        <span class="s2">&quot;beem&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">exe</span><span class="o">.</span><span class="n">beem</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/protein/MMCIF2PDB.py&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">RMSD</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.protein.RMSD"></span><a class="mkapi-docs-link" title="biopipen.ns.protein.RMSD" href="../../biopipen.ns.protein/#biopipen.ns.protein.RMSD">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the RMSD between two structures.</span>

<span class="sd">    See also https://github.com/charnley/rmsd.</span>

<span class="sd">    If the input is in mmCIF format, convert it to PDB first.</span>

<span class="sd">    Input:</span>
<span class="sd">        infile1: The first structure file.</span>
<span class="sd">        infile2: The second structure file.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output file containing the RMSD value.</span>

<span class="sd">    Envs:</span>
<span class="sd">        beem: The path to the BeEM executable.</span>
<span class="sd">        calculate_rmsd: The path to the calculate_rmsd executable.</span>
<span class="sd">        conv_tool (choice): The tool to use for conversion.</span>
<span class="sd">            - maxit: Use MAXIT.</span>
<span class="sd">            - beem: Use BeEM.</span>
<span class="sd">        ca_only (flag): Whether to calculate RMSD using only C-alpha atoms.</span>
<span class="sd">        duel (choice): How to handle the duel atoms. Default is &quot;keep&quot;.</span>
<span class="sd">            - keep: Keep both atoms.</span>
<span class="sd">            - keep_first: Keep the first atom.</span>
<span class="sd">            - keep_last: Keep the last atom.</span>
<span class="sd">            - average: Average the coordinates.</span>
<span class="sd">        reorder (flag): Whether to reorder the atoms in the structures.</span>
<span class="sd">        &lt;more&gt;: Other options for calculate_rmsd.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infile1:file, infile2:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.infile1 | stem}}-{{in.infile2 | stem}}.rmsd.txt&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;maxit&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">exe</span><span class="o">.</span><span class="n">maxit</span><span class="p">,</span>
        <span class="s2">&quot;beem&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">exe</span><span class="o">.</span><span class="n">beem</span><span class="p">,</span>
        <span class="s2">&quot;calculate_rmsd&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">exe</span><span class="o">.</span><span class="n">calculate_rmsd</span><span class="p">,</span>
        <span class="s2">&quot;conv_tool&quot;</span><span class="p">:</span> <span class="s2">&quot;maxit&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ca_only&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;duel&quot;</span><span class="p">:</span> <span class="s2">&quot;keep&quot;</span><span class="p">,</span>
        <span class="s2">&quot;reorder&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/protein/RMSD.py&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">PDB2Fasta</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.protein.PDB2Fasta"></span><a class="mkapi-docs-link" title="biopipen.ns.protein.PDB2Fasta" href="../../biopipen.ns.protein/#biopipen.ns.protein.PDB2Fasta">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert PDB file to FASTA file.</span>

<span class="sd">    Input:</span>
<span class="sd">        infile: The input PDB file.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output FASTA file.</span>

<span class="sd">    Envs:</span>
<span class="sd">        chains (auto): The chains to extract. A list of chain IDs or separated by</span>
<span class="sd">            commas.</span>
<span class="sd">            If None, extract all chains.</span>
<span class="sd">        wrap (type=int): The number of residues per line in the output FASTA</span>
<span class="sd">            file. Set to 0 to disable wrapping.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.infile | stem}}.fasta&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">python</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;chains&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;wrap&quot;</span><span class="p">:</span> <span class="mi">80</span><span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/protein/PDB2Fasta.py&quot;</span>
</code></pre></div>
</div>
</div>
              
            </div>
          </div><footer>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  <div class="footer-notes">
    <a href="https://github.com/pwwang/biopipen" title="Open repository in GitHub">Github Repo</a>
    <div class="footer-sep"></div>Built with <a href="https://www.mkdocs.org/">MkDocs</a> using theme <a href="https://github.com/pwwang/mkdocs-rtd">mkdocs-rtd</a>.
  </div>
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
  </span>
</div>
    <script src="../../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../../..";</script>
    <script src="../../../js/theme_extra.js"></script>
    <script src="../../../js/theme.js"></script>
    <script src="../../../js/version-select.js"></script>
      <script src="../../../js/mkapi.js"></script>
      <script src="../../../extra.js"></script>
      <script src="../../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
