<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Scrna metabolic - biopipen</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="//use.fontawesome.com/releases/v5.8.1/css/all.css" rel="stylesheet" />
        <link href="//use.fontawesome.com/releases/v5.8.1/css/v4-shims.css" rel="stylesheet" />
        <link href="../../css/mkapi-common.css" rel="stylesheet" />
        <link href="../../style.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Scrna metabolic";
        var mkdocs_page_input_path = "pipelines/scrna_metabolic.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.configure({ ignoreUnescapedHTML: true }); hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../..">
          <img src="../../img/logo.png" class="logo" alt="Logo"/>
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="nav-scrollable-wrapper">
      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Pipelines</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../cnvkit_pipeline/">CNVkit Pipeline</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../cellranger_pipeline/">CellRanger Pipeline</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">API</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.utils</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.utils/">biopipen.utils</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.utils.vcf/">biopipen.utils.vcf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.utils.reference/">biopipen.utils.reference</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.utils.misc/">biopipen.utils.misc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.utils.common_docstrs/">biopipen.utils.common_docstrs</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.utils.gene/">biopipen.utils.gene</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.core</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.core/">biopipen.core</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.core.proc/">biopipen.core.proc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.core.defaults/">biopipen.core.defaults</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.core.config/">biopipen.core.config</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.core.filters/">biopipen.core.filters</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.core.testing/">biopipen.core.testing</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.ns</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns/">biopipen.ns</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.protein/">biopipen.ns.protein</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.cellranger/">biopipen.ns.cellranger</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.cnvkit/">biopipen.ns.cnvkit</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.cnv/">biopipen.ns.cnv</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.web/">biopipen.ns.web</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.cellranger_pipeline/">biopipen.ns.cellranger_pipeline</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.bam/">biopipen.ns.bam</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.scrna/">biopipen.ns.scrna</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.tcr/">biopipen.ns.tcr</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.tcgamaf/">biopipen.ns.tcgamaf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.vcf/">biopipen.ns.vcf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.gsea/">biopipen.ns.gsea</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.scrna_metabolic_landscape/">biopipen.ns.scrna_metabolic_landscape</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.delim/">biopipen.ns.delim</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.snp/">biopipen.ns.snp</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.plot/">biopipen.ns.plot</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.misc/">biopipen.ns.misc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.rnaseq/">biopipen.ns.rnaseq</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.gene/">biopipen.ns.gene</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.stats/">biopipen.ns.stats</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.regulatory/">biopipen.ns.regulatory</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.cnvkit_pipeline/">biopipen.ns.cnvkit_pipeline</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../api/biopipen.ns.bed/">biopipen.ns.bed</a>
                </li>
    </ul>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../CHANGELOG/">Change log</a>
                </li>
              </ul>
      </div>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">biopipen</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Scrna metabolic</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/pwwang/biopipen/edit/master/docs/pipelines/scrna_metabolic.md">Edit on pwwang/biopipen</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="scrna_metabolic">scrna_metabolic</h2>
<p>Metabolic landscape analysis for single-cell RNA-seq data</p>
<p>An abstract from https://github.com/LocasaleLab/Single-Cell-Metabolic-Landscape</p>
<h3 id="reference">Reference</h3>
<blockquote>
<p>Xiao, Zhengtao, Ziwei Dai, and Jason W. Locasale. "Metabolic landscape of the tumor microenvironment at single cell resolution." Nature communications 10.1 (2019): 1-12.</p>
</blockquote>
<h3 id="run-the-pipeline">Run the pipeline</h3>
<h4 id="run-from-cli">Run from CLI:</h4>
<div class="highlight"><pre><span></span><code>pipen<span class="w"> </span>run<span class="w"> </span>scrna_metabolic_landscape<span class="w"> </span>ScrnaMetabolicLandscape<span class="w"> </span><span class="o">[</span>options<span class="o">]</span>
</code></pre></div>
<h4 id="serve-as-part-of-a-pipeline">Serve as part of a pipeline:</h4>
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">biopipen.ns.scrna_metabolic_landscape</span><span class="w"> </span><span class="kn">import</span> <span class="n">ScrnaMetabolicLandscape</span>

<span class="n">pipeline</span> <span class="o">=</span> <span class="n">ScrnaMetabolicLandscape</span><span class="p">(</span><span class="o">&lt;</span><span class="n">options</span><span class="o">&gt;</span><span class="p">)</span>

<span class="c1"># You can specify dependencies so that the whole metabolic landscape pipeline</span>
<span class="c1"># works as a part of another pipeline</span>
</code></pre></div>
<h3 id="inputs">Inputs</h3>
<ul>
<li>
<p><code>metafile</code>: Either a metafile or an rds file of a Seurat object.
    If it is a metafile, it should have two columns: <code>Sample</code> and <code>RNAData</code>.
    <code>Sample</code> should be the first column with unique identifiers for the samples
    and <code>RNAData</code> indicates where the barcodes, genes, expression matrices are.
    The data will be loaded and an unsupervised clustering will be done.
    Currently only 10X data is supported.
    If it is an rds file, the seurat object will be used directly.</p>
</li>
<li>
<p><code>gmtfile</code>: The GMT file with the metabolic pathways. The gene names should
    match the gene names in the gene list in RNAData or the Seurat object.
    You can also provide a URL to the GMT file.</p>
</li>
<li>
<p><code>group_by</code>: Group the data by the given column in the metadata. For example, <code>cluster</code>.</p>
</li>
<li>
<p><code>subset_by</code>: (Optional) Subset the data by the given column in the metadata.
    For example, <code>Response</code>. <code>NA</code> values will be removed in this column.
    If None, the data will not be subsetted.</p>
</li>
<li>
<p><code>mutaters</code>: (Optional) Add new columns to the metadata for grouping/subsetting.
    They are passed to <code>sobj@meta.data |&gt; mutate(...)</code>. For example,
    <code>{"timepoint": "if_else(treatment == 'control', 'pre', 'post')"}</code>
    will add a new column <code>timepoint</code> to the metadata with values of
    <code>pre</code> and <code>post</code> based on the <code>treatment</code> column.</p>
</li>
<li>
<p><code>cases</code>: (Optional) Multiple cases for the analysis. If you have multiple different
    grouping/subsetting scenarios, you can specify them here. Each case can have
    its own <code>subset_by</code>, <code>group_by</code>, and analysis parameters.</p>
</li>
</ul>
<h3 id="advanced-configuration">Advanced Configuration</h3>
<h4 id="multiple-cases">Multiple Cases</h4>
<p>You can define multiple analysis cases with different grouping/subsetting strategies:</p>
<div class="highlight"><pre><span></span><code><span class="k">[ScrnaMetabolicLandscape]</span>
<span class="n">metafile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;test_data/scrna_metabolic/seurat_obj.rds&quot;</span>
<span class="n">gmtfile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;test_data/scrna_metabolic/KEGG_metabolism.gmt&quot;</span>

<span class="k">[ScrnaMetabolicLandscape.MetabolicPathwayActivity.envs.cases]</span>
<span class="s2">&quot;By Treatment&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">{</span><span class="w"> </span><span class="n">group_by</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s2">&quot;seurat_clusters&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">subset_by</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s2">&quot;treatment&quot;</span><span class="w"> </span><span class="p">}</span>
<span class="s2">&quot;By Response&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">{</span><span class="w"> </span><span class="n">group_by</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s2">&quot;seurat_clusters&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">subset_by</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s2">&quot;response&quot;</span><span class="w"> </span><span class="p">}</span>
</code></pre></div>
<h4 id="custom-plotting">Custom Plotting</h4>
<p>Each process supports customizable plots:</p>
<div class="highlight"><pre><span></span><code><span class="k">[ScrnaMetabolicLandscape.MetabolicPathwayActivity.envs.plots]</span>
<span class="s2">&quot;Custom Heatmap&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">plot_type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;heatmap&quot;</span><span class="err">,</span>
<span class="w">    </span><span class="n">show_row_names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">true</span><span class="err">,</span>
<span class="w">    </span><span class="n">devpars</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">{</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="mi">1200</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="mi">800</span><span class="p">,</span><span class="w"> </span><span class="n">res</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="mi">150</span><span class="w"> </span><span class="p">}</span>
<span class="err">}</span>
<span class="s2">&quot;Custom Violin&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">plot_type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;violin&quot;</span><span class="err">,</span>
<span class="w">    </span><span class="n">add_box</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">true</span><span class="err">,</span>
<span class="w">    </span><span class="n">devpars</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">{</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="mi">1000</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="mi">600</span><span class="w"> </span><span class="p">}</span>
<span class="err">}</span>

<span class="k">[ScrnaMetabolicLandscape.MetabolicFeatures.envs.plots]</span>
<span class="s2">&quot;Top 5 Summary&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">plot_type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;summary&quot;</span><span class="err">,</span>
<span class="w">    </span><span class="n">top_term</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">5</span><span class="err">,</span>
<span class="w">    </span><span class="n">level</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;subset&quot;</span>
<span class="err">}</span>
</code></pre></div>
<h3 id="a-step-by-step-example">A step-by-step example</h3>
<h4 id="prepare-the-seurat-object">Prepare the seurat object</h4>
<p>Using the data from:</p>
<blockquote>
<p>Yost KE, Satpathy AT, Wells DK, Qi Y et al. Clonal replacement of tumor-specific T cells following PD-1 blockade. Nat Med 2019 Aug;25(8):1251-1259. PMID: 31359002</p>
</blockquote>
<div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>

<span class="c1"># Download data (tcell rds and metadata) from</span>
<span class="c1"># https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE123813</span>

<span class="n">count_file</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;test_data/scrna_metabolic/GSE123813_bcc_scRNA_counts.txt.gz&quot;</span>
<span class="n">meta_file</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;test_data/scrna_metabolic/GSE123813_bcc_all_metadata.txt.gz&quot;</span>

<span class="n">counts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.table</span><span class="p">(</span><span class="n">count_file</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">check.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">F</span><span class="p">)</span>
<span class="n">metadata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.table</span><span class="p">(</span><span class="n">meta_file</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">check.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">F</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># Subset 1000 cells for just demo purpose</span>
<span class="n">counts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">counts</span><span class="p">[,</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="nf">ncol</span><span class="p">(</span><span class="n">counts</span><span class="p">),</span><span class="w"> </span><span class="m">1000</span><span class="p">)]</span>
<span class="n">metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">metadata</span><span class="p">[</span><span class="nf">colnames</span><span class="p">(</span><span class="n">counts</span><span class="p">),]</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># Create seurat object</span>
<span class="n">seurat_obj</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">CreateSeuratObject</span><span class="p">(</span><span class="n">counts</span><span class="o">=</span><span class="n">counts</span><span class="p">)</span>
<span class="n">seurat_obj</span><span class="o">@</span><span class="n">meta.data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">cbind</span><span class="p">(</span>
<span class="w">    </span><span class="n">seurat_obj</span><span class="o">@</span><span class="n">meta.data</span><span class="p">,</span>
<span class="w">    </span><span class="n">metadata</span><span class="p">[</span><span class="nf">rownames</span><span class="p">(</span><span class="n">seurat_obj</span><span class="o">@</span><span class="n">meta.data</span><span class="p">),]</span>
<span class="p">)</span>
<span class="n">seurat_obj</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">NormalizeData</span><span class="p">(</span><span class="n">seurat_obj</span><span class="p">)</span>
<span class="n">all.genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">seurat_obj</span><span class="p">)</span>
<span class="n">seurat_obj</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ScaleData</span><span class="p">(</span><span class="n">seurat_obj</span><span class="p">,</span><span class="w"> </span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">all.genes</span><span class="p">)</span>
<span class="n">seurat_obj</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindVariableFeatures</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">seurat_obj</span><span class="p">)</span>
<span class="n">seurat_obj</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">RunPCA</span><span class="p">(</span><span class="n">seurat_obj</span><span class="p">,</span><span class="w"> </span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">VariableFeatures</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">seurat_obj</span><span class="p">))</span>

<span class="c1"># Output exceeds the size limit. Open the full output data in a text editor</span>

<span class="c1"># Warning message:</span>
<span class="c1"># &quot;Feature names cannot have underscores (&#39;_&#39;), replacing with dashes (&#39;-&#39;)&quot;</span>
<span class="c1"># Centering and scaling data matrix</span>

<span class="c1"># PC_ 1</span>
<span class="c1"># Positive:  CALD1, EMP2, SPARC, CAV1, EMP1, ACTN1, KRT17, BGN, DSP, RAB13</span>
<span class="c1"># 	   RND3, S100A16, KRT14, S100A14, S100A2, CD9, FHL2, IER3, GJB2, TRIM29</span>
<span class="c1"># 	   TSC22D1, KRT15, SFN, FSTL1, DDR1, IGFBP7, JUP, PTRF, KRT5, FOSL1</span>
<span class="c1"># Negative:  CD74, CRIP1, RARRES3, RGS1, LAT, CD69, NKG7, HLA-DPA1, TNFRSF4, CD27</span>
<span class="c1"># 	   GZMA, HLA-DRB1, CXCR6, CTSW, GPR183, LDLRAD4, ICOS, HIST1H4C, GZMK, AC092580.4</span>
<span class="c1"># 	   SLC9A3R1, CCR7, S100A4, HLA-DPB1, HMGB2, GBP5, SELL, CXCR3, LAG3, HLA-DRB5</span>
<span class="c1"># PC_ 2</span>
<span class="c1"># Positive:  DSP, DSC3, SFN, SERPINB5, KRT17, S100A14, KRT15, KRT16, IRF6, TRIM29</span>
<span class="c1"># 	   KRT14, LGALS7B, JUP, PERP, TACSTD2, GJB2, KRT5, KRT6B, DDR1, S100A2</span>
<span class="c1"># 	   PKP1, CDH3, KRT6A, FXYD3, GJB3, MPZL2, CXADR, DSC2, DSG3, GRHL3</span>
<span class="c1"># Negative:  COL1A2, COL3A1, LUM, COL6A1, MXRA8, FN1, CTSK, COL1A1, COL6A3, DCN</span>
<span class="c1"># 	   MMP2, COL6A2, PRRX1, FKBP10, TNFAIP6, FAP, PCOLCE, PDGFRB, NNMT, AEBP1</span>
<span class="c1"># 	   C1S, CCDC80, SFRP2, RCN3, PDPN, SERPINF1, COL5A2, CTHRC1, COL5A1, COL12A1</span>
<span class="c1"># PC_ 3</span>
<span class="c1"># Positive:  LYZ, TYROBP, SPI1, FCER1G, KYNU, C15orf48, BCL2A1, HLA-DRA, CD68, AIF1</span>
<span class="c1"># 	   CST3, CTSZ, SERPINA1, CSF2RA, HLA-DRB5, SLC7A11, LST1, MS4A7, ALDH2, FAM49A</span>
<span class="c1"># 	   IFI30, HLA-DRB1, GPR157, PLEK, HLA-DPB1, IL1B, CD86, CCDC88A, HLA-DPA1, RNF144B</span>
<span class="c1"># Negative:  MT2A, MT1X, BGN, MT1E, COL6A2, COL1A2, COL6A1, COL5A2, MXRA8, DCN</span>
<span class="c1"># 	   COL12A1, COL3A1, COL1A1, LUM, MFAP2, PCOLCE, MT1F, MMP2, COL5A1, COL6A3</span>
<span class="c1"># 	   PRRX1, C1S, AEBP1, CTSK, FAP, LAT, PDGFRB, RARRES3, THY1, EFEMP2</span>
<span class="c1"># 	   GJB3, LYPD3, GRHL3, PVRL4, KRT16, TACSTD2, GJB5, SERPINB13, MPZL2, KRT23</span>
<span class="c1"># Negative:  UBE2C, GTSE1, BIRC5, RRM2, CCNA2, TYMS, TOP2A, TK1, DLGAP5, MKI67</span>
<span class="c1"># 	   PKMYT1, CENPA, KIFC1, CDCA5, UHRF1, ASF1B, AURKB, FAM111B, TROAP, CKAP2L</span>
<span class="c1"># 	   HJURP, ESCO2, FOXM1, CDK1, ZWINT, E2F2, CLSPN, HIST1H1B, CDT1, MCM10</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># By default, the pipeline will do the clustering using the SeuratClustering process</span>
<span class="c1"># If you want to do the clustering yourself, you can set `is_seurat = true`</span>
<span class="c1"># when running the pipeline, which will skip the clustering step.</span>

<span class="n">seurat_obj</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindNeighbors</span><span class="p">(</span><span class="n">seurat_obj</span><span class="p">,</span><span class="w"> </span><span class="n">dims</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">)</span>
<span class="n">seurat_obj</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">FindClusters</span><span class="p">(</span><span class="n">seurat_obj</span><span class="p">,</span><span class="w"> </span><span class="n">resolution</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="nf">Idents</span><span class="p">(</span><span class="n">seurat_obj</span><span class="p">))</span>
<span class="c1"># Computing nearest neighbor graph</span>

<span class="c1"># Computing SNN</span>

<span class="c1"># Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span>

<span class="c1"># Number of nodes: 1000</span>
<span class="c1"># Number of edges: 30631</span>

<span class="c1"># Running Louvain algorithm...</span>
<span class="c1"># Maximum modularity in 10 random starts: 0.8795</span>
<span class="c1"># Number of communities: 9</span>
<span class="c1"># Elapsed time: 0 seconds</span>

<span class="c1"># bcc.su009.post.tcell_CCATTCGCAATCTACG 0</span>
<span class="c1"># bcc.su004.pre.tcell_ACAGCTACACTTCTGC  0</span>
<span class="c1"># bcc.su006.pre.tcell_CGTGTAAAGTGACTCT  1</span>
<span class="c1"># bcc.su001.post.tcell_CCTTTCTGTACCGTTA 2</span>
<span class="c1"># bcc.su007.pre.tcell_ATTTCTGAGAAGGACA  1</span>
<span class="c1"># bcc.su009.pre.tcell_AGACGTTTCCTGCAGG8 8</span>
<span class="c1"># Levels:</span>
<span class="c1"># &#39;0&#39;&#39;1&#39;&#39;2&#39;&#39;3&#39;&#39;4&#39;&#39;5&#39;&#39;6&#39;&#39;7&#39;&#39;8&#39;</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># Check the meta.data</span>
<span class="nf">head</span><span class="p">(</span><span class="n">seurat_obj</span><span class="o">@</span><span class="n">meta.data</span><span class="p">)</span>
<span class="c1"># 	orig.ident	nCount_RNA	nFeature_RNA	patient	treatment	sort	cluster	UMAP1	UMAP2	RNA_snn_res.0.5	seurat_clusters</span>
<span class="c1"># &lt;fct&gt;	&lt;dbl&gt;	&lt;int&gt;	&lt;chr&gt;	&lt;chr&gt;	&lt;chr&gt;	&lt;chr&gt;	&lt;dbl&gt;	&lt;dbl&gt;	&lt;fct&gt;	&lt;fct&gt;</span>
<span class="c1"># bcc.su009.post.tcell_CCATTCGCAATCTACG	bcc.su009.post.tcell	4242	1726	su009	post	CD45+ CD3+	CD8_mem_T_cells	-9.1816435	0.7484789	0	0</span>
<span class="c1"># bcc.su004.pre.tcell_ACAGCTACACTTCTGC	bcc.su004.pre.tcell	3159	1466	su004	pre	CD45+ CD3+	CD8_ex_T_cells	4.1067562	3.3754938	0	0</span>
<span class="c1"># bcc.su006.pre.tcell_CGTGTAAAGTGACTCT	bcc.su006.pre.tcell	3279	1401	su006	pre	CD45+ CD3+	Tregs	0.4816457	11.9388428	1	1</span>
<span class="c1"># bcc.su001.post.tcell_CCTTTCTGTACCGTTA	bcc.su001.post.tcell	5057	2218	su001	post	CD45+ CD3+	CD8_ex_T_cells	2.3045983	7.4856248	2	2</span>
<span class="c1"># bcc.su007.pre.tcell_ATTTCTGAGAAGGACA	bcc.su007.pre.tcell	3701	1413	su007	pre	CD45+ CD3+	CD8_mem_T_cells	-1.9617293	5.5546365	1	1</span>
<span class="c1"># bcc.su009.pre.tcell_AGACGTTTCCTGCAGG	bcc.su009.pre.tcell	3891	1593	su009	pre	CD45+ CD3+	Tcell_prolif	5.2243228	-1.0460106	8	8</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># save seurat object</span>
<span class="nf">saveRDS</span><span class="p">(</span><span class="n">seurat_obj</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;test_data/scrna_metabolic/seurat_obj.rds&quot;</span><span class="p">)</span>
</code></pre></div>
<h4 id="prepare-the-pathway-file">Prepare the pathway file</h4>
<p>A set of collected metabolic pathways can be found here:</p>
<p>https://github.com/LocasaleLab/Single-Cell-Metabolic-Landscape/blob/master/Data/KEGG_metabolism.gmt</p>
<p>Download and save it to <code>test_data/scrna_metabolic/KEGG_metabolism.gmt</code></p>
<h4 id="prepare-the-configuration-file">Prepare the configuration file</h4>
<p>Save at <code>test_data/scrna_metabolic/config.toml</code>:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Pipeline configuration</span>
<span class="k">[ScrnaMetabolicLandscape]</span>
<span class="n">metafile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;test_data/scrna_metabolic/seurat_obj.rds&quot;</span>
<span class="n">gmtfile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;test_data/scrna_metabolic/KEGG_metabolism.gmt&quot;</span>
<span class="n">group_by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;seurat_clusters&quot;</span>
<span class="n">subset_by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;timepoint&quot;</span>

<span class="k">[ScrnaMetabolicLandscape.mutaters]</span>
<span class="n">timepoint</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;if_else(patient != &#39;su001&#39;, NA_character_, treatment)&quot;</span>

<span class="c1"># Optional: Configure individual processes</span>
<span class="k">[ScrnaMetabolicLandscape.MetabolicPathwayActivity.envs]</span>
<span class="n">ntimes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">1000</span>

<span class="k">[ScrnaMetabolicLandscape.MetabolicFeatures.envs.plots]</span>
<span class="s2">&quot;Summary Plot&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">{</span><span class="w"> </span><span class="n">plot_type</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s2">&quot;summary&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">top_term</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="mi">5</span><span class="w"> </span><span class="p">}</span>
</code></pre></div>
<h4 id="run-the-pipeline_1">Run the pipeline</h4>
<div class="highlight"><pre><span></span><code>pipen<span class="w"> </span>run<span class="w"> </span>scrna_metabolic_landscape<span class="w"> </span>ScrnaMetabolicLandscape<span class="w"> </span>--config<span class="w"> </span>test_data/scrna_metabolic/config.toml
</code></pre></div>
<h4 id="check-out-the-resultsreports">Check out the results/reports</h4>
<p>The results can be found at <code>./ScrnaMetabolicLandscape_results/</code>, and reports can be found at <code>./ScrnaMetabolicLandscape_results/REPORTS</code>. To check out the reports, open <code>./ScrnaMetabolicLandscape_results/REPORTS/index.html</code> in your browser.</p>
<p>There are 3 parts of the results:</p>
<ul>
<li>
<p><code>MetabolicPathwayActivity</code>:</p>
<p>The pathway activities for groups (defined by <code>group_by</code> in the configuration) for each subset (defined by <code>subset_by</code>)</p>
</li>
<li>
<p><code>MetabolicPathwayHeterogeneity</code>:</p>
<p>Pathway heterogeneity for groups for each subset</p>
</li>
<li>
<p><code>MetabolicFeatures</code>:</p>
<p>The pathway enrichment analysis in detail for each group against the rest of the groups in each subset.</p>
</li>
</ul>
              
            </div>
          </div><footer>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  <div class="footer-notes">
    <a href="https://github.com/pwwang/biopipen" title="Open repository in GitHub">Github Repo</a>
    <div class="footer-sep"></div>Built with <a href="https://www.mkdocs.org/">MkDocs</a> using theme <a href="https://github.com/pwwang/mkdocs-rtd">mkdocs-rtd</a>.
  </div>
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
    <script src="../../js/version-select.js"></script>
      <script src="../../js/mkapi.js"></script>
      <script src="../../extra.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
