<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../img/favicon.ico" />
    <title>biopipen.ns.stats - biopipen</title>
    <link rel="stylesheet" href="../../../css/theme.css" />
    <link rel="stylesheet" href="../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="//use.fontawesome.com/releases/v5.8.1/css/all.css" rel="stylesheet" />
        <link href="//use.fontawesome.com/releases/v5.8.1/css/v4-shims.css" rel="stylesheet" />
        <link href="../../../css/mkapi-common.css" rel="stylesheet" />
        <link href="../../../style.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "biopipen.ns.stats";
        var mkdocs_page_input_path = "api/source/biopipen.ns.stats.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.configure({ ignoreUnescapedHTML: true }); hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../..">
          <img src="../../../img/logo.png" class="logo" alt="Logo"/>
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="nav-scrollable-wrapper">
      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Pipelines</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../../pipelines/cnvkit_pipeline/">CNVkit Pipeline</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../../pipelines/cellranger_pipeline/">CellRanger Pipeline</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">API</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.utils</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils/">biopipen.utils</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.gene/">biopipen.utils.gene</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.reference/">biopipen.utils.reference</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.misc/">biopipen.utils.misc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.vcf/">biopipen.utils.vcf</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.core</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core/">biopipen.core</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.proc/">biopipen.core.proc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.filters/">biopipen.core.filters</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.testing/">biopipen.core.testing</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.config/">biopipen.core.config</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.defaults/">biopipen.core.defaults</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.ns</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns/">biopipen.ns</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.gene/">biopipen.ns.gene</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cellranger/">biopipen.ns.cellranger</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.scrna_metabolic_landscape/">biopipen.ns.scrna_metabolic_landscape</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.protein/">biopipen.ns.protein</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.delim/">biopipen.ns.delim</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.rnaseq/">biopipen.ns.rnaseq</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.plot/">biopipen.ns.plot</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.tcgamaf/">biopipen.ns.tcgamaf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.regulatory/">biopipen.ns.regulatory</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.stats/">biopipen.ns.stats</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cellranger_pipeline/">biopipen.ns.cellranger_pipeline</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.bed/">biopipen.ns.bed</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnv/">biopipen.ns.cnv</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.snp/">biopipen.ns.snp</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.web/">biopipen.ns.web</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.misc/">biopipen.ns.misc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnvkit/">biopipen.ns.cnvkit</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnvkit_pipeline/">biopipen.ns.cnvkit_pipeline</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.tcr/">biopipen.ns.tcr</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.scrna/">biopipen.ns.scrna</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.vcf/">biopipen.ns.vcf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.gsea/">biopipen.ns.gsea</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.bam/">biopipen.ns.bam</a>
                </li>
    </ul>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../../CHANGELOG/">Change log</a>
                </li>
              </ul>
      </div>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../..">biopipen</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../.." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">biopipen.ns.stats</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/pwwang/biopipen/edit/master/docs/api/source/biopipen.ns.stats.md">Edit on pwwang/biopipen</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="mkapi-node">
<h1 id="" class="mkapi-object mkapi-object-code plain">
<span class="mkapi-object-kind mkapi-object-kind-code">SOURCE CODE</span>
<span class="mkapi-object-prefix">biopipen.ns.</span><span class="mkapi-object-name">stats</span>
<span id="biopipen.ns.stats"></span><a class="mkapi-docs-link" href="../../biopipen.ns.stats">DOCS</a>
</h1>

<div class="mkapi-code">
<div class="codehilite"><pre><span></span><code><span class="sd">&quot;&quot;&quot;Provides processes for statistics.&quot;&quot;&quot;</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">..core.proc</span><span class="w"> </span><span class="kn">import</span> <span class="n">Proc</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">..core.config</span><span class="w"> </span><span class="kn">import</span> <span class="n">config</span>


<span class="k">class</span><span class="w"> </span><span class="nc">ChowTest</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.stats.ChowTest"></span><a class="mkapi-docs-link" title="biopipen.ns.stats.ChowTest" href="../../biopipen.ns.stats/#biopipen.ns.stats.ChowTest">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Massive Chow tests.</span>

<span class="sd">    See Also https://en.wikipedia.org/wiki/Chow_test</span>

<span class="sd">    Input:</span>
<span class="sd">        infile: The input data file. The rows are samples and the columns are</span>
<span class="sd">            features. It must be tab-delimited.</span>
<span class="sd">            ```</span>
<span class="sd">            Sample   F1   F2   F3   ...   Fn</span>
<span class="sd">            S1       1.2  3.4  5.6        7.8</span>
<span class="sd">            S2       2.3  4.5  6.7        8.9</span>
<span class="sd">            ...</span>
<span class="sd">            Sm       5.6  7.8  9.0        1.2</span>
<span class="sd">            ```</span>
<span class="sd">        groupfile: The group file. The rows are the samples and the columns</span>
<span class="sd">            are the groupings. It must be tab-delimited.</span>
<span class="sd">            ```</span>
<span class="sd">            Sample   G1   G2   G3   ...   Gk</span>
<span class="sd">            S1       0    1    0          0</span>
<span class="sd">            S2       2    1    0          NA  # exclude this sample</span>
<span class="sd">            ...</span>
<span class="sd">            Sm       1    0    0          0</span>
<span class="sd">            ```</span>
<span class="sd">        fmlfile: The formula file. The first column is grouping and the</span>
<span class="sd">            second column is the formula. It must be tab-delimited.</span>
<span class="sd">            ```</span>
<span class="sd">            Group   Formula  ...  # Other columns to be added to outfile</span>
<span class="sd">            G1      Fn ~ F1 + Fx + Fy   # Fx, Fy could be covariates</span>
<span class="sd">            G1      Fn ~ F2 + Fx + Fy</span>
<span class="sd">            ...</span>
<span class="sd">            Gk      Fn ~ F3 + Fx + Fy</span>
<span class="sd">            ```</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output file. It is a tab-delimited file with the first</span>
<span class="sd">            column as the grouping and the second column as the p-value.</span>
<span class="sd">            ```</span>
<span class="sd">            Group  Formula  ...  Pooled  Groups  SSR  SumSSR  Fstat  Pval  Padj</span>
<span class="sd">            G1     Fn ~ F1       0.123   2       1    0.123   0.123  0.123 0.123</span>
<span class="sd">            G1     Fn ~ F2       0.123   2       1    0.123   0.123  0.123 0.123</span>
<span class="sd">            ...</span>
<span class="sd">            Gk     Fn ~ F3       0.123   2       1    0.123   0.123  0.123 0.123</span>
<span class="sd">            ```</span>

<span class="sd">    Envs:</span>
<span class="sd">        padj (choice): The method for p-value adjustment.</span>
<span class="sd">            - none: No p-value adjustment (no Padj column in outfile).</span>
<span class="sd">            - holm: Holm-Bonferroni method.</span>
<span class="sd">            - hochberg: Hochberg method.</span>
<span class="sd">            - hommel: Hommel method.</span>
<span class="sd">            - bonferroni: Bonferroni method.</span>
<span class="sd">            - BH: Benjamini-Hochberg method.</span>
<span class="sd">            - BY: Benjamini-Yekutieli method.</span>
<span class="sd">            - fdr: FDR correction method.</span>
<span class="sd">        transpose_input (flag): Whether to transpose the input file.</span>
<span class="sd">        transpose_group (flag): Whether to transpose the group file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infile:file, groupfile:file, fmlfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.infile | stem}}.chowtest.txt&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;padj&quot;</span><span class="p">:</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span>
        <span class="s2">&quot;transpose_input&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;transpose_group&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/stats/ChowTest.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">Mediation</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.stats.Mediation"></span><a class="mkapi-docs-link" title="biopipen.ns.stats.Mediation" href="../../biopipen.ns.stats/#biopipen.ns.stats.Mediation">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Mediation analysis.</span>

<span class="sd">    The flowchart of mediation analysis:</span>

<span class="sd">    ![Mediation Analysis](https://library.virginia.edu/sites/default/files/inline-images/mediation_flowchart-1.png)</span>

<span class="sd">    Reference:</span>
<span class="sd">        - &lt;https://library.virginia.edu/data/articles/introduction-to-mediation-analysis&gt;</span>
<span class="sd">        - &lt;https://en.wikipedia.org/wiki/Mediation_(statistics)&gt;</span>
<span class="sd">        - &lt;https://tilburgsciencehub.com/topics/analyze/regression/linear-regression/mediation-analysis/&gt;</span>
<span class="sd">        - &lt;https://ademos.people.uic.edu/Chapter14.html&gt;</span>

<span class="sd">    Input:</span>
<span class="sd">        infile: The input data file. The rows are samples and the columns are</span>
<span class="sd">            features. It must be tab-delimited.</span>
<span class="sd">            ```</span>
<span class="sd">            Sample   F1   F2   F3   ...   Fn</span>
<span class="sd">            S1       1.2  3.4  5.6        7.8</span>
<span class="sd">            S2       2.3  4.5  6.7        8.9</span>
<span class="sd">            ...</span>
<span class="sd">            Sm       5.6  7.8  9.0        1.2</span>
<span class="sd">            ```</span>
<span class="sd">        fmlfile: The formula file.</span>
<span class="sd">            ```</span>
<span class="sd">            Case   M   Y   X   Cov     Model_M    Model_Y</span>
<span class="sd">            Case1  F1  F2  F3  F4,F5   glm        lm</span>
<span class="sd">            ...</span>
<span class="sd">            ```</span>
<span class="sd">            Where Y is the outcome variable, X is the predictor variable, M is the</span>
<span class="sd">            mediator variable, and Case is the case name. Model_M and Model_Y are the</span>
<span class="sd">            models for M and Y, respectively.</span>
<span class="sd">            `envs.cases` will be ignored if this is provided.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output file.</span>
<span class="sd">            Columns to help understand the results:</span>
<span class="sd">            Total Effect: a total effect of X on Y (without M) (`Y ~ X`).</span>
<span class="sd">            ADE: A Direct Effect of X on Y after taking into account a mediation effect of M (`Y ~ X + M`).</span>
<span class="sd">            ACME: The Mediation Effect, the total effect minus the direct effect,</span>
<span class="sd">            which equals to a product of a coefficient of X in the second step and a coefficient of M in the last step.</span>
<span class="sd">            The goal of mediation analysis is to obtain this indirect effect and see if it&#39;s statistically significant.</span>

<span class="sd">    Envs:</span>
<span class="sd">        ncores (type=int): Number of cores to use for parallelization for cases.</span>
<span class="sd">        sims (type=int): Number of Monte Carlo draws for nonparametric bootstrap or quasi-Bayesian approximation.</span>
<span class="sd">            Will be passed to `mediation::mediate` function.</span>
<span class="sd">        args (ns): Other arguments passed to `mediation::mediate` function.</span>
<span class="sd">            - &lt;more&gt;: More arguments passed to `mediation::mediate` function.</span>
<span class="sd">                See: &lt;https://rdrr.io/cran/mediation/man/mediate.html&gt;</span>
<span class="sd">        padj (choice): The method for (ACME) p-value adjustment.</span>
<span class="sd">            - none: No p-value adjustment (no Padj column in outfile).</span>
<span class="sd">            - holm: Holm-Bonferroni method.</span>
<span class="sd">            - hochberg: Hochberg method.</span>
<span class="sd">            - hommel: Hommel method.</span>
<span class="sd">            - bonferroni: Bonferroni method.</span>
<span class="sd">            - BH: Benjamini-Hochberg method.</span>
<span class="sd">            - BY: Benjamini-Yekutieli method.</span>
<span class="sd">            - fdr: FDR correction method.</span>
<span class="sd">        cases (type=json): The cases for mediation analysis.</span>
<span class="sd">            Ignored if `in.fmlfile` is provided.</span>
<span class="sd">            A json/dict with case names as keys and values as a dict of M, Y, X, Cov, Model_M, Model_Y.</span>
<span class="sd">            For example:</span>
<span class="sd">            ```json</span>
<span class="sd">            {</span>
<span class="sd">                &quot;Case1&quot;: {</span>
<span class="sd">                    &quot;M&quot;: &quot;F1&quot;,</span>
<span class="sd">                    &quot;Y&quot;: &quot;F2&quot;,</span>
<span class="sd">                    &quot;X&quot;: &quot;F3&quot;,</span>
<span class="sd">                    &quot;Cov&quot;: &quot;F4,F5&quot;,</span>
<span class="sd">                    &quot;Model_M&quot;: &quot;glm&quot;,</span>
<span class="sd">                    &quot;Model_Y&quot;: &quot;lm&quot;</span>
<span class="sd">                },</span>
<span class="sd">                ...</span>
<span class="sd">            }</span>
<span class="sd">            ```</span>
<span class="sd">        transpose_input (flag): Whether to transpose the input file.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infile:file, fmlfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.infile | stem}}.mediation.txt&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;sims&quot;</span><span class="p">:</span> <span class="mi">1000</span><span class="p">,</span>
        <span class="s2">&quot;args&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;padj&quot;</span><span class="p">:</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span>
        <span class="s2">&quot;cases&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;transpose_input&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/stats/Mediation.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">LiquidAssoc</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.stats.LiquidAssoc"></span><a class="mkapi-docs-link" title="biopipen.ns.stats.LiquidAssoc" href="../../biopipen.ns.stats/#biopipen.ns.stats.LiquidAssoc">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Liquid association tests.</span>

<span class="sd">    See Also https://github.com/gundt/fastLiquidAssociation</span>
<span class="sd">    Requires https://github.com/pwwang/fastLiquidAssociation</span>

<span class="sd">    Input:</span>
<span class="sd">        infile: The input data file. The rows are samples and the columns are</span>
<span class="sd">            features. It must be tab-delimited.</span>
<span class="sd">            ```</span>
<span class="sd">            Sample   F1   F2   F3   ...   Fn</span>
<span class="sd">            S1       1.2  3.4  5.6        7.8</span>
<span class="sd">            S2       2.3  4.5  6.7        8.9</span>
<span class="sd">            ...</span>
<span class="sd">            Sm       5.6  7.8  9.0        1.2</span>
<span class="sd">            ```</span>
<span class="sd">            The features (columns) will be tested pairwise, which will be the X and</span>
<span class="sd">            Y columns in the result of `fastMLA`</span>
<span class="sd">        covfile: The covariate file. The rows are the samples and the columns</span>
<span class="sd">            are the covariates. It must be tab-delimited.</span>
<span class="sd">            If provided, the data in `in.infile` will be adjusted by covariates by</span>
<span class="sd">            regressing out the covariates and the residuals will be used for</span>
<span class="sd">            liquid association tests.</span>
<span class="sd">        groupfile: The group file. The rows are the samples and the columns</span>
<span class="sd">            are the groupings. It must be tab-delimited.</span>
<span class="sd">            ```</span>
<span class="sd">            Sample   G1   G2   G3   ...   Gk</span>
<span class="sd">            S1       0    1    0          0</span>
<span class="sd">            S2       2    1    0          NA  # exclude this sample</span>
<span class="sd">            ...</span>
<span class="sd">            Sm       1    0    0          0</span>
<span class="sd">            ```</span>
<span class="sd">            This will be served as the Z column in the result of `fastMLA`</span>
<span class="sd">            This can be omitted. If so, `envs.nvec` should be specified, which is</span>
<span class="sd">            to select column from `in.infile` as Z.</span>
<span class="sd">        fmlfile: The formula file. The 3 columns are X3, X12 and X21. The results</span>
<span class="sd">            will be filtered based on the formula. It must be tab-delimited without</span>
<span class="sd">            header.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output file.</span>
<span class="sd">            ```</span>
<span class="sd">            X12  X21  X3  rhodiff  MLA value  estimates  san.se  wald  Pval  model</span>
<span class="sd">            C38  C46  C5  0.87  0.32  0.67  0.20  10.87  0  F</span>
<span class="sd">            C46  C38  C5  0.87  0.32  0.67  0.20  10.87  0  F</span>
<span class="sd">            C27  C39  C4  0.94  0.34  1.22  0.38  10.03  0  F</span>
<span class="sd">            ```</span>

<span class="sd">    Envs:</span>
<span class="sd">        nvec: The column index (1-based) of Z in `in.infile`, if `in.groupfile` is</span>
<span class="sd">            omitted. You can specify multiple columns by comma-seperated values, or</span>
<span class="sd">            a range of columns by `-`. For example, `1,3,5-7,9`. It also supports</span>
<span class="sd">            column names. For example, `F1,F3`. `-` is not supported for column</span>
<span class="sd">            names.</span>
<span class="sd">        x: Similar as `nvec`, but limit X group to given features.</span>
<span class="sd">            The rest of features (other than X and Z) in `in.infile` will</span>
<span class="sd">            be used as Y.</span>
<span class="sd">            The features in `in.infile` will still be tested pairwise, but only</span>
<span class="sd">            features in X and Y will be kept.</span>
<span class="sd">        topn (type=int): Number of results to return by `fastMLA`, ordered from</span>
<span class="sd">            highest `|MLA|` value descending.</span>
<span class="sd">            The default of the package is 2000, but here we set to 1e6 to return as</span>
<span class="sd">            many results as possible (also good to do pvalue adjustment).</span>
<span class="sd">        rvalue (type=float): Tolerance value for LA approximation. Lower values of</span>
<span class="sd">            rvalue will cause a more thorough search, but take longer.</span>
<span class="sd">        cut (type=int): Value passed to the GLA function to create buckets</span>
<span class="sd">            (equal to number of buckets+1). Values placing between 15-30 samples per</span>
<span class="sd">            bucket are optimal. Must be a positive integer&gt;1. By default,</span>
<span class="sd">            `max(ceiling(nrow(data)/22), 4)` is used.</span>
<span class="sd">        ncores (type=int): Number of cores to use for parallelization.</span>
<span class="sd">        padj (choice): The method for p-value adjustment.</span>
<span class="sd">            - none: No p-value adjustment (no Padj column in outfile).</span>
<span class="sd">            - holm: Holm-Bonferroni method.</span>
<span class="sd">            - hochberg: Hochberg method.</span>
<span class="sd">            - hommel: Hommel method.</span>
<span class="sd">            - bonferroni: Bonferroni method.</span>
<span class="sd">            - BH: Benjamini-Hochberg method.</span>
<span class="sd">            - BY: Benjamini-Yekutieli method.</span>
<span class="sd">            - fdr: FDR correction method.</span>
<span class="sd">        transpose_input (flag): Whether to transpose the input file.</span>
<span class="sd">        transpose_group (flag): Whether to transpose the group file.</span>
<span class="sd">        transpose_cov (flag): Whether to transpose the covariate file.</span>
<span class="sd">        xyz_names: The names of X12, X21 and X3 in the final output file. Separated</span>
<span class="sd">            by comma. For example, `X12,X21,X3`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infile:file, covfile:file, groupfile:file, fmlfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.infile | stem}}.liquidassoc.txt&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;nvec&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;topn&quot;</span><span class="p">:</span> <span class="mf">1e6</span><span class="p">,</span>
        <span class="s2">&quot;rvalue&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span>
        <span class="s2">&quot;cut&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;padj&quot;</span><span class="p">:</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span>
        <span class="s2">&quot;transpose_input&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;transpose_group&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;transpose_cov&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;xyz_names&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/stats/LiquidAssoc.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">DiffCoexpr</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.stats.DiffCoexpr"></span><a class="mkapi-docs-link" title="biopipen.ns.stats.DiffCoexpr" href="../../biopipen.ns.stats/#biopipen.ns.stats.DiffCoexpr">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Differential co-expression analysis.</span>

<span class="sd">    See also &lt;https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-497&gt;</span>
<span class="sd">    and &lt;https://github.com/DavisLaboratory/dcanr/blob/8958d61788937eef3b7e2b4118651cbd7af7469d/R/inference_methods.R#L199&gt;.</span>

<span class="sd">    Input:</span>
<span class="sd">        infile: The input data file. The rows are samples and the columns are</span>
<span class="sd">            features. It must be tab-delimited.</span>
<span class="sd">            ```</span>
<span class="sd">            Sample   F1   F2   F3   ...   Fn</span>
<span class="sd">            S1       1.2  3.4  5.6        7.8</span>
<span class="sd">            S2       2.3  4.5  6.7        8.9</span>
<span class="sd">            ...</span>
<span class="sd">            Sm       5.6  7.8  9.0        1.2</span>
<span class="sd">            ```</span>
<span class="sd">        groupfile: The group file. The rows are the samples and the columns</span>
<span class="sd">            are the groupings. It must be tab-delimited.</span>
<span class="sd">            ```</span>
<span class="sd">            Sample   G1   G2   G3   ...   Gk</span>
<span class="sd">            S1       0    1    0          0</span>
<span class="sd">            S2       2    1    0          NA  # exclude this sample</span>
<span class="sd">            ...</span>
<span class="sd">            Sm       1    0    0          0</span>
<span class="sd">            ```</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output file. It is a tab-delimited file with the first</span>
<span class="sd">            column as the feature pair and the second column as the p-value.</span>
<span class="sd">            ```</span>
<span class="sd">            Group  Feature1  Feature2  Pval  Padj</span>
<span class="sd">            G1     F1        F2        0.123 0.123</span>
<span class="sd">            G1     F1        F3        0.123 0.123</span>
<span class="sd">            ...</span>
<span class="sd">            ```</span>

<span class="sd">    Envs:</span>
<span class="sd">        method (choice): The method used to calculate the differential</span>
<span class="sd">            co-expression.</span>
<span class="sd">            - pearson: Pearson correlation.</span>
<span class="sd">            - spearman: Spearman correlation.</span>
<span class="sd">        beta: The beta value for the differential co-expression analysis.</span>
<span class="sd">        padj (choice): The method for p-value adjustment.</span>
<span class="sd">            - none: No p-value adjustment (no Padj column in outfile).</span>
<span class="sd">            - holm: Holm-Bonferroni method.</span>
<span class="sd">            - hochberg: Hochberg method.</span>
<span class="sd">            - hommel: Hommel method.</span>
<span class="sd">            - bonferroni: Bonferroni method.</span>
<span class="sd">            - BH: Benjamini-Hochberg method.</span>
<span class="sd">            - BY: Benjamini-Yekutieli method.</span>
<span class="sd">            - fdr: FDR correction method.</span>
<span class="sd">        perm_batch (type=int): The number of permutations to run in each batch</span>
<span class="sd">        seed (type=int): The seed for random number generation</span>
<span class="sd">        ncores (type=int): The number of cores to use for parallelization</span>
<span class="sd">        transpose_input (flag): Whether to transpose the input file.</span>
<span class="sd">        transpose_group (flag): Whether to transpose the group file.</span>
<span class="sd">    &quot;&quot;&quot;</span>  <span class="c1"># noqa: E501</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infile:file, groupfile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.infile | stem}}.diffcoexpr.txt&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;pearson&quot;</span><span class="p">,</span>
        <span class="s2">&quot;beta&quot;</span><span class="p">:</span> <span class="mi">6</span><span class="p">,</span>
        <span class="s2">&quot;padj&quot;</span><span class="p">:</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span>
        <span class="s2">&quot;perm_batch&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span>
        <span class="s2">&quot;seed&quot;</span><span class="p">:</span> <span class="mi">8525</span><span class="p">,</span>
        <span class="s2">&quot;ncores&quot;</span><span class="p">:</span> <span class="n">config</span><span class="o">.</span><span class="n">misc</span><span class="o">.</span><span class="n">ncores</span><span class="p">,</span>
        <span class="s2">&quot;transpose_input&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;transpose_group&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/stats/DiffCoexpr.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">MetaPvalue</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.stats.MetaPvalue"></span><a class="mkapi-docs-link" title="biopipen.ns.stats.MetaPvalue" href="../../biopipen.ns.stats/#biopipen.ns.stats.MetaPvalue">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calulation of meta p-values.</span>

<span class="sd">    If there is only one input file, only the p-value adjustment will be performed.</span>

<span class="sd">    Input:</span>
<span class="sd">        infiles: The input files. Each file is a tab-delimited file with multiple</span>
<span class="sd">            columns. There should be ID column(s) to match the rows in other files and</span>
<span class="sd">            p-value column(s) to be combined. The records will be full-joined by ID.</span>
<span class="sd">            When only one file is provided, only the pvalue adjustment will be</span>
<span class="sd">            performed when `envs.padj` is not `none`, otherwise the input file will</span>
<span class="sd">            be copied to `out.outfile`.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output file. It is a tab-delimited file with the first column as</span>
<span class="sd">            the ID and the second column as the combined p-value.</span>
<span class="sd">            ```</span>
<span class="sd">            ID  ID1 ...  Pval   Padj</span>
<span class="sd">            a   x   ...  0.123  0.123</span>
<span class="sd">            b   y   ...  0.123  0.123</span>
<span class="sd">            ...</span>
<span class="sd">            ```</span>

<span class="sd">    Envs:</span>
<span class="sd">        id_cols: The column names used in all `in.infiles` as ID columns. Multiple</span>
<span class="sd">            columns can be specified by comma-seperated values. For example, `ID1,ID2`,</span>
<span class="sd">            where `ID1` is the ID column in the first file and `ID2` is the ID column</span>
<span class="sd">            in the second file.</span>
<span class="sd">            If `id_exprs` is specified, this should be a single column name for the new</span>
<span class="sd">            ID column in each `in.infiles` and the final `out.outfile`.</span>
<span class="sd">        id_exprs: The R expressions for each `in.infiles` to get ID column(s).</span>
<span class="sd">        pval_cols: The column names used in all `in.infiles` as p-value columns.</span>
<span class="sd">            Different columns can be specified by comma-seperated values for each</span>
<span class="sd">            `in.infiles`. For example, `Pval1,Pval2`.</span>
<span class="sd">        method (choice): The method used to calculate the meta-pvalue.</span>
<span class="sd">            - fisher: Fisher&#39;s method.</span>
<span class="sd">            - sumlog: Sum of logarithms (same as Fisher&#39;s method)</span>
<span class="sd">            - logitp: Logit method.</span>
<span class="sd">            - sumz: Sum of z method (Stouffer&#39;s method).</span>
<span class="sd">            - meanz: Mean of z method.</span>
<span class="sd">            - meanp: Mean of p method.</span>
<span class="sd">            - invt: Inverse t method.</span>
<span class="sd">            - sump: Sum of p method (Edgington&#39;s method).</span>
<span class="sd">            - votep: Vote counting method.</span>
<span class="sd">            - wilkinsonp: Wilkinson&#39;s method.</span>
<span class="sd">            - invchisq: Inverse chi-square method.</span>
<span class="sd">        keep_single (flag): Whether to keep the original p-value when there is only one</span>
<span class="sd">            p-value.</span>
<span class="sd">        na: The method to handle NA values. -1 to skip the record. Otherwise NA</span>
<span class="sd">            will be replaced by the given value.</span>
<span class="sd">        padj (choice): The method for p-value adjustment.</span>
<span class="sd">            - none: No p-value adjustment (no Padj column in outfile).</span>
<span class="sd">            - holm: Holm-Bonferroni method.</span>
<span class="sd">            - hochberg: Hochberg method.</span>
<span class="sd">            - hommel: Hommel method.</span>
<span class="sd">            - bonferroni: Bonferroni method.</span>
<span class="sd">            - BH: Benjamini-Hochberg method.</span>
<span class="sd">            - BY: Benjamini-Yekutieli method.</span>
<span class="sd">            - fdr: FDR correction method.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infiles:files&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.infiles | first | stem}}.metapval.txt&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;id_cols&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;id_exprs&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;pval_cols&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;fisher&quot;</span><span class="p">,</span>
        <span class="s2">&quot;na&quot;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
        <span class="s2">&quot;keep_single&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;padj&quot;</span><span class="p">:</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/stats/MetaPvalue.R&quot;</span>


<span class="k">class</span><span class="w"> </span><span class="nc">MetaPvalue1</span><span class="p">(</span><span class="n">Proc</span><span class="p">):</span><span id="biopipen.ns.stats.MetaPvalue1"></span><a class="mkapi-docs-link" title="biopipen.ns.stats.MetaPvalue1" href="../../biopipen.ns.stats/#biopipen.ns.stats.MetaPvalue1">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calulation of meta p-values.</span>

<span class="sd">    Unlike `MetaPvalue`, this process only accepts one input file.</span>

<span class="sd">    The p-values will be grouped by the ID columns and combined by the selected method.</span>

<span class="sd">    Input:</span>
<span class="sd">        infile: The input file.</span>
<span class="sd">            The file is a tab-delimited file with multiple</span>
<span class="sd">            columns. There should be ID column(s) to group the rows where</span>
<span class="sd">            p-value column(s) to be combined.</span>

<span class="sd">    Output:</span>
<span class="sd">        outfile: The output file. It is a tab-delimited file with the first column as</span>
<span class="sd">            the ID and the second column as the combined p-value.</span>
<span class="sd">            ```</span>
<span class="sd">            ID  ID1 ...  Pval   Padj</span>
<span class="sd">            a   x   ...  0.123  0.123</span>
<span class="sd">            b   y   ...  0.123  0.123</span>
<span class="sd">            ...</span>
<span class="sd">            ```</span>

<span class="sd">    Envs:</span>
<span class="sd">        id_cols: The column names used in `in.infile` as ID columns. Multiple</span>
<span class="sd">            columns can be specified by comma-seperated values. For example, `ID1,ID2`.</span>
<span class="sd">        pval_col: The column name used in `in.infile` as p-value column.</span>
<span class="sd">        method (choice): The method used to calculate the meta-pvalue.</span>
<span class="sd">            - fisher: Fisher&#39;s method.</span>
<span class="sd">            - sumlog: Sum of logarithms (same as Fisher&#39;s method)</span>
<span class="sd">            - logitp: Logit method.</span>
<span class="sd">            - sumz: Sum of z method (Stouffer&#39;s method).</span>
<span class="sd">            - meanz: Mean of z method.</span>
<span class="sd">            - meanp: Mean of p method.</span>
<span class="sd">            - invt: Inverse t method.</span>
<span class="sd">            - sump: Sum of p method (Edgington&#39;s method).</span>
<span class="sd">            - votep: Vote counting method.</span>
<span class="sd">            - wilkinsonp: Wilkinson&#39;s method.</span>
<span class="sd">            - invchisq: Inverse chi-square method.</span>
<span class="sd">        na: The method to handle NA values. -1 to skip the record. Otherwise NA</span>
<span class="sd">            will be replaced by the given value.</span>
<span class="sd">        keep_single (flag): Whether to keep the original p-value when there is only one</span>
<span class="sd">            p-value.</span>
<span class="sd">        padj (choice): The method for p-value adjustment.</span>
<span class="sd">            - none: No p-value adjustment (no Padj column in outfile).</span>
<span class="sd">            - holm: Holm-Bonferroni method.</span>
<span class="sd">            - hochberg: Hochberg method.</span>
<span class="sd">            - hommel: Hommel method.</span>
<span class="sd">            - bonferroni: Bonferroni method.</span>
<span class="sd">            - BH: Benjamini-Hochberg method.</span>
<span class="sd">            - BY: Benjamini-Yekutieli method.</span>
<span class="sd">            - fdr: FDR correction method.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;infile:file&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;outfile:file:{{in.infile | stem}}.metapval.txt&quot;</span>
    <span class="n">lang</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">lang</span><span class="o">.</span><span class="n">rscript</span>
    <span class="n">envs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;id_cols&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;pval_col&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;fisher&quot;</span><span class="p">,</span>
        <span class="s2">&quot;na&quot;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
        <span class="s2">&quot;keep_single&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;padj&quot;</span><span class="p">:</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">script</span> <span class="o">=</span> <span class="s2">&quot;file://../scripts/stats/MetaPvalue1.R&quot;</span>
</code></pre></div>
</div>
</div>
              
            </div>
          </div><footer>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  <div class="footer-notes">
    <a href="https://github.com/pwwang/biopipen" title="Open repository in GitHub">Github Repo</a>
    <div class="footer-sep"></div>Built with <a href="https://www.mkdocs.org/">MkDocs</a> using theme <a href="https://github.com/pwwang/mkdocs-rtd">mkdocs-rtd</a>.
  </div>
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
  </span>
</div>
    <script src="../../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../../..";</script>
    <script src="../../../js/theme_extra.js"></script>
    <script src="../../../js/theme.js"></script>
    <script src="../../../js/version-select.js"></script>
      <script src="../../../js/mkapi.js"></script>
      <script src="../../../extra.js"></script>
      <script src="../../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
