## 0.1.9

- ğŸ› Load `all_config_annotations.csv` if `filtered_contig_annotations.csv` doesn't exist for `tcr.ImmunarchLoad`
- ğŸ› Calculate diversity for all clones only if filtering by clone sizes failed for `tcr.ImmunarchAdvanced`
- ğŸš‘ Fix seurat object creating when expressions are named "Gene Expression" for scrna.SeuratPreparing
- âœ¨ Add `tcr.TCRClustering`
- âœ¨ Add `raw` to immdata for `tcr.immunarchLoading`
- âœ¨ Add `on_raw` env to `tcr.TCRClustering`
- âœ¨ Add `bam.ControlFREEC`

## 0.1.8

- âœ¨ Add tcr.Attach2Seurat

## 0.1.7

- â• Add datar dep for scrna_metabolic pipeline
- ğŸš‘ Fix scrna_metabolic.MetabolicPathwayActivity
- âœ¨ Add bcftools.BcftoolsFilter
- ğŸ‘½ï¸ Don't wrap job report in `report_jobs` report macro (to adopt pipen-report 0.2)
- âœ¨ Add more options for scrna.DimPlots

## 0.1.6

- âœ¨ Convert CNVpytor results to gff and bed
- ğŸš‘ Make scrna_metabolic pipeline work standalone
- â• Add datar dep for scrna_metabolic pipeline
- ğŸš‘ Fix scrna_metabolic.MetabolicPathwayActivity
- âœ¨ Add bcftools.BcftoolsFilter

## 0.1.5

- âœ¨ Add features and fix issues for immunopipe 0.0.4
- âœ¨ Add some vcf processes

## 0.1.4

- ğŸ› Fix bam.CNVpytor when snpfile is not provided
- âœ¨ Add metabolic pathway analysis for single-cell RNA-seq data

## 0.1.3

- Add gsea.GSEA and scrna.SCImpute
- Add gene name conversions
- Add gsea.FGSEA
- Add venn plots and refactor ImmunarchFilter
- Add plot.Heatmap
- Reuse plot.Heatmap for scrna.GeneExpressionInvestigation
- Attach metadata to seurat object in scrna.SeuratPreparing
- Add envs.group_subset for scrna.GeneExpressionInvestigation
- Fix typo for scrna.GeneExpressionInvestigation
- Add docs


## 0.1.2

- âœ¨ Add envs.qc for scrna.SeuratPreparing

## 0.1.1

- Finish processes for immunopipe

## 0.1.0

- Adopt pipen 0.2+
