## 0.4.6

- âœ¨  [vcf.TruvariBench] Allow `multimatch` to be passed
- âœ¨  [vcf.TruvariConsistency] Add report

## 0.4.5

- âœ¨ [bam.CNVpytor] Generate and fix VCF file as result
- ğŸ“ [vcf.TruvariBench] Update docs to show other arguments for `truvari bench`
- âœ¨ [vcf.TruvariBench] Allow `sizemax` to be passed
- âœ¨ [bed.BedConsensus] Add process and tests
- âœ¨ [core] Add `ref.genome` to configurations
- âš¡ï¸ [bed.BedConsensus] Parallelize and speed up
- ğŸ’š [test] Add bedtools to env `bio`
- ğŸ’š [test] Add chromsome sizes to reference
- ğŸ’š [test] Add r-gsea_r to env `r`
- ğŸ’š [scrna.ScFGSEA] Fix testsâ

## 0.4.4

- ğŸ› [scrna.SeuratPreparing] Fix after tidyseurat being used
- ğŸ› [scrna.SeuratPreparing] Fix object `Sample` not found
- ğŸ“ [Housekeeping] Fix API docs
- ğŸ“ [Housekeeping] Make apis show neater docs

## 0.4.3

- âœ¨ [scrna] Add `filter` for cases in CellsDistribution, MarkersFinder and ScFGSEA
- âœ¨ [utils] Allow gg object for ggs in plot.R
- ğŸ› [scrna_metabolic] Fix reports
- ğŸ› [scrna_metabolic] Fix multiple cases
- ğŸ› [scrna_metabolic] Fix rmagic for normalization
- âš¡ï¸ [scrna.SeuratClusterStats] Add common gene list
- âš¡ï¸ [scrna.MarkersFinder] Add `filter2` to filter after mutaters
- ğŸ› [tcr.Immunarch] Fix missing library tibble in script
- âš¡ï¸ [scrna.ScFGSEA] Make ident hierarchical

## 0.4.2

- ğŸ’š [Housekeeping] Fix CI deploy
- âš¡ï¸ [processes] Use faster do_call() instead of do.call()
- ğŸ“ [tcr] Fix some docstrings with `{{` and `}}`
- âœ… [vcf.TruvariBench] Add ref for test
- ğŸ©¹ [tcr.TCRClustering] FIx VGeneScores.txt being generated in current dir
- ğŸ“ [scrna.SeuratPreparing] Update docstring and refactor script
- âœ¨ [scrna.SeuratClustering] Allow dims to be expanded in arguments
- ğŸ“ [scrna.MarkersFinder] Adopt reduced case configuration level

## 0.4.1

### General
- ğŸ‘· [Housekeeping] Add deploy in CI
- ğŸšš [Housekeeping] Move tests/test_tcr/TCRClustering to tests/test_tcr/TCRClusteringStats
- ğŸ”§ [Tests] Add r-tidyseurat to env_r.toml

### Processes
- ğŸ©¹ [scrna.CellsDistribution] Reduce envs.cases levels
- ğŸ©¹ [scrna.CellsDistribution] Allow acurate sizes to be used in orderby
- ğŸ©¹ [scrna.ScFGSEA] Reduce envs.cases levels
- âœ¨ [scrna.ScFGSEA] Allow `{ident}` or `{cluster}` as placeholder in cases
- âœ¨ [scrna.SeuratClusterStats] Add dimplots
- ğŸš‘ [scrna.SeuratClusterStats] Limit 20 genes by default
- ğŸ› [tcr.ImmunarchLoading] Fix multiple "Source" columns in data
- ğŸ©¹ [tcr.TCRClustering] Make clusterfile as a meta file that can be used by SeuratMetadataMutater
- âœ¨ [tcr.TCRClusteringStats] Add shared clusters by grouping
- ğŸ“ [tcr.TCRClusteringStats] Don't show shared TCR clusters between groups if not configured
- ğŸ“ [gsea.FGSEA] Limit pagesize to 10 in report
- âœ¨ [vcf.TruvariBenchSummary] Add process and test
- âœ¨ [vcf.TruvariBenchSummary] Add default `input_data`
- âœï¸ [bed.Bed2Vcf] Fix typos in doc
- âœ¨ [bed.Bed2Vcf] Allow records to be skipped
- âœ… [vcf.TruvariBench] Add ref for test

## 0.4.0

- âœ¨ [scrna.CellsDistribution] Add process and test
- ğŸ—‘ï¸ Remove `namespaces` (use `ns` instead)

## 0.3.2

- âœ… Allow tests to run locally only
- ğŸ’š Add pipen-args for tests
- âœ… [plot.Heatmap] Fix test
- âœ… [pipeline.scrna_metabolic] Add ARGS in run.env
- âœ… [scrna.ScFGSEA] Add test
- âœ¨ [tcr.TCRClusteringStats] Add process
- âœ… [tcr.TCRClustering] Use env r for testing
- âœ… [tcr.TCRClustering] Add test
- âœ… [pipeline.scrna_metabolic] Add test
- âœ… [gsea.GSEA] Add tests
- âœ… [gsea.FGSEA] Add tests
- âœ… [plot.Heatmap] Add tests
- âœ… [gene.GeneNameConversion] Add tests
- âœ… [utils.gene] Add tests
- ğŸ’š [bed.Bed2Vcf] Fix test
- âœ… [vcf.VcfFix] Add test
- âœ… [misc.File2Proc] Use base container for test
- âœ… [misc.File2Proc] Fix test
- ğŸ©¹ [scrna.ExprImpute] Use if-statement for requirements
- âœ¨ [scrna.SeuratClusterStats] Add process and test

## 0.3.1

- ğŸ—‘ï¸ Deprecate `biopipen.namespaces`, use `biopipen.ns` instead
- âœ¨ [bed.Bed2Vcf] Add bed.Bed2Vcf
- âœ¨ [vcf.VcfFix] Add vcf.VcfFix
- ğŸ› [vcf.vcfFix] Fix when a flag in INFO
- âœ¨ [vcf.TruvariBench] Add vcf.TruvariBench
- âœ¨ [vcf.TruvariConsistency] Add vcf.TruvariConsistency
- ğŸ› [utils.reference] Fix typo in tabix_index
- ğŸ› [vcf.VcfIndex] Fix vcf.VcfIndex
- âœ¨ [bed.Bed2Vcf] Allow to ignore non-existing contigs and index the output file
- âœ¨ [misc.Shell] Add misc.Shell to run a shell command

## 0.3.0

- â™»ï¸ Refactor some processes for immunopipe
- ğŸ©¹ [scrna.SeuratPreparing] Remove tmp datadir for scrna.SeuratPreparing if exsits
- ğŸ©¹ [scrna.SeuratPreparing] Add a TODO comment in scrna.SeuratPreparing (#26)
- âœ¨ [scrna.Subset10X] Add `scrna.Subset10X`
- ğŸ’¥ [tcr.Immunarch] Merge `tcr.ImmunarchBasic` and `tcr.ImmunarchAdvanced` into `tcr.Immunarch`
- ğŸ©¹ [tcr.VJUsage] Fix R script being generated at current direct for `tcr.VJUsage`
- âœ¨ [scrna.SeuratMetadataMutater] Add `scrna.SeuratMetadataMutater`
- ğŸ› [tcr.Immunarch] Fix clonotype tracking not selecting top clones by given top N
- â™»ï¸ [pipeline.scrna_metabolic] Refactor scrna_metabolic
- ğŸ“ [pipeline.scrna_metabolic] Update docs for scrna_metabolic pipeline
- âœ¨ [pipeline.scrna_metabolic] Allow scrna_metabolic pipeline to handle multiple cases
- ğŸš‘ [scrna.ExprImpute] Fix reticulate not using right python
- ğŸš‘ [scrna.SeuratMetadataMutater] Fix error when input mutaters in None
- ğŸš‘ [scrna_metabolic.MetabolicInputs] Fix diot not imported in script

## 0.2.1

- User rtoml over toml

## 0.2.0

- ğŸ“Œ Pin deps for docs
- Don't link non-existing files for misc.Glob2Dir
- Upgrade datar to 0.8
- â¬†ï¸ Upgrade pipen to v0.3
- âš¡ï¸ Load 10X TCR and RNA-seq data files more robustly for scrna.SeuratPreparing and tcr.ImmunarchLoading


## 0.1.9

- ğŸ› Load `all_config_annotations.csv` if `filtered_contig_annotations.csv` doesn't exist for `tcr.ImmunarchLoad`
- ğŸ› Calculate diversity for all clones only if filtering by clone sizes failed for `tcr.ImmunarchAdvanced`
- ğŸš‘ Fix seurat object creating when expressions are named "Gene Expression" for scrna.SeuratPreparing
- âœ¨ Add `tcr.TCRClustering`
- âœ¨ Add `raw` to immdata for `tcr.immunarchLoading`
- âœ¨ Add `on_raw` env to `tcr.TCRClustering`
- âœ¨ Add `bam.ControlFREEC`

## 0.1.8

- âœ¨ Add tcr.Attach2Seurat

## 0.1.7

- â• Add datar dep for scrna_metabolic pipeline
- ğŸš‘ Fix scrna_metabolic.MetabolicPathwayActivity
- âœ¨ Add bcftools.BcftoolsFilter
- ğŸ‘½ï¸ Don't wrap job report in `report_jobs` report macro (to adopt pipen-report 0.2)
- âœ¨ Add more options for scrna.DimPlots

## 0.1.6

- âœ¨ Convert CNVpytor results to gff and bed
- ğŸš‘ Make scrna_metabolic pipeline work standalone
- â• Add datar dep for scrna_metabolic pipeline
- ğŸš‘ Fix scrna_metabolic.MetabolicPathwayActivity
- âœ¨ Add bcftools.BcftoolsFilter

## 0.1.5

- âœ¨ Add features and fix issues for immunopipe 0.0.4
- âœ¨ Add some vcf processes

## 0.1.4

- ğŸ› Fix bam.CNVpytor when snpfile is not provided
- âœ¨ Add metabolic pathway analysis for single-cell RNA-seq data

## 0.1.3

- Add gsea.GSEA and scrna.SCImpute
- Add gene name conversions
- Add gsea.FGSEA
- Add venn plots and refactor ImmunarchFilter
- Add plot.Heatmap
- Reuse plot.Heatmap for scrna.GeneExpressionInvestigation
- Attach metadata to seurat object in scrna.SeuratPreparing
- Add envs.group_subset for scrna.GeneExpressionInvestigation
- Fix typo for scrna.GeneExpressionInvestigation
- Add docs


## 0.1.2

- âœ¨ Add envs.qc for scrna.SeuratPreparing

## 0.1.1

- Finish processes for immunopipe

## 0.1.0

- Adopt pipen 0.2+
