<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../img/favicon.ico" />
    <title>biopipen.utils.gene - biopipen</title>
    <link rel="stylesheet" href="../../../css/theme.css" />
    <link rel="stylesheet" href="../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="//use.fontawesome.com/releases/v5.8.1/css/all.css" rel="stylesheet" />
        <link href="//use.fontawesome.com/releases/v5.8.1/css/v4-shims.css" rel="stylesheet" />
        <link href="../../../css/mkapi-common.css" rel="stylesheet" />
        <link href="../../../style.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "biopipen.utils.gene";
        var mkdocs_page_input_path = "api/source/biopipen.utils.gene.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.configure({ ignoreUnescapedHTML: true }); hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../..">
          <img src="../../../img/logo.png" class="logo" alt="Logo"/>
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="nav-scrollable-wrapper">
      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Pipelines</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../../pipelines/cnvkit_pipeline/">CNVkit Pipeline</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../../pipelines/cellranger_pipeline/">CellRanger Pipeline</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">API</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.utils</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils/">biopipen.utils</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.gene/">biopipen.utils.gene</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.reference/">biopipen.utils.reference</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.misc/">biopipen.utils.misc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.utils.vcf/">biopipen.utils.vcf</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.core</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core/">biopipen.core</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.proc/">biopipen.core.proc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.filters/">biopipen.core.filters</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.testing/">biopipen.core.testing</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.config/">biopipen.core.config</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.core.defaults/">biopipen.core.defaults</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">biopipen.ns</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns/">biopipen.ns</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.gene/">biopipen.ns.gene</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cellranger/">biopipen.ns.cellranger</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.scrna_metabolic_landscape/">biopipen.ns.scrna_metabolic_landscape</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.protein/">biopipen.ns.protein</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.delim/">biopipen.ns.delim</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.rnaseq/">biopipen.ns.rnaseq</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.plot/">biopipen.ns.plot</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.tcgamaf/">biopipen.ns.tcgamaf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.regulatory/">biopipen.ns.regulatory</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.stats/">biopipen.ns.stats</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cellranger_pipeline/">biopipen.ns.cellranger_pipeline</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.bed/">biopipen.ns.bed</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnv/">biopipen.ns.cnv</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.snp/">biopipen.ns.snp</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.web/">biopipen.ns.web</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.misc/">biopipen.ns.misc</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnvkit/">biopipen.ns.cnvkit</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.cnvkit_pipeline/">biopipen.ns.cnvkit_pipeline</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.tcr/">biopipen.ns.tcr</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.scrna/">biopipen.ns.scrna</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.vcf/">biopipen.ns.vcf</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.gsea/">biopipen.ns.gsea</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../biopipen.ns.bam/">biopipen.ns.bam</a>
                </li>
    </ul>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../../CHANGELOG/">Change log</a>
                </li>
              </ul>
      </div>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../..">biopipen</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../.." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">biopipen.utils.gene</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/pwwang/biopipen/edit/master/docs/api/source/biopipen.utils.gene.md">Edit on pwwang/biopipen</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="mkapi-node">
<h1 id="" class="mkapi-object mkapi-object-code plain">
<span class="mkapi-object-kind mkapi-object-kind-code">SOURCE CODE</span>
<span class="mkapi-object-prefix">biopipen.utils.</span><span class="mkapi-object-name">gene</span>
<span id="biopipen.utils.gene"></span><a class="mkapi-docs-link" href="../../biopipen.utils.gene">DOCS</a>
</h1>

<div class="mkapi-code">
<div class="codehilite"><pre><span></span><code><span class="sd">&quot;&quot;&quot;Do gene name conversion&quot;&quot;&quot;</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">__future__</span><span class="w"> </span><span class="kn">import</span> <span class="n">annotations</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">re</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">contextlib</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">mygene</span><span class="w"> </span><span class="kn">import</span> <span class="n">MyGeneInfo</span>

<span class="n">mygene</span> <span class="o">=</span> <span class="n">MyGeneInfo</span><span class="p">()</span>


<span class="k">class</span><span class="w"> </span><span class="nc">QueryGenesNotFound</span><span class="p">(</span><span class="ne">ValueError</span><span class="p">):</span><span id="biopipen.utils.gene.QueryGenesNotFound"></span><a class="mkapi-docs-link" title="biopipen.utils.gene.QueryGenesNotFound" href="../../biopipen.utils.gene/#biopipen.utils.gene.QueryGenesNotFound">DOCS</a>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;When genes cannot be found&quot;&quot;&quot;</span>


<span class="k">def</span><span class="w"> </span><span class="nf">gene_name_conversion</span><span class="p">(</span><span id="biopipen.utils.gene.gene_name_conversion"></span><a class="mkapi-docs-link" title="biopipen.utils.gene.gene_name_conversion" href="../../biopipen.utils.gene/#biopipen.utils.gene.gene_name_conversion">DOCS</a>
    <span class="n">genes</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">infmt</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">outfmt</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">dup</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;first&quot;</span><span class="p">,</span>
    <span class="n">species</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;human&quot;</span><span class="p">,</span>
    <span class="n">notfound</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;na&quot;</span><span class="p">,</span>
    <span class="n">suppress_messages</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert gene names using MyGeneInfo</span>

<span class="sd">    Args:</span>
<span class="sd">        genes: A character/integer vector of gene names/ids</span>
<span class="sd">        species: A character vector of species names</span>
<span class="sd">        infmt: A character vector of input gene name formats</span>
<span class="sd">            See the available scopes at</span>
<span class="sd">            https://docs.mygene.info/en/latest/doc/data.html#available-fields</span>
<span class="sd">            You can use ensg as a shortcut for ensembl.gene</span>
<span class="sd">        outfmt: A character vector of output gene name formats</span>
<span class="sd">        dup: How to deal with duplicate gene names found.</span>
<span class="sd">            first: keep the first one (default), sorted by score descendingly</span>
<span class="sd">            last: keep the last one, sorted by score descendingly</span>
<span class="sd">            all: keep all of them, each will be a separate row</span>
<span class="sd">            &lt;X&gt;: combine them into a single string, separated by X</span>
<span class="sd">        notfound: How to deal with gene names that are not found</span>
<span class="sd">            error: stop with an error message</span>
<span class="sd">            use-query: use the query gene name as the converted gene name</span>
<span class="sd">            skip: skip the gene names that are not found</span>
<span class="sd">            ignore: Same as &quot;skip&quot;</span>
<span class="sd">            na: use NA as the converted gene name (default)</span>
<span class="sd">        suppress_messages: Suppress the messages while querying</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dataframe with the query gene names and the converted gene names</span>
<span class="sd">        When a gene name is not found, the converted name will be &quot;NA&quot;</span>
<span class="sd">        When duplicate gene names are found, the one with the highest score will be kept</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">notfound</span> <span class="o">=</span> <span class="n">notfound</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">notfound</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;error&quot;</span><span class="p">,</span> <span class="s2">&quot;use-query&quot;</span><span class="p">,</span> <span class="s2">&quot;skip&quot;</span><span class="p">,</span> <span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;`notfound` of `gene_name_conversion` must be one of &quot;</span>
            <span class="s2">&quot;&#39;error&#39;, &#39;use-query&#39;, &#39;skip&#39;, &#39;ignore&#39;, &#39;na&#39;&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">infmt</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;ensg&quot;</span><span class="p">,</span> <span class="s2">&quot;ensmusg&quot;</span><span class="p">]:</span>
        <span class="n">infmt</span> <span class="o">=</span> <span class="s2">&quot;ensembl.gene&quot;</span>
    <span class="k">if</span> <span class="n">outfmt</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;ensg&quot;</span><span class="p">,</span> <span class="s2">&quot;ensmusg&quot;</span><span class="p">]:</span>
        <span class="n">outfmt</span> <span class="o">=</span> <span class="s2">&quot;ensembl.gene&quot;</span>

    <span class="n">orig_genes</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[:]</span>
    <span class="k">if</span> <span class="n">infmt</span> <span class="o">==</span> <span class="s2">&quot;ensembl.gene&quot;</span><span class="p">:</span>
        <span class="c1"># Remove version numbers from ensembl gene ids</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\\</span><span class="s2">..*&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">]</span>

    <span class="n">query_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;query&quot;</span><span class="p">:</span> <span class="n">genes</span><span class="p">,</span> <span class="s2">&quot;orig&quot;</span><span class="p">:</span> <span class="n">orig_genes</span><span class="p">})</span>

    <span class="k">if</span> <span class="n">suppress_messages</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">contextlib</span><span class="o">.</span><span class="n">redirect_stdout</span><span class="p">(</span><span class="kc">None</span><span class="p">):</span>
            <span class="n">out</span> <span class="o">=</span> <span class="n">mygene</span><span class="o">.</span><span class="n">querymany</span><span class="p">(</span>
                <span class="n">genes</span><span class="p">,</span>
                <span class="n">scopes</span><span class="o">=</span><span class="n">infmt</span><span class="p">,</span>
                <span class="n">fields</span><span class="o">=</span><span class="n">outfmt</span><span class="p">,</span>
                <span class="n">species</span><span class="o">=</span><span class="n">species</span><span class="p">,</span>
                <span class="n">as_dataframe</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">df_index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">mygene</span><span class="o">.</span><span class="n">querymany</span><span class="p">(</span>
            <span class="n">genes</span><span class="p">,</span>
            <span class="n">scopes</span><span class="o">=</span><span class="n">infmt</span><span class="p">,</span>
            <span class="n">fields</span><span class="o">=</span><span class="n">outfmt</span><span class="p">,</span>
            <span class="n">species</span><span class="o">=</span><span class="n">species</span><span class="p">,</span>
            <span class="n">as_dataframe</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">df_index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">out</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;query&quot;</span><span class="p">:</span> <span class="n">genes</span><span class="p">,</span> <span class="s2">&quot;converted&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;NA&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)})</span>

    <span class="k">if</span> <span class="n">dup</span> <span class="o">==</span> <span class="s2">&quot;first&quot;</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">out</span>
            <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;_score&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;query&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">dup</span> <span class="o">==</span> <span class="s2">&quot;last&quot;</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">out</span>
            <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;_score&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;query&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">dup</span> <span class="o">!=</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">out</span>
            <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;_score&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;query&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="n">outfmt</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">dup</span><span class="si">}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">unique</span><span class="p">()])})</span>
            <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="p">)</span>

    <span class="n">out</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">query_df</span><span class="p">,</span> <span class="n">out</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;query&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;query&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;orig&quot;</span><span class="p">:</span> <span class="s2">&quot;query&quot;</span><span class="p">})</span>

    <span class="k">if</span> <span class="n">notfound</span> <span class="o">==</span> <span class="s2">&quot;error&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">out</span><span class="p">[</span><span class="n">outfmt</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
            <span class="n">nagenes</span> <span class="o">=</span> <span class="n">out</span><span class="p">[</span><span class="n">out</span><span class="p">[</span><span class="n">outfmt</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()][</span><span class="s2">&quot;query&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
            <span class="k">raise</span> <span class="n">QueryGenesNotFound</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Query genes not found: </span><span class="si">{</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">nagenes</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">notfound</span> <span class="o">==</span> <span class="s2">&quot;use-query&quot;</span><span class="p">:</span>
        <span class="n">out</span><span class="p">[</span><span class="n">outfmt</span><span class="p">]</span> <span class="o">=</span> <span class="n">out</span><span class="p">[</span><span class="n">outfmt</span><span class="p">]</span><span class="o">.</span><span class="n">combine_first</span><span class="p">(</span><span class="n">out</span><span class="p">[</span><span class="s2">&quot;query&quot;</span><span class="p">])</span>
    <span class="k">elif</span> <span class="n">notfound</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;skip&quot;</span><span class="p">,</span> <span class="s2">&quot;ignore&quot;</span><span class="p">]:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="n">outfmt</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>  <span class="c1"># notfound == &quot;na&quot;</span>
        <span class="n">out</span><span class="p">[</span><span class="n">outfmt</span><span class="p">]</span> <span class="o">=</span> <span class="n">out</span><span class="p">[</span><span class="n">outfmt</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;NA&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">out</span>
</code></pre></div>
</div>
</div>
              
            </div>
          </div><footer>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  <div class="footer-notes">
    <a href="https://github.com/pwwang/biopipen" title="Open repository in GitHub">Github Repo</a>
    <div class="footer-sep"></div>Built with <a href="https://www.mkdocs.org/">MkDocs</a> using theme <a href="https://github.com/pwwang/mkdocs-rtd">mkdocs-rtd</a>.
  </div>
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
  </span>
</div>
    <script src="../../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../../..";</script>
    <script src="../../../js/theme_extra.js"></script>
    <script src="../../../js/theme.js"></script>
    <script src="../../../js/version-select.js"></script>
      <script src="../../../js/mkapi.js"></script>
      <script src="../../../extra.js"></script>
      <script src="../../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
